{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fantastic-leeds",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sys\n",
    "import os\n",
    "import re\n",
    "import csv\n",
    "import collections as co\n",
    "import itertools as it\n",
    "import umap\n",
    "import hdbscan\n",
    "import time \n",
    "import progressbar\n",
    "import math\n",
    "import random\n",
    "import fileinput\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.transforms as transforms\n",
    "import scipy.spatial.distance as ssd\n",
    "import seaborn as sns\n",
    "\n",
    "from Bio import Align, AlignIO, Phylo\n",
    "from Bio.Align.Applications import MafftCommandline\n",
    "from Bio.Seq import Seq \n",
    "from Bio.Phylo.TreeConstruction import DistanceTreeConstructor, DistanceMatrix, DistanceCalculator\n",
    "from ete3 import PhyloTree, Tree, faces, AttrFace, CircleFace, TreeStyle, NodeStyle, TextFace, SequenceFace\n",
    "from IPython.display import display\n",
    "from matplotlib import colors \n",
    "from scipy.cluster import hierarchy\n",
    "from sklearn.preprocessing import normalize\n",
    "from sklearn.decomposition import PCA\n",
    "from statistics import mean\n",
    "from scipy import stats\n",
    "from kneed import DataGenerator, KneeLocator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "hybrid-gasoline",
   "metadata": {},
   "outputs": [],
   "source": [
    "class frequency(object):\n",
    "    \n",
    "    def __init__(self, k = 7, convert = 0):\n",
    "    \n",
    "        self.k = k\n",
    "        self.convert = convert\n",
    "        self.exist = co.defaultdict(int) \n",
    "        self.keys = list(self.exist.keys())\n",
    "        self.col = len(self.keys)\n",
    "        self.row = 0\n",
    "        self.matrix = np.empty((self.row, self.col, ),dtype = \"float64\")\n",
    "        self.amino = co.defaultdict(str, {\n",
    "            'AAA':'K', 'AAC':'N', 'AAG':'K', 'AAT':'N',\n",
    "            'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "            'AGA':'R', 'AGC':'S', 'AGG':'R', 'AGT':'S',\n",
    "            'ATA':'I', 'ATC':'I', 'ATG':'M', 'ATT':'I',\n",
    "            'CAA':'Q', 'CAC':'H', 'CAG':'Q', 'CAT':'H',\n",
    "            'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "            'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "            'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "            'GAA':'E', 'GAC':'D', 'GAG':'E', 'GAT':'D',\n",
    "            'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "            'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "            'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',    \n",
    "            'TAA':'Y', 'TAC':'*', 'TAG':'*', 'TAT':'Y',\n",
    "            'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "            'TGA':'*', 'TGC':'C', 'TGG':'W', 'TGT':'C',\n",
    "            'TTA':'L', 'TTC':'F', 'TTG':'L', 'TTT':'F'\n",
    "        })\n",
    "                \n",
    "    def translate(self, read):\n",
    "    \n",
    "        chain = ''\n",
    "\n",
    "        for i in range(len(read) - 2):\n",
    "            trip = read[i:i+3]\n",
    "            chain += self.amino[trip]\n",
    "\n",
    "        return(chain)\n",
    "    \n",
    "    \n",
    "    def adjust_to_data(self, infile):\n",
    "    \n",
    "        self.row = infile.shape[0]\n",
    "            \n",
    "        for line, read in infile.itertuples(index=True, name=None):\n",
    "\n",
    "            if self.convert == 1:\n",
    "                seq = self.translate(read)\n",
    "                del read\n",
    "\n",
    "                num = len(seq) - self.k + 1\n",
    "\n",
    "                for i in range(num):\n",
    "                    kmer = seq[i:i+self.k]\n",
    "                    self.exist[kmer] = 0\n",
    "\n",
    "            else:\n",
    "                seq = read\n",
    "                del read\n",
    "\n",
    "                num = len(seq) - self.k + 1\n",
    "\n",
    "                if re.match('^[ACGT]*$', seq): \n",
    "                    for i in range(num):\n",
    "                        kmer = seq[i:i+self.k]\n",
    "                        self.exist[kmer] = 0\n",
    "                else:\n",
    "                    for i in range(num):\n",
    "                        kmer = seq[i:i+self.k]\n",
    "                        if re.match('^[ACGT]*$', kmer): \n",
    "                            self.exist[kmer] = 0\n",
    "            \n",
    "        self.keys = list(self.exist.keys())\n",
    "        self.col = len(self.keys)\n",
    "        self.matrix = np.empty((self.row, self.col, ), dtype=\"float64\")\n",
    "        \n",
    "        del seq\n",
    "    \n",
    "    \n",
    "    def calculate_frequence(self, infile):\n",
    "        \n",
    "        for line, read in infile.itertuples(index=True, name=None): \n",
    "                 \n",
    "            if self.convert == 1:\n",
    "                seq = self.translate(read)\n",
    "                del read\n",
    "\n",
    "                counts = self.exist.copy()\n",
    "                num = len(seq) - self.k + 1\n",
    "\n",
    "                for i in range(num):\n",
    "                    kmer = seq[i:i+self.k]\n",
    "                    counts[kmer] += 1\n",
    "\n",
    "            else:\n",
    "                seq = read\n",
    "                del read\n",
    "\n",
    "                counts = self.exist.copy()\n",
    "                num = len(seq) - self.k + 1\n",
    "\n",
    "                if re.match('^[ACGT]*$', seq): \n",
    "                    for i in range(num):\n",
    "                        kmer = seq[i:i+self.k]\n",
    "                        counts[kmer] += 1\n",
    "                else:\n",
    "                    for i in range(num):\n",
    "                        kmer = seq[i:i+self.k]\n",
    "                        if re.match('^[ACGT]*$', kmer): \n",
    "                            counts[kmer] += 1\n",
    "\n",
    "            vector = np.array(list(counts.values()), dtype = \"float64\")\n",
    "\n",
    "            self.matrix[line] = vector/vector.sum()\n",
    "            \n",
    "            counts.clear()\n",
    "            del vector\n",
    "            del seq\n",
    "            del counts\n",
    "    \n",
    "    \n",
    "    def get_keys(self):\n",
    "        \n",
    "        return(self.keys)\n",
    "    \n",
    "    def get_matrix(self):\n",
    "        \n",
    "        return(self.matrix)\n",
    "\n",
    "\n",
    "# In[3]:\n",
    "\n",
    "\n",
    "def subsplit(subtype):\n",
    "    if re.match('^[H][0-9]+N[0-9]+$', subtype): \n",
    "        H = re.search('[H][0-9]+', subtype).group(0)\n",
    "        N = re.search('[N][0-9]+', subtype).group(0)\n",
    "    else:\n",
    "        H = ''#np.nan\n",
    "        N = ''#np.nan\n",
    "    return(H, N)\n",
    "\n",
    "\n",
    "# In[4]:\n",
    "\n",
    "\n",
    "def convert_fasta(infile, outfile):\n",
    "\n",
    "    #with some help from https://stackoverflow.com/questions/50856538/how-to-convert-multiline-fasta-files-to-singleline-fasta-files-without-biopython\n",
    "    \n",
    "    with open(infile) as f_input, open(outfile, 'w') as f_output:\n",
    "\n",
    "        block = []\n",
    "\n",
    "        for line in f_input:\n",
    "    \n",
    "            if line.startswith('>'):\n",
    "\n",
    "                if block:\n",
    "                    f_output.write(new_line + ',' + \"\".join(block) + '\\n')\n",
    "                    block = []\n",
    "                \n",
    "                new_line = line.strip().replace('|', ',')\n",
    "            else:\n",
    "                block.append(line.strip())\n",
    "    \n",
    "        if block:\n",
    "            f_output.write(new_line + ',' + \"\".join(block) + '\\n')\n",
    "\n",
    "\n",
    "# In[5]:\n",
    "\n",
    "\n",
    "def get_vectors(sequence, accession, metric, neigh, comp, pca):\n",
    "\n",
    "    freq = frequency(k = 7, convert = 0)\n",
    "    freq.adjust_to_data(sequence)\n",
    "    freq.calculate_frequence(sequence)\n",
    "\n",
    "    #Kevins approach\n",
    "    #freq = frequency(k = 7)\n",
    "    #freq.calculate_frequence(sequence)\n",
    "    \n",
    "    if pca:\n",
    "        matrix = freq.get_matrix()\n",
    "        pca_init = PCA(n_components=pca)\n",
    "        pca_matr = pca_init.fit_transform(matrix)\n",
    "        explained_var = pca_init.explained_variance_ratio_.sum()\n",
    "        #pca_matr = freq.get_matrix()\n",
    "\n",
    "        dataframe_pca = pd.concat([accession, pd.DataFrame(pca_matr)], axis=1, copy = False, ignore_index = False).set_index('accession')\n",
    "    else:\n",
    "        pca_matr = freq.get_matrix()\n",
    "        dataframe_pca = pd.DataFrame()\n",
    "        explained_var = -1\n",
    "    \n",
    "    \n",
    "    if (pca and not comp) or (pca <= comp):\n",
    "        reduced = pca_matr\n",
    "\n",
    "    elif neigh and comp:\n",
    "        reduced = umap.UMAP(\n",
    "            n_neighbors = neigh,\n",
    "            min_dist = 0.0,\n",
    "            n_components = comp,\n",
    "            random_state = 42,\n",
    "            #metric = 'euclidean',\n",
    "        ).fit_transform(pca_matr)\n",
    "        \n",
    "    else:\n",
    "        reduced = pca_matr\n",
    "    \n",
    "    #approximation of cosine by l2 normalization (https://github.com/scikit-learn-contrib/hdbscan/issues/69)\n",
    "    if metric == 'cosine':\n",
    "        cos_red = normalize(reduced, norm='l2')\n",
    "        dataframe = pd.concat([accession, pd.DataFrame(cos_red)], axis=1, copy = False, ignore_index = False).set_index('accession')\n",
    "    else:\n",
    "        dataframe = pd.concat([accession, pd.DataFrame(reduced)], axis=1, copy = False, ignore_index = False).set_index('accession')\n",
    "    \n",
    "    return(dataframe, dataframe_pca, explained_var)\n",
    "\n",
    "\n",
    "# In[6]:\n",
    "\n",
    "\n",
    "def get_elbow(dataframe, extra, accession, metric, min_clust, sample, kneedle):\n",
    "    \n",
    "    #with some help from https://joernhees.de/blog/2015/08/26/scipy-hierarchical-clustering-and-dendrogram-tutorial/\n",
    "    \n",
    "    if metric == 'cosine':\n",
    "        metric = 'euclidean'\n",
    "    \n",
    "    clusterer_best = hdbscan.HDBSCAN(\n",
    "        min_samples = sample,\n",
    "        min_cluster_size = min_clust,\n",
    "        gen_min_span_tree = True,\n",
    "        metric = metric,\n",
    "    ).fit(dataframe)\n",
    "    \n",
    "    label = clusterer_best.labels_\n",
    "    label_list = label.tolist()\n",
    "    \n",
    "    if -1 in label_list:\n",
    "        label_list.remove(-1)\n",
    "    \n",
    "    n_cluster = len(set(label_list))\n",
    "    \n",
    "    linkage = clusterer_best.single_linkage_tree_.to_pandas()\n",
    "    numpy_linkage = linkage.drop(columns=['parent']).to_numpy()\n",
    "    \n",
    "    dist = numpy_linkage[:, 2]\n",
    "    dist_rev = dist[::-1]\n",
    "    \n",
    "    if kneedle == -1:\n",
    "        dist_area = dist_rev[0:n_cluster]\n",
    "    else:\n",
    "        dist_area = dist_rev[0:kneedle]\n",
    "    \n",
    "    idxs = np.arange(1, len(dist_area) + 1)\n",
    "\n",
    "    kn = KneeLocator(idxs, dist_area,\n",
    "        curve='convex',\n",
    "        direction='decreasing',\n",
    "        interp_method='polynomial',\n",
    "        online=True,\n",
    "        S = 1.0,\n",
    "    )\n",
    "    \n",
    "    n_cluster_raw = kn.knee\n",
    "    n_cluster_norm = kn.norm_knee\n",
    "    \n",
    "    #linkage.set_index('parent', inplace = True)\n",
    "    elbow = pd.DataFrame({'n_cluster': idxs, \n",
    "                          'distance': dist_area, \n",
    "                          'x_normalized':kn.x_normalized, \n",
    "                          'y_normalized':kn.y_normalized, \n",
    "                          'x_difference':kn.x_difference, \n",
    "                          'y_difference':kn.y_difference}).set_index('n_cluster')\n",
    "    epsilon_best = elbow.loc[n_cluster_raw][['distance']].item()\n",
    "    linkage.set_index('parent', inplace = True)\n",
    "    \n",
    "    return(linkage, elbow, epsilon_best, n_cluster_raw, n_cluster_norm)\n",
    "\n",
    "\n",
    "# In[7]:\n",
    "\n",
    "\n",
    "def get_cluster(epsilon_best, dataframe, extra, accession, metric, min_clust, sample):\n",
    "    \n",
    "    #approximation see above\n",
    "    if metric == 'cosine':\n",
    "        metric = 'euclidean'\n",
    "    \n",
    "    clusterer_best = hdbscan.HDBSCAN(\n",
    "        min_samples = sample,\n",
    "        min_cluster_size = min_clust,\n",
    "        cluster_selection_epsilon = epsilon_best,\n",
    "        gen_min_span_tree=True,\n",
    "        metric = metric,\n",
    "    ).fit(dataframe)\n",
    "\n",
    "    label = clusterer_best.labels_\n",
    "    label_list = label.tolist()\n",
    "    \n",
    "    unclustered = label_list.count(-1)\n",
    "    if -1 in label_list:\n",
    "        label_list.remove(-1)\n",
    "    \n",
    "    n_cluster = len(set(label_list))\n",
    "    \n",
    "    cluster = pd.concat([pd.DataFrame(label, columns = ['cluster']), extra, accession], axis=1, copy = False).set_index('accession')\n",
    "    #linkage = clusterer_best.single_linkage_tree_.to_pandas()\n",
    "    #linkage.set_index('parent', inplace = True)\n",
    "    \n",
    "    N_unmatch = cluster.query('cluster != -1').groupby('cluster').filter(lambda x: x['N'].replace('', np.nan).nunique() > 1).cluster.nunique()\n",
    "    H_unmatch = cluster.query('cluster != -1').groupby('cluster').filter(lambda x: x['H'].replace('', np.nan).nunique() > 1).cluster.nunique() \n",
    "    \n",
    "    cluster.insert(2, 'centroid', False)\n",
    "\n",
    "    num = cluster['cluster'].max()+1\n",
    "    values = [True]*num\n",
    "    accessions = []\n",
    "\n",
    "    for i in range(num):\n",
    "\n",
    "        query = cluster.query('cluster == @i')\n",
    "        match = query.index.values.tolist()\n",
    "        sub = dataframe.loc[match]\n",
    "\n",
    "        if not sub.empty:\n",
    "\n",
    "            tuple_min = (0, '')\n",
    "            for name, row in sub.iterrows():\n",
    "\n",
    "                if not tuple_min[1]:\n",
    "                    dist_min = ssd.cdist([row], sub, metric).mean()\n",
    "                    tuple_min = (dist_min, name)\n",
    "                else:\n",
    "                    dist_min = ssd.cdist([row], sub, metric).mean()\n",
    "                    if dist_min < tuple_min[0]:\n",
    "                        tuple_min = (dist_min, name)\n",
    "\n",
    "            accessions.append(tuple_min[1])\n",
    "\n",
    "    cluster.update(pd.DataFrame(values, columns=['centroid'], index = accessions))\n",
    "    \n",
    "    return(cluster, n_cluster, unclustered, H_unmatch, N_unmatch)\n",
    "\n",
    "\n",
    "# In[8]:\n",
    "\n",
    "\n",
    "def plot_density(density, segments, outpath, render, accuracy, suffix = ''):\n",
    "\n",
    "    for seg in segments:\n",
    "             \n",
    "        with sns.axes_style(\"darkgrid\"):\n",
    "            fig, ax = plt.subplots(figsize=(4,4))\n",
    "            sns.histplot(data = density.query('segment == 4'), x = \"size\", kde=False, log_scale=False, element=\"step\", fill = True, bins = round(density.query('segment == 4')['size'].max()*accuracy), ax = ax)\n",
    "            plt.xlabel(\"log(Count)\")\n",
    "            plt.ylabel(\"#Cluster\")\n",
    "            plt.tight_layout()\n",
    "            plt.close(fig)\n",
    "            fig.savefig(outpath + 'Cluster_Distribution_Segment_' + str(seg) + str(suffix) + '.' + render)\n",
    "    \n",
    "        with sns.axes_style(\"darkgrid\"):\n",
    "            fig, ax = plt.subplots(figsize=(4,4))\n",
    "            sns.histplot(data = density.query('segment == 4'), x = \"size\", kde=False, log_scale=True, element=\"step\", fill = True, bins = round(density.query('segment == 4')['size'].max()*accuracy), ax = ax)\n",
    "            plt.xlabel(\"log(Count)\")\n",
    "            plt.ylabel(\"#Cluster\")\n",
    "            plt.tight_layout()\n",
    "            plt.close(fig)\n",
    "            fig.savefig(outpath + 'Cluster_Distribution_Log_Segment_' + str(seg) + str(suffix) + '.' + render)\n",
    "\n",
    "# In[9]:\n",
    "\n",
    "\n",
    "def plot_elbow(elbow, information, segments, outfolder, render, suffix = ''):\n",
    "\n",
    "        \n",
    "    for seg in segments:\n",
    "    \n",
    "        with sns.axes_style(\"darkgrid\"):\n",
    "            fig, ax = plt.subplots(figsize=(4,4))\n",
    "        \n",
    "            sns.lineplot(data = elbow.query('segment == @seg'), x = \"n_cluster\", y = \"distance\", ax = ax)\n",
    "\n",
    "            y_max = elbow.query('segment == @seg')['distance'].max().item()\n",
    "            x_max = elbow.query('segment == @seg')['n_cluster'].max().item()\n",
    "\n",
    "            y_note = information.query('segment == @seg')['epsilon'].item()\n",
    "            x_note = information.query('segment == @seg')['n_cluster_raw'].item()\n",
    "            z_note = information.query('segment == @seg')['n_cluster'].item()\n",
    "\n",
    "            ax.axhline(y = y_note, alpha=1, color=\"y\", label = f\"Distance = {y_note:0.3f}\")\n",
    "            ax.axvline(x = x_note, alpha=1, color=\"r\", label = f\"#Cluster = {x_note:0.0f}\")\n",
    "            ax.axvline(x = z_note, alpha=1, color=\"g\", label = f\"#Cluster = {z_note:0.0f}\")\n",
    "\n",
    "            #ax.text(x_max, y_note, f\"{y_note:0.3f}\", va='bottom', ha='right', color = \"r\", alpha=1, size=10)\n",
    "            #ax.text(x_note, y_max, f\"{x_note:0.0f}\", rotation='vertical', va='top', ha='left', color = \"r\", alpha=1, size=10)\n",
    "            #ax.text(z_note, y_max, f\"{z_note:0.0f}\", rotation='vertical', va='top', ha='right', color = \"g\", alpha=1, size=10)\n",
    "\n",
    "            plt.xlabel(\"#Cluster\")\n",
    "            plt.ylabel(\"Distance\")\n",
    "        \n",
    "            plt.legend()\n",
    "            plt.tight_layout()\n",
    "            plt.close(fig)\n",
    "            fig.savefig(outfolder + 'Cluster_Elbow_Knee_Segment_' + str(seg) + str(suffix) + '.' + render)\n",
    "\n",
    "        \n",
    "        with sns.axes_style(\"darkgrid\"):\n",
    "            fig, ax = plt.subplots(figsize=(4,4))\n",
    "        \n",
    "            sns.lineplot(data = elbow.query('segment == @seg'), x = \"x_normalized\", y = \"y_normalized\", ax = ax)\n",
    "            sns.lineplot(data = elbow.query('segment == @seg'), x = \"x_difference\", y = \"y_difference\", color=\"g\", ax = ax)\n",
    "        \n",
    "            z_note = information.query('segment == @seg')['n_cluster_norm'].item()\n",
    "            ax.axvline(x = z_note, alpha=1, color=\"r\", label = f\"norm(#Cluster) = {z_note:0.3f}\")\n",
    "            #ax.text(z_note, y_max*0.0, f\"{z_note:0.3f}\", rotation='vertical', va='bottom', ha='left', color = \"r\", alpha=1, size=10)\n",
    "\n",
    "            plt.xlabel(\"norm(#Cluster)\")\n",
    "            plt.ylabel(\"Difference\")\n",
    "        \n",
    "            plt.legend()\n",
    "            plt.tight_layout()\n",
    "            plt.close(fig)\n",
    "            fig.savefig(outfolder + 'Cluster_Knee_Segment_' + str(seg) + str(suffix) + '.' + render)\n",
    "\n",
    "\n",
    "# In[10]:\n",
    "\n",
    "\n",
    "#Modified Version of the code presented on https://github.com/scipy/scipy/issues/8274\n",
    "def getNewick(node, newick, parentdist, leaf_names):\n",
    "    \n",
    "    if node.is_leaf():\n",
    "        return(leaf_names[node.id] + \":\" + np.format_float_positional(parentdist - node.dist) + newick)\n",
    "        #return \"%s:%.5f%s\" % (leaf_names[node.id], parentdist - node.dist, newick)\n",
    "    \n",
    "    else:\n",
    "        if len(newick) > 0:\n",
    "            newick = \"):\" + np.format_float_positional(parentdist - node.dist) + newick\n",
    "            #newick = \"):%.5f%s\" % (parentdist - node.dist, newick)\n",
    "        else:\n",
    "            newick = \");\"\n",
    "        newick = getNewick(node.get_left(), newick, node.dist, leaf_names)\n",
    "        newick = getNewick(node.get_right(), \",\" + newick, node.dist, leaf_names)\n",
    "        #newick = getNewick(node.get_right(), \",%s\" % (newick), node.dist, leaf_names)\n",
    "        \n",
    "        newick = \"(\" + newick\n",
    "        #newick = \"(%s\" % (newick)\n",
    "        return(newick)\n",
    "\n",
    "\n",
    "# In[11]:\n",
    "\n",
    "\n",
    "def alignment(fasta, threads, outfasta, outmsa, treeout):    \n",
    "\n",
    "    fasta.to_csv(outfasta, header=None, index=True, sep='\\n', mode='w')\n",
    "\n",
    "    mafft_cline = MafftCommandline(input=outfasta, thread=threads,  treeout = treeout)\n",
    "    #mafft_cline = MafftCommandline(input=outfasta, thread=threads,  treeout = False)\n",
    "    stdout, stderr = mafft_cline()\n",
    "\n",
    "    #os.remove('tmp.fasta') \n",
    "    with open(outmsa, \"w\") as handle:\n",
    "        handle.write(stdout)\n",
    "    \n",
    "    return(stdout)\n",
    "\n",
    "\n",
    "# In[12]:\n",
    "\n",
    "\n",
    "def fill_gaps(group, protein):\n",
    "    \n",
    "    set_protein = set(group[protein].tolist())\n",
    "    if '' in set_protein:\n",
    "        set_protein.discard('')\n",
    "    if len(set_protein) == 1:\n",
    "        group['protein'] = list(set_protein)[0]\n",
    "    else:\n",
    "        group['protein'] = group[protein]\n",
    "        group['protein'].replace('', 'mixed', inplace = True)\n",
    "        group['protein'].fillna('mixed', inplace = True)\n",
    "        \n",
    "    return(group)\n",
    "\n",
    "\n",
    "# In[13]:\n",
    "\n",
    "\n",
    "def get_acc(dictionary_tree, list_value):\n",
    "    list_accessions = []\n",
    "    for key, value in dictionary_tree.items():\n",
    "         if value in list_value:\n",
    "            list_accessions.append(key)\n",
    "    return(list_accessions)\n",
    "\n",
    "\n",
    "# In[14]:\n",
    "\n",
    "\n",
    "def layout(node):\n",
    "    \n",
    "    if node.is_leaf():\n",
    "        N = AttrFace(\"name\", fsize=10)\n",
    "        faces.add_face_to_node(N, node, 0, position=\"aligned\")\n",
    "    \n",
    "        M = AttrFace(\"size\", fsize=10, text_prefix='|Size ', text_suffix='|')\n",
    "        faces.add_face_to_node(M, node, 1, position=\"aligned\")    \n",
    "\n",
    "\n",
    "# In[15]:\n",
    "\n",
    "\n",
    "def get_tree(cluster, upload, segment, prot, list_color_hex, list_prune = [], list_focus = [], collapse = True, path_phylo = '', msa_dict = co.defaultdict(str), area = [], linkage = pd.DataFrame()):\n",
    "\n",
    "    if prot:\n",
    "    #Dataframes Creation and Correction\n",
    "        dataframe_tree_query = cluster.query('segment == @segment').join(upload.set_index('accession')[['host', 'date']])[['cluster', prot, 'host', 'date']]\n",
    "        dataframe_tree_corr = dataframe_tree_query.groupby('cluster').apply(fill_gaps, protein = prot).drop(prot, axis = 1)\n",
    "        dictionary_tree = dataframe_tree_corr.to_dict()\n",
    "        list_tree_label = dataframe_tree_corr.index.values\n",
    "    else:\n",
    "        dataframe_tree_query = cluster.query('segment == @segment').join(upload.set_index('accession')[['host', 'date']])[['cluster', 'host', 'date']]\n",
    "        dataframe_tree_query['protein'] = 'mixed'\n",
    "        dictionary_tree = dataframe_tree_query.to_dict()\n",
    "        list_tree_label = dataframe_tree_query.index.values\n",
    "        \n",
    "    #Tree Backbone Creation:\n",
    "    tree_style = TreeStyle()\n",
    "    \n",
    "    if not path_phylo and not linkage.empty:\n",
    "        array_tree_link = linkage.query('segment == @segment').drop(columns=['segment', 'parent']).to_numpy()\n",
    "        tree_hierarchy = hierarchy.to_tree(array_tree_link, False)\n",
    "        tree_newick = getNewick(tree_hierarchy, \"\", tree_hierarchy.dist, list_tree_label)\n",
    "        tree = Tree(tree_newick)\n",
    "    \n",
    "    elif path_phylo and linkage.empty:    \n",
    "        tree = Tree(path_phylo)\n",
    "        for leaf in tree.iter_leaves():\n",
    "            leaf_name = leaf.name.split('_')[1]\n",
    "            if not leaf_name == 'Cluster':\n",
    "                leaf.name = '>' + leaf_name\n",
    "            else:\n",
    "                leaf.name = leaf.name.split('_')[2]\n",
    "                \n",
    "    else:\n",
    "        print('Linkage Tree and Phylo Tree both or none given. Empty Tree generated.')\n",
    "        tree = Tree()\n",
    "        return(tree, tree_style)\n",
    "        \n",
    "    #General Appereance Settings:\n",
    "    for node in tree.traverse():\n",
    "        node.img_style[\"vt_line_width\"] = 1\n",
    "        node.img_style[\"hz_line_width\"] = 1\n",
    "        node.img_style[\"vt_line_type\"] = 0\n",
    "        node.img_style[\"hz_line_type\"] = 0\n",
    "        node.img_style[\"fgcolor\"] = \"#000000\"\n",
    "        node.img_style[\"shape\"] = \"circle\"\n",
    "        node.img_style[\"vt_line_color\"] = \"#000000\"\n",
    "        node.img_style[\"hz_line_color\"] = \"#000000\"\n",
    "        node.img_style['bgcolor'] = 'lightgrey'\n",
    "        node.protein = 'mixed'\n",
    "\n",
    "    #List and Dictionary Creation\n",
    "    #dict_color_protein = co.defaultdict(str)\n",
    "    list_cluster = list(set(dictionary_tree['cluster'].values()))\n",
    "    list_protein = list(set(dictionary_tree['protein'].values()))\n",
    "    list_host = list(set(dictionary_tree['host'].values()))\n",
    "    list_color = []\n",
    "    list_cluster_existing = []\n",
    "    \n",
    "    #Initial Leaf Settings:\n",
    "    for leaf in tree.iter_leaves():\n",
    "        if leaf.name.startswith('>'):\n",
    "            leaf.host = dictionary_tree['host'].get(leaf.name, \"none\")\n",
    "            leaf.date = dictionary_tree['date'].get(leaf.name, \"none\")\n",
    "            leaf.cluster = dictionary_tree['cluster'].get(leaf.name, \"none\")\n",
    "            leaf.protein = dictionary_tree['protein'].get(leaf.name, \"mixed\")\n",
    "            leaf.name = leaf.name[1:]\n",
    "            leaf.accession = True\n",
    "        else:\n",
    "            leaf_member = get_acc(dictionary_tree['cluster'], [int(leaf.name)])\n",
    "            leaf.protein = dictionary_tree['protein'].get(leaf_member[0], \"mixed\")\n",
    "            leaf.size = len(leaf_member)\n",
    "            leaf.accession = False\n",
    "    \n",
    "    #Cluster Node Creation and Settings:\n",
    "    if -1 in list_cluster:\n",
    "        list_cluster.remove(-1)\n",
    "    \n",
    "    for cluster in list_cluster:\n",
    "        try:\n",
    "            list_member = tree.search_nodes(cluster=cluster)\n",
    "            node_cluster = tree.get_common_ancestor(list_member)\n",
    "            node_cluster.name = str(cluster)\n",
    "            node_cluster.accession = False\n",
    "            node_cluster.size = len(list_member)\n",
    "            list_cluster_existing.append(cluster)\n",
    "        except:\n",
    "            pass\n",
    "            \n",
    "    #Serotype Coloring (maybe include mixed in the coloring):\n",
    "    if 'mixed' in list_protein:\n",
    "        list_protein.remove('mixed')\n",
    "\n",
    "    if prot:\n",
    "\n",
    "        for protein in list_protein:\n",
    "\n",
    "            for tree_protein in tree.get_monophyletic(values=[protein], target_attr = \"protein\"):\n",
    "\n",
    "                color = list_color_hex[int(protein[1:])-1]\n",
    "                #dict_color_protein[color] = protein\n",
    "                #tree_protein.protein = protein\n",
    "                #tree_protein.img_style['bgcolor'] = color\n",
    "\n",
    "                for node_protein in tree_protein.traverse():\n",
    "\n",
    "                    node_protein.img_style['bgcolor'] = color\n",
    "                    node_protein.protein = protein\n",
    "\n",
    "    #Host Monophyletic Lines Drawing:\n",
    "    for host in list_host:\n",
    "        for tree_host in tree.get_monophyletic(values=[host], target_attr = \"host\"):\n",
    "            \n",
    "            #tree_host.img_style[\"vt_line_type\"] = 1\n",
    "            #tree_host.img_style[\"hz_line_type\"] = 1\n",
    "            #tree_host.host = host\n",
    "            \n",
    "            for node_host in tree_host.traverse():\n",
    "                node_host.img_style[\"vt_line_type\"] = 1\n",
    "                node_host.img_style[\"hz_line_type\"] = 1\n",
    "                node_host.host = host\n",
    "\n",
    "    #Pruning according to Settings:\n",
    "    if not list_prune:\n",
    "        if not list_focus:\n",
    "            if collapse == True:\n",
    "                list_prune = list(map(str, list_cluster_existing))\n",
    "        else:\n",
    "            list_prune = [accession[1:] for accession in get_acc(dictionary_tree['cluster'], list_focus)]\n",
    "            if collapse == False:\n",
    "                list_prune.extend(list(map(str, list(set(list_cluster_existing).difference(set(list_focus))))))\n",
    "    \n",
    "    if list_prune:\n",
    "        try:\n",
    "            tree.prune(list_prune)\n",
    "        except:\n",
    "            print('No pruning performed.')\n",
    "\n",
    "    for leaf in tree:\n",
    "        if leaf.accession == True:\n",
    "            leaf.add_face(AttrFace(\"cluster\", fsize=10, text_prefix='Cluster: '), 0, position=\"aligned\")\n",
    "            leaf.add_face(AttrFace(\"name\", fsize=10, text_prefix='|', text_suffix='|'), 1, position=\"aligned\")\n",
    "            leaf.add_face(AttrFace(\"date\", fsize=10, text_prefix='Date: '), 2, position=\"aligned\")\n",
    "            if msa_dict:\n",
    "                seq = msa_dict.get(leaf.name, \"\")\n",
    "                if area:\n",
    "                    area_seq = \"...\".join([seq[start:end] for start, end in area])\n",
    "                    leaf.add_face(SequenceFace(area_seq), 3, position=\"aligned\")\n",
    "                else:\n",
    "                    leaf.add_face(SequenceFace(seq), 3, position=\"aligned\")\n",
    "        else:\n",
    "            leaf.add_face(AttrFace(\"name\", fsize=10, text_prefix='Cluster: '), 0, position=\"aligned\")\n",
    "            leaf.add_face(AttrFace(\"size\", fsize=10, text_prefix='|Size: ', text_suffix='|'), 1, position=\"aligned\")\n",
    "            if msa_dict:\n",
    "                seq = msa_dict.get(leaf.name, \"\")\n",
    "                if area:\n",
    "                    area_seq = \"...\".join([seq[start:end] for start, end in area])\n",
    "                    leaf.add_face(SequenceFace(area_seq), 2, position=\"aligned\")\n",
    "                else:\n",
    "                    leaf.add_face(SequenceFace(seq), 2, position=\"aligned\")\n",
    "        \n",
    "        list_color.append(leaf.img_style['bgcolor'])\n",
    "\n",
    "    if prot:\n",
    "        set_color = set(list_color)\n",
    "        column = 0\n",
    "        for color_hex in list_color_hex:\n",
    "            if color_hex in set_color:\n",
    "                face_protein = TextFace(prot + str(list_color_hex.index(color_hex)+1))\n",
    "                face_protein.margin_top = 10\n",
    "                face_protein.margin_right = 10\n",
    "                face_protein.margin_left = 10\n",
    "                face_protein.margin_bottom = 10\n",
    "                face_protein.border.width = 1\n",
    "                face_protein.background.color = color_hex\n",
    "                tree_style.legend.add_face(face_protein, column=column)\n",
    "                column += 1\n",
    "        tree_style.legend_position = 1\n",
    "            \n",
    "    tree_style.show_leaf_name = False\n",
    "    tree_style.draw_guiding_lines = True\n",
    "    tree_style.guiding_lines_color = 'black'\n",
    "    \n",
    "    if not msa_dict:\n",
    "        tree_style.mode = \"c\"\n",
    "\n",
    "    return(tree, tree_style, tree_newick)\n",
    "\n",
    "\n",
    "def pairmsa(nuc1, nuc2):\n",
    "        \n",
    "    seq1 = Seq(nuc1.item())\n",
    "    seq2 = Seq(nuc2.item()) \n",
    "\n",
    "    aligner = Align.PairwiseAligner()\n",
    "    alignments = aligner.align(seq1, seq2) \n",
    "    \n",
    "    #alignments = pairwise2.align.globalmx(seq1, seq2, 1, -1) \n",
    "    \n",
    "    return(alignments[0].score)\n",
    "\n",
    "def sample_difference(cluster, genome, segment, tree, threshold = -1, n = -1, index = []):\n",
    "    \n",
    "    if not index:\n",
    "        num = cluster['cluster'].max()+1\n",
    "        index = [i for i in range(num)]\n",
    "    else:\n",
    "        num = len(index)\n",
    "    \n",
    "    pair = np.empty([num, num])\n",
    "    \n",
    "    for x, j in enumerate(index):\n",
    "        for y, k in enumerate(index):\n",
    "            \n",
    "            if j != k:\n",
    "                dist = tree.get_distance(str(j), str(k))\n",
    "            else:\n",
    "                dist = 0.0\n",
    "            \n",
    "            if threshold == -1 or dist <= threshold:\n",
    "            \n",
    "                if n == -1:\n",
    "                    sample_x = cluster.query('segment == @segment & cluster == @j & centroid == True').join(genome)[['genome']]\n",
    "                    sample_y = cluster.query('segment == @segment & cluster == @k & centroid == True').join(genome)[['genome']]\n",
    "\n",
    "                else:\n",
    "                    query_x = cluster.query('segment == @segment & cluster == @j').join(genome)[['genome']]\n",
    "                    if len(query_x) >= n:\n",
    "                        sample_x = query_x.sample(n=n, random_state=42)\n",
    "                    else:\n",
    "                        sample_x = query_x\n",
    "\n",
    "                    query_y = cluster.query('segment == @segment & cluster == @k').join(genome)[['genome']]\n",
    "                    if len(query_y) >= n:\n",
    "                        sample_y = query_y.sample(n=n, random_state=42)\n",
    "                    else:\n",
    "                        sample_y = query_y\n",
    "\n",
    "                dist = ssd.cdist(sample_x, sample_y, metric = pairmsa)\n",
    "\n",
    "                dist_mean = pd.DataFrame(dist, dtype = 'float64').mean().mean()\n",
    "                len_mean = (sample_x['genome'].str.len().mean() + sample_y['genome'].str.len().mean())/2\n",
    "                \n",
    "                pair[x,y] = dist_mean/len_mean\n",
    "                \n",
    "            else:\n",
    "                pair[x,y] = 0.0\n",
    "            \n",
    "    dataframe_pair = pd.DataFrame(pair, index = index, columns = index)\n",
    "\n",
    "    return(dataframe_pair)\n",
    "\n",
    "\n",
    "def plot_difference(proof, segments, outpath, render):\n",
    "\n",
    "    with sns.axes_style(\"darkgrid\"):\n",
    "        fig, ax = plt.subplots(figsize=(8,6))\n",
    "        ax = sns.heatmap(proof, linewidths=.5)\n",
    "        plt.xlabel(\"Cluster\")\n",
    "        plt.ylabel(\"Cluster\")\n",
    "        plt.tight_layout()\n",
    "        plt.close(fig)\n",
    "        fig.savefig(outpath + 'Cluster_Difference_Segment_' + str(segments) + '.' + render)\n",
    "\n",
    "# In[23]:\n",
    "\n",
    "\n",
    "def main(infile, outfolder, segments, custom, metric, min_clust, sample, umap_neigh, umap_comp, pca_comp, kneedle, render, epsilon):#, recreate):\n",
    "\n",
    "    Hs = ['H1', 'H2', 'H3', 'H4', 'H5', 'H6', 'H7', 'H8', 'H9', 'H10', 'H11', 'H12', 'H13', 'H14', 'H15', 'H16', 'H17', 'H18']\n",
    "    Ns = ['N1', 'N2', 'N3', 'N4', 'N5', 'N6', 'N7', 'N8', 'N9', 'N10', 'N11']\n",
    "    tree_hex = ['#FF0029','#377EB8','#66A61E','#984EA3','#00D2D5','#FF7F00','#AF8D00','#7F80CD','#B3E900','#C42E60','#A65628','#F781BF','#8DD3C7','#BEBADA','#FB8072','#80B1D3','#FDB462','#FCCDE5']\n",
    "\n",
    "    try:\n",
    "        if not outfolder.endswith(\"/\"):\n",
    "            outfolder += \"/\"\n",
    "\n",
    "        if not os.path.isdir(outfolder):\n",
    "            os.mkdir(outfolder)\n",
    "\n",
    "        #if recreate == True:\n",
    "        for f in os.listdir(outfolder):\n",
    "            os.remove(os.path.join(outfolder, f))\n",
    "    except:\n",
    "        sys.exit('Outfolder not reachable. Exit.')\n",
    "\n",
    "    #Dict Creation\n",
    "    dataframe_pca_dict = co.defaultdict(list)\n",
    "    dataframe_dict = co.defaultdict(list)\n",
    "    cluster_dict = co.defaultdict(list)\n",
    "    information_dict = co.defaultdict(list)\n",
    "    linkage_dict = co.defaultdict(list)\n",
    "    elbow_dict = co.defaultdict(list)\n",
    "\n",
    "    #Upload Fasta\n",
    "    try:\n",
    "        upload = pd.read_csv('fasta.csv', sep = ',', na_filter = False, header = None)\n",
    "\n",
    "    except:\n",
    "        try:\n",
    "            #Convert the Fasta to Csv\n",
    "            convert_fasta(infile, 'fasta.csv')\n",
    "            upload = pd.read_csv('fasta.csv', sep = ',', na_filter = False, header = None)\n",
    "        except:\n",
    "            sys.exit('Infile not reachable or wrong format. Exit.')\n",
    "\n",
    "    upload.columns = custom\n",
    "    upload.query('curation == \"Pass\"', inplace = True)\n",
    "    upload[['H', 'N']] = upload.subtype.apply(lambda x: subsplit(x)).tolist()\n",
    "    #upload.to_csv(outfolder + 'upload.csv', index=False, header=True, sep=',', mode='w')\n",
    "\n",
    "    genome = upload[['accession', 'genome']].copy()\n",
    "    genome.set_index('accession', inplace = True)\n",
    "\n",
    "    try:\n",
    "\n",
    "        #dataframe_pca = pd.read_csv(outfolder + 'dataframe_pca.csv', sep = ',', na_filter = False, header = 0, index_col = 0)\n",
    "        #dataframe = pd.read_csv(outfolder + 'dataframe.csv', sep = ',', na_filter = False, header = 0, index_col = 0)\n",
    "        cluster = pd.read_csv(outfolder + 'cluster.csv', sep = ',', na_filter = False, header = 0, index_col = 0)\n",
    "        linkage = pd.read_csv(outfolder + 'linkage.csv', sep = ',', na_filter = False, header = 0)\n",
    "        information = pd.read_csv(outfolder + 'information.csv', sep = ',', na_filter = False, header = 0, index_col = 0)\n",
    "        density = pd.read_csv(outfolder + 'density.csv', sep = ',', na_filter = False, header = 0)\n",
    "        elbow = pd.read_csv(outfolder + 'elbow.csv', sep = ',', na_filter = False, header = 0)\n",
    "\n",
    "    except:\n",
    "\n",
    "        #Progress Variables\n",
    "        widgets = [' [', progressbar.Timer(format = 'elapsed time: %(elapsed)s'), '] ', progressbar.Bar('#'),' (', progressbar.ETA(), ') ', ] \n",
    "        bar = progressbar.ProgressBar(max_value = len(segments), widgets = widgets).start() \n",
    "\n",
    "        #Pipeline\n",
    "        for progress, segment in enumerate(segments, 1):\n",
    "\n",
    "            subset = upload.query('segment == @segment').reset_index()\n",
    "            sequence = subset[['genome']].copy()\n",
    "            accession = subset[['accession']].copy()\n",
    "            extra = subset[['H', 'N']].copy()\n",
    "\n",
    "            dataframe_dict[segment], dataframe_pca_dict[segment], explained_var = get_vectors(sequence, accession, metric, umap_neigh, umap_comp, pca_comp)\n",
    "            linkage_dict[segment], elbow_dict[segment], epsilon_best, n_cluster_raw, n_cluster_norm = get_elbow(dataframe_dict[segment], extra, accession, metric, min_clust, sample, kneedle)\n",
    "\n",
    "            cluster_dict[segment], n_cluster, unclustered, H_unmatch, N_unmatch = get_cluster(epsilon_best, dataframe_dict[segment], extra, accession, metric, min_clust, sample)\n",
    "            information_dict[segment] = [n_cluster, n_cluster_raw, n_cluster_norm, unclustered, H_unmatch, N_unmatch, epsilon_best, explained_var]\n",
    "\n",
    "            bar.update(progress)\n",
    "\n",
    "        bar.finish()    \n",
    "\n",
    "        ########################################################################################################\n",
    "\n",
    "        #dataframe_pca = pd.concat(dataframe_pca_dict)\n",
    "        #dataframe_pca.index.set_names([\"segment\", \"accession\"], inplace=True)\n",
    "        #dataframe_pca.reset_index(level = \"segment\", inplace=True, drop = True)\n",
    "\n",
    "        #dataframe = pd.concat(dataframe_dict)\n",
    "        #dataframe.index.set_names([\"segment\", \"accession\"], inplace=True)\n",
    "        #dataframe.reset_index(level = \"segment\", inplace=True, drop = True)\n",
    "\n",
    "        cluster = pd.concat(cluster_dict)\n",
    "        cluster.index.set_names([\"segment\", \"accession\"], inplace=True)\n",
    "        cluster.reset_index(level = \"segment\", inplace=True)\n",
    "\n",
    "        linkage = pd.concat(linkage_dict)\n",
    "        linkage.index.set_names([\"segment\", 'parent'], inplace=True)\n",
    "        linkage.reset_index(level = [\"segment\", 'parent'], inplace=True)\n",
    "\n",
    "        information = pd.DataFrame.from_dict(information_dict, orient='index', columns=['n_cluster', 'n_cluster_raw', 'n_cluster_norm', 'unclustered', 'H_unmatch', 'N_unmatch', 'epsilon', 'explained_var'])\n",
    "        information.index.set_names([\"segment\"], inplace=True)\n",
    "\n",
    "        elbow = pd.concat(elbow_dict)\n",
    "        elbow.index.set_names(['segment', 'n_cluster'], inplace=True)\n",
    "        elbow.reset_index(level = [\"segment\", \"n_cluster\"], inplace=True)\n",
    "\n",
    "        density = pd.DataFrame(cluster.value_counts(subset=['segment', 'cluster']), columns = ['size'])\n",
    "        density.reset_index(level = [\"segment\", \"cluster\"], inplace=True)\n",
    "\n",
    "        ########################################################################################################\n",
    "\n",
    "        #correcting the KNOWN errors in the database\n",
    "        if '>MK237334' in cluster.index:\n",
    "            cluster.loc['>MK237334','H'] = \"H1\"\n",
    "\n",
    "        if '>MH346657' in cluster.index:\n",
    "            cluster.loc['>MH346657','N'] = \"N1\"\n",
    "\n",
    "        if '>MK237837' in cluster.index:\n",
    "            cluster.loc['>MK237837','N'] = \"N1\"\n",
    "\n",
    "        if '>MG925516' in cluster.index:\n",
    "            cluster.loc['>MG925516','N'] = \"N2\"\n",
    "\n",
    "        ########################################################################################################\n",
    "\n",
    "        #dataframe_pca.to_csv(outfolder + 'dataframe_pca.csv', index=True, header=True, sep=',', mode='w')\n",
    "        #dataframe.to_csv(outfolder + 'dataframe.csv', index=True, header=True, sep=',', mode='w')\n",
    "\n",
    "        cluster.to_csv(outfolder + 'cluster.csv', index=True, header=True, sep=',', mode='w')\n",
    "        #genome.to_csv(outfolder + 'genome.csv', index=True, header=True, sep=',', mode='w')\n",
    "        information.to_csv(outfolder + 'information.csv', index=True, header=True, sep=',', mode='w')\n",
    "\n",
    "        elbow.to_csv(outfolder + 'elbow.csv', index=False, header=True, sep=',', mode='w')\n",
    "        density.to_csv(outfolder + 'density.csv', index=False, header=True, sep=',', mode='w')\n",
    "        linkage.to_csv(outfolder + 'linkage.csv', index=False, header=True, sep=',', mode='w')\n",
    "\n",
    "        ########################################################################################################    \n",
    "\n",
    "    plot_elbow(elbow, information, segments, outfolder, render)\n",
    "    plot_density(density, segments, outfolder, render, 1)\n",
    "\n",
    "    anti_dict = {1:None, 2:None, 3:None, 4:'H', 5:None, 6:'N', 7:None, 8:None}\n",
    "\n",
    "    for seg in segments:\n",
    "\n",
    "        tree, ts, newick = get_tree(cluster, upload, seg, anti_dict[seg], tree_hex, linkage = linkage)\n",
    "        tree.render(outfolder + 'Clustertree_Segment_' + str(seg) + '.' + render, tree_style = ts)\n",
    "        with open(outfolder + 'Newick_Segment_' + str(seg) + '.txt', 'w') as f:\n",
    "            f.write(newick)\n",
    "        proof = sample_difference(cluster, genome, seg, tree)\n",
    "        proof.to_csv(outfolder + 'Pairwise_Segment_' + str(seg) + '.csv', index=True, header=True, sep=',', mode='w')\n",
    "        plot_difference(proof, 4, outfolder, render)\n",
    "\n",
    "    os.remove(\"fasta.csv\") \n",
    "\t\n",
    "# In[29]:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "smart-tutorial",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_difference(proof, segments, outpath, render):\n",
    "\n",
    "    with sns.axes_style(\"darkgrid\"):\n",
    "        fig, ax = plt.subplots(figsize=(8,6))\n",
    "        ax = sns.heatmap(proof, linewidths=.5)\n",
    "        plt.xlabel(\"Cluster\")\n",
    "        plt.ylabel(\"Cluster\")\n",
    "        plt.tight_layout()\n",
    "        plt.close(fig)\n",
    "        fig.savefig(outpath + 'Cluster_Difference_Segment_' + str(segments) + '.' + render)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "psychological-terminology",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pairmsa(nuc1, nuc2):\n",
    "        \n",
    "    seq1 = Seq(nuc1.item())\n",
    "    seq2 = Seq(nuc2.item()) \n",
    "\n",
    "    aligner = Align.PairwiseAligner()\n",
    "    alignments = aligner.align(seq1, seq2) \n",
    "    \n",
    "    #alignments = pairwise2.align.globalmx(seq1, seq2, 1, -1) \n",
    "    \n",
    "    return(alignments[0].score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "announced-design",
   "metadata": {},
   "outputs": [],
   "source": [
    "outfolder = 'Results/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "stone-portable",
   "metadata": {},
   "outputs": [],
   "source": [
    "convert_fasta('../../Desktop/A.fasta', '../../Desktop/fasta.csv')\n",
    "upload = pd.read_csv('../../Desktop/fasta.csv', sep = ',', na_filter = False, header = None)\n",
    "\n",
    "upload.columns = ['accession', 'strain', 'segment', 'protein', 'genus', 'subtype', 'date', 'host', 'curation', 'genome']\n",
    "upload.query('curation == \"Pass\"', inplace = True)\n",
    "upload[['H', 'N']] = upload.subtype.apply(lambda x: subsplit(x)).tolist()\n",
    "\n",
    "genome = upload[['accession', 'genome']].copy()\n",
    "genome.set_index('accession', inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "specific-speaker",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster = pd.read_csv(outfolder + 'cluster.csv', sep = ',', na_filter = False, header = 0, index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "id": "cognitive-sequence",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import pairwise2\n",
    "from multiprocessing import Pool\n",
    "import functools\n",
    "\n",
    "def pairmsa2(nuc1, nuc2):\n",
    "        \n",
    "    seq1 = Seq(nuc1.item())\n",
    "    seq2 = Seq(nuc2.item()) \n",
    "\n",
    "    alignments = pairwise2.align.globalxx(seq1, seq2, score_only = True) \n",
    "    \n",
    "    return(alignments)\n",
    "\n",
    "def worker(x, y, j, k, n, cl):\n",
    "           \n",
    "    if n == -1:\n",
    "        sample_x = cl.query('cluster == @j & centroid == True').join(genome)[['genome']]\n",
    "        sample_y = cl.query('cluster == @k & centroid == True').join(genome)[['genome']]\n",
    "\n",
    "    else:\n",
    "        query_x = cl.query('cluster == @j').join(genome)[['genome']]\n",
    "        if len(query_x) >= n:\n",
    "            sample_x = query_x.sample(n=n, random_state=42)\n",
    "        else:\n",
    "            sample_x = query_x\n",
    "\n",
    "        query_y = cl.query('cluster == @k').join(genome)[['genome']]\n",
    "        if len(query_y) >= n:\n",
    "            sample_y = query_y.sample(n=n, random_state=42)\n",
    "        else:\n",
    "            sample_y = query_y\n",
    "\n",
    "    dist_mean = ssd.cdist(sample_x, sample_y, metric = pairmsa2).mean()\n",
    "    len_mean = (sample_x['genome'].str.len().mean() + sample_y['genome'].str.len().mean())/2\n",
    "\n",
    "    return((x,y,dist_mean/len_mean))\n",
    "\n",
    "def sample_difference(cluster, genome, segment, proc = 8, n = -1):\n",
    "\n",
    "    cl = cluster.query('segment == @segment') \n",
    "    names = list(set(cl['cluster']).difference({-1}))\n",
    "    num = len(names)\n",
    "        \n",
    "    params = []\n",
    "    dist_matrix = np.zeros([num, num])\n",
    "\n",
    "    for x, j in enumerate(names):\n",
    "        for y, k in enumerate(names[:names.index(j)+1]):\n",
    "            params.append((x, y, j, k, n, cl))\n",
    "\n",
    "    with Pool(processes=proc) as pool:\n",
    "        result = pool.starmap(worker, params)\n",
    "\n",
    "    for x, y, dist in result:\n",
    "        dist_matrix[x,y] = dist\n",
    "    \n",
    "    dist_dataframe = pd.DataFrame(dist_matrix, index = names, columns = names)\n",
    "    \n",
    "    return(dist_dataframe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "undefined-charlotte",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>46</th>\n",
       "      <th>47</th>\n",
       "      <th>48</th>\n",
       "      <th>49</th>\n",
       "      <th>50</th>\n",
       "      <th>51</th>\n",
       "      <th>52</th>\n",
       "      <th>53</th>\n",
       "      <th>54</th>\n",
       "      <th>55</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.934816</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.783949</td>\n",
       "      <td>0.977619</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.750807</td>\n",
       "      <td>0.738941</td>\n",
       "      <td>0.950710</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.748160</td>\n",
       "      <td>0.738134</td>\n",
       "      <td>0.851631</td>\n",
       "      <td>0.936520</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.854389</td>\n",
       "      <td>0.784220</td>\n",
       "      <td>0.749937</td>\n",
       "      <td>0.749702</td>\n",
       "      <td>0.949022</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.840005</td>\n",
       "      <td>0.778977</td>\n",
       "      <td>0.745428</td>\n",
       "      <td>0.743418</td>\n",
       "      <td>0.844897</td>\n",
       "      <td>0.921687</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.784556</td>\n",
       "      <td>0.799953</td>\n",
       "      <td>0.744342</td>\n",
       "      <td>0.741278</td>\n",
       "      <td>0.782100</td>\n",
       "      <td>0.780424</td>\n",
       "      <td>0.951716</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.804715</td>\n",
       "      <td>0.840063</td>\n",
       "      <td>0.755351</td>\n",
       "      <td>0.745741</td>\n",
       "      <td>0.805334</td>\n",
       "      <td>0.794352</td>\n",
       "      <td>0.842448</td>\n",
       "      <td>0.977010</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.685939</td>\n",
       "      <td>0.689590</td>\n",
       "      <td>0.691690</td>\n",
       "      <td>0.690281</td>\n",
       "      <td>0.684042</td>\n",
       "      <td>0.684396</td>\n",
       "      <td>0.685473</td>\n",
       "      <td>0.684307</td>\n",
       "      <td>0.981559</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.683152</td>\n",
       "      <td>0.694775</td>\n",
       "      <td>0.691760</td>\n",
       "      <td>0.691238</td>\n",
       "      <td>0.688755</td>\n",
       "      <td>0.688081</td>\n",
       "      <td>0.695693</td>\n",
       "      <td>0.692221</td>\n",
       "      <td>0.792899</td>\n",
       "      <td>0.968273</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.685014</td>\n",
       "      <td>0.695168</td>\n",
       "      <td>0.695434</td>\n",
       "      <td>0.694043</td>\n",
       "      <td>0.688178</td>\n",
       "      <td>0.688559</td>\n",
       "      <td>0.691181</td>\n",
       "      <td>0.690430</td>\n",
       "      <td>0.793459</td>\n",
       "      <td>0.849706</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>0.728886</td>\n",
       "      <td>0.727167</td>\n",
       "      <td>0.727936</td>\n",
       "      <td>0.741250</td>\n",
       "      <td>0.730548</td>\n",
       "      <td>0.730278</td>\n",
       "      <td>0.721784</td>\n",
       "      <td>0.720931</td>\n",
       "      <td>0.688170</td>\n",
       "      <td>0.690929</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.688921</td>\n",
       "      <td>0.696588</td>\n",
       "      <td>0.692216</td>\n",
       "      <td>0.692601</td>\n",
       "      <td>0.688108</td>\n",
       "      <td>0.684220</td>\n",
       "      <td>0.687794</td>\n",
       "      <td>0.691487</td>\n",
       "      <td>0.835340</td>\n",
       "      <td>0.811437</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.725224</td>\n",
       "      <td>0.720202</td>\n",
       "      <td>0.722864</td>\n",
       "      <td>0.730578</td>\n",
       "      <td>0.720414</td>\n",
       "      <td>0.723844</td>\n",
       "      <td>0.718734</td>\n",
       "      <td>0.724085</td>\n",
       "      <td>0.681713</td>\n",
       "      <td>0.688961</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.729473</td>\n",
       "      <td>0.726418</td>\n",
       "      <td>0.725197</td>\n",
       "      <td>0.738215</td>\n",
       "      <td>0.731152</td>\n",
       "      <td>0.727414</td>\n",
       "      <td>0.722559</td>\n",
       "      <td>0.731567</td>\n",
       "      <td>0.688550</td>\n",
       "      <td>0.686971</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>0.744255</td>\n",
       "      <td>0.747013</td>\n",
       "      <td>0.773161</td>\n",
       "      <td>0.764520</td>\n",
       "      <td>0.744402</td>\n",
       "      <td>0.740711</td>\n",
       "      <td>0.742413</td>\n",
       "      <td>0.748835</td>\n",
       "      <td>0.688991</td>\n",
       "      <td>0.687687</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>0.697606</td>\n",
       "      <td>0.702401</td>\n",
       "      <td>0.704359</td>\n",
       "      <td>0.702809</td>\n",
       "      <td>0.701534</td>\n",
       "      <td>0.701398</td>\n",
       "      <td>0.695016</td>\n",
       "      <td>0.702711</td>\n",
       "      <td>0.685002</td>\n",
       "      <td>0.690607</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0.788325</td>\n",
       "      <td>0.866975</td>\n",
       "      <td>0.736418</td>\n",
       "      <td>0.738782</td>\n",
       "      <td>0.784557</td>\n",
       "      <td>0.784291</td>\n",
       "      <td>0.794389</td>\n",
       "      <td>0.843431</td>\n",
       "      <td>0.690093</td>\n",
       "      <td>0.693465</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>0.784287</td>\n",
       "      <td>0.851129</td>\n",
       "      <td>0.737207</td>\n",
       "      <td>0.733792</td>\n",
       "      <td>0.779929</td>\n",
       "      <td>0.777619</td>\n",
       "      <td>0.796996</td>\n",
       "      <td>0.827492</td>\n",
       "      <td>0.690166</td>\n",
       "      <td>0.693032</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>0.737965</td>\n",
       "      <td>0.736958</td>\n",
       "      <td>0.769602</td>\n",
       "      <td>0.766593</td>\n",
       "      <td>0.739857</td>\n",
       "      <td>0.737867</td>\n",
       "      <td>0.734781</td>\n",
       "      <td>0.747455</td>\n",
       "      <td>0.685368</td>\n",
       "      <td>0.690778</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>0.687067</td>\n",
       "      <td>0.693814</td>\n",
       "      <td>0.690586</td>\n",
       "      <td>0.696170</td>\n",
       "      <td>0.691335</td>\n",
       "      <td>0.690163</td>\n",
       "      <td>0.688544</td>\n",
       "      <td>0.694346</td>\n",
       "      <td>0.823948</td>\n",
       "      <td>0.802855</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>0.687116</td>\n",
       "      <td>0.690694</td>\n",
       "      <td>0.692040</td>\n",
       "      <td>0.694352</td>\n",
       "      <td>0.689987</td>\n",
       "      <td>0.687206</td>\n",
       "      <td>0.688456</td>\n",
       "      <td>0.692051</td>\n",
       "      <td>0.835392</td>\n",
       "      <td>0.809665</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>0.680353</td>\n",
       "      <td>0.690178</td>\n",
       "      <td>0.688861</td>\n",
       "      <td>0.690286</td>\n",
       "      <td>0.685481</td>\n",
       "      <td>0.683610</td>\n",
       "      <td>0.684907</td>\n",
       "      <td>0.682722</td>\n",
       "      <td>0.819087</td>\n",
       "      <td>0.774057</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>0.743997</td>\n",
       "      <td>0.737109</td>\n",
       "      <td>0.762498</td>\n",
       "      <td>0.764336</td>\n",
       "      <td>0.736986</td>\n",
       "      <td>0.741156</td>\n",
       "      <td>0.728631</td>\n",
       "      <td>0.741744</td>\n",
       "      <td>0.687240</td>\n",
       "      <td>0.693568</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>0.738776</td>\n",
       "      <td>0.729953</td>\n",
       "      <td>0.758641</td>\n",
       "      <td>0.757234</td>\n",
       "      <td>0.731889</td>\n",
       "      <td>0.731237</td>\n",
       "      <td>0.728438</td>\n",
       "      <td>0.743948</td>\n",
       "      <td>0.684695</td>\n",
       "      <td>0.682731</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>0.702732</td>\n",
       "      <td>0.706165</td>\n",
       "      <td>0.708429</td>\n",
       "      <td>0.715292</td>\n",
       "      <td>0.703266</td>\n",
       "      <td>0.700912</td>\n",
       "      <td>0.707462</td>\n",
       "      <td>0.709255</td>\n",
       "      <td>0.689281</td>\n",
       "      <td>0.690929</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>0.713155</td>\n",
       "      <td>0.711962</td>\n",
       "      <td>0.722878</td>\n",
       "      <td>0.723102</td>\n",
       "      <td>0.713194</td>\n",
       "      <td>0.707450</td>\n",
       "      <td>0.718164</td>\n",
       "      <td>0.717054</td>\n",
       "      <td>0.687752</td>\n",
       "      <td>0.691585</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>0.684091</td>\n",
       "      <td>0.690592</td>\n",
       "      <td>0.688112</td>\n",
       "      <td>0.689219</td>\n",
       "      <td>0.682467</td>\n",
       "      <td>0.685541</td>\n",
       "      <td>0.688200</td>\n",
       "      <td>0.688216</td>\n",
       "      <td>0.816531</td>\n",
       "      <td>0.788674</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>0.698752</td>\n",
       "      <td>0.692118</td>\n",
       "      <td>0.697944</td>\n",
       "      <td>0.703711</td>\n",
       "      <td>0.698021</td>\n",
       "      <td>0.695307</td>\n",
       "      <td>0.694548</td>\n",
       "      <td>0.701628</td>\n",
       "      <td>0.688789</td>\n",
       "      <td>0.686925</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>0.705598</td>\n",
       "      <td>0.695493</td>\n",
       "      <td>0.712660</td>\n",
       "      <td>0.712546</td>\n",
       "      <td>0.700052</td>\n",
       "      <td>0.699860</td>\n",
       "      <td>0.701532</td>\n",
       "      <td>0.702778</td>\n",
       "      <td>0.694551</td>\n",
       "      <td>0.700835</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>0.706860</td>\n",
       "      <td>0.715474</td>\n",
       "      <td>0.717571</td>\n",
       "      <td>0.714976</td>\n",
       "      <td>0.710234</td>\n",
       "      <td>0.703376</td>\n",
       "      <td>0.715087</td>\n",
       "      <td>0.725350</td>\n",
       "      <td>0.682435</td>\n",
       "      <td>0.690065</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>0.718824</td>\n",
       "      <td>0.714309</td>\n",
       "      <td>0.723794</td>\n",
       "      <td>0.721167</td>\n",
       "      <td>0.711160</td>\n",
       "      <td>0.707423</td>\n",
       "      <td>0.716006</td>\n",
       "      <td>0.723137</td>\n",
       "      <td>0.691859</td>\n",
       "      <td>0.690037</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>0.701668</td>\n",
       "      <td>0.697649</td>\n",
       "      <td>0.704354</td>\n",
       "      <td>0.710482</td>\n",
       "      <td>0.707100</td>\n",
       "      <td>0.698657</td>\n",
       "      <td>0.694239</td>\n",
       "      <td>0.705609</td>\n",
       "      <td>0.686181</td>\n",
       "      <td>0.683526</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>0.691024</td>\n",
       "      <td>0.691129</td>\n",
       "      <td>0.698760</td>\n",
       "      <td>0.706687</td>\n",
       "      <td>0.697803</td>\n",
       "      <td>0.695513</td>\n",
       "      <td>0.694895</td>\n",
       "      <td>0.698695</td>\n",
       "      <td>0.679965</td>\n",
       "      <td>0.693369</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>0.700370</td>\n",
       "      <td>0.706583</td>\n",
       "      <td>0.710043</td>\n",
       "      <td>0.708987</td>\n",
       "      <td>0.700352</td>\n",
       "      <td>0.699349</td>\n",
       "      <td>0.700087</td>\n",
       "      <td>0.703738</td>\n",
       "      <td>0.684724</td>\n",
       "      <td>0.690792</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>0.681056</td>\n",
       "      <td>0.691002</td>\n",
       "      <td>0.684735</td>\n",
       "      <td>0.685174</td>\n",
       "      <td>0.683047</td>\n",
       "      <td>0.684746</td>\n",
       "      <td>0.683687</td>\n",
       "      <td>0.689509</td>\n",
       "      <td>0.817576</td>\n",
       "      <td>0.785813</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>0.690633</td>\n",
       "      <td>0.692940</td>\n",
       "      <td>0.690036</td>\n",
       "      <td>0.689574</td>\n",
       "      <td>0.686112</td>\n",
       "      <td>0.688727</td>\n",
       "      <td>0.686761</td>\n",
       "      <td>0.690451</td>\n",
       "      <td>0.729768</td>\n",
       "      <td>0.731972</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>0.687736</td>\n",
       "      <td>0.692359</td>\n",
       "      <td>0.693352</td>\n",
       "      <td>0.694434</td>\n",
       "      <td>0.689820</td>\n",
       "      <td>0.694187</td>\n",
       "      <td>0.690536</td>\n",
       "      <td>0.690528</td>\n",
       "      <td>0.743410</td>\n",
       "      <td>0.748876</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>0.690687</td>\n",
       "      <td>0.690091</td>\n",
       "      <td>0.690112</td>\n",
       "      <td>0.694178</td>\n",
       "      <td>0.689018</td>\n",
       "      <td>0.689289</td>\n",
       "      <td>0.680350</td>\n",
       "      <td>0.685460</td>\n",
       "      <td>0.741756</td>\n",
       "      <td>0.736080</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>0.692125</td>\n",
       "      <td>0.691724</td>\n",
       "      <td>0.691936</td>\n",
       "      <td>0.692440</td>\n",
       "      <td>0.686848</td>\n",
       "      <td>0.689465</td>\n",
       "      <td>0.690511</td>\n",
       "      <td>0.689152</td>\n",
       "      <td>0.739334</td>\n",
       "      <td>0.740906</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>0.684560</td>\n",
       "      <td>0.691371</td>\n",
       "      <td>0.686024</td>\n",
       "      <td>0.687559</td>\n",
       "      <td>0.687008</td>\n",
       "      <td>0.686352</td>\n",
       "      <td>0.685443</td>\n",
       "      <td>0.688599</td>\n",
       "      <td>0.703583</td>\n",
       "      <td>0.700832</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>0.703343</td>\n",
       "      <td>0.702358</td>\n",
       "      <td>0.703257</td>\n",
       "      <td>0.705792</td>\n",
       "      <td>0.705602</td>\n",
       "      <td>0.697844</td>\n",
       "      <td>0.707694</td>\n",
       "      <td>0.709830</td>\n",
       "      <td>0.687566</td>\n",
       "      <td>0.687421</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>0.705778</td>\n",
       "      <td>0.701213</td>\n",
       "      <td>0.703342</td>\n",
       "      <td>0.707115</td>\n",
       "      <td>0.704742</td>\n",
       "      <td>0.699392</td>\n",
       "      <td>0.703710</td>\n",
       "      <td>0.701941</td>\n",
       "      <td>0.681906</td>\n",
       "      <td>0.684698</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>0.694266</td>\n",
       "      <td>0.687586</td>\n",
       "      <td>0.691015</td>\n",
       "      <td>0.692222</td>\n",
       "      <td>0.690423</td>\n",
       "      <td>0.690612</td>\n",
       "      <td>0.685751</td>\n",
       "      <td>0.688394</td>\n",
       "      <td>0.703370</td>\n",
       "      <td>0.700512</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>0.688326</td>\n",
       "      <td>0.688645</td>\n",
       "      <td>0.691236</td>\n",
       "      <td>0.691835</td>\n",
       "      <td>0.688623</td>\n",
       "      <td>0.685965</td>\n",
       "      <td>0.686672</td>\n",
       "      <td>0.687833</td>\n",
       "      <td>0.706672</td>\n",
       "      <td>0.697084</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>0.689960</td>\n",
       "      <td>0.690364</td>\n",
       "      <td>0.689147</td>\n",
       "      <td>0.689936</td>\n",
       "      <td>0.693764</td>\n",
       "      <td>0.694116</td>\n",
       "      <td>0.682763</td>\n",
       "      <td>0.690901</td>\n",
       "      <td>0.692914</td>\n",
       "      <td>0.707947</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>0.693766</td>\n",
       "      <td>0.689802</td>\n",
       "      <td>0.693989</td>\n",
       "      <td>0.694470</td>\n",
       "      <td>0.693055</td>\n",
       "      <td>0.695302</td>\n",
       "      <td>0.688101</td>\n",
       "      <td>0.695158</td>\n",
       "      <td>0.697632</td>\n",
       "      <td>0.705065</td>\n",
       "      <td>...</td>\n",
       "      <td>0.947391</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>0.692385</td>\n",
       "      <td>0.688524</td>\n",
       "      <td>0.697319</td>\n",
       "      <td>0.695939</td>\n",
       "      <td>0.690428</td>\n",
       "      <td>0.688607</td>\n",
       "      <td>0.686674</td>\n",
       "      <td>0.693175</td>\n",
       "      <td>0.698229</td>\n",
       "      <td>0.700515</td>\n",
       "      <td>...</td>\n",
       "      <td>0.769756</td>\n",
       "      <td>0.942668</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>0.693111</td>\n",
       "      <td>0.692672</td>\n",
       "      <td>0.694074</td>\n",
       "      <td>0.691459</td>\n",
       "      <td>0.688377</td>\n",
       "      <td>0.698549</td>\n",
       "      <td>0.691526</td>\n",
       "      <td>0.691099</td>\n",
       "      <td>0.694301</td>\n",
       "      <td>0.699509</td>\n",
       "      <td>...</td>\n",
       "      <td>0.772508</td>\n",
       "      <td>0.753439</td>\n",
       "      <td>0.979212</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>0.691011</td>\n",
       "      <td>0.688959</td>\n",
       "      <td>0.694905</td>\n",
       "      <td>0.693554</td>\n",
       "      <td>0.689306</td>\n",
       "      <td>0.691381</td>\n",
       "      <td>0.685588</td>\n",
       "      <td>0.688846</td>\n",
       "      <td>0.701178</td>\n",
       "      <td>0.703580</td>\n",
       "      <td>...</td>\n",
       "      <td>0.785098</td>\n",
       "      <td>0.767045</td>\n",
       "      <td>0.795907</td>\n",
       "      <td>0.906631</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>0.693976</td>\n",
       "      <td>0.694550</td>\n",
       "      <td>0.691529</td>\n",
       "      <td>0.692475</td>\n",
       "      <td>0.688289</td>\n",
       "      <td>0.689564</td>\n",
       "      <td>0.687903</td>\n",
       "      <td>0.688923</td>\n",
       "      <td>0.703577</td>\n",
       "      <td>0.702598</td>\n",
       "      <td>...</td>\n",
       "      <td>0.787864</td>\n",
       "      <td>0.777159</td>\n",
       "      <td>0.806227</td>\n",
       "      <td>0.842220</td>\n",
       "      <td>0.953763</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>0.679432</td>\n",
       "      <td>0.690538</td>\n",
       "      <td>0.684203</td>\n",
       "      <td>0.686743</td>\n",
       "      <td>0.682930</td>\n",
       "      <td>0.681391</td>\n",
       "      <td>0.683244</td>\n",
       "      <td>0.689697</td>\n",
       "      <td>0.823738</td>\n",
       "      <td>0.791087</td>\n",
       "      <td>...</td>\n",
       "      <td>0.700246</td>\n",
       "      <td>0.694669</td>\n",
       "      <td>0.695050</td>\n",
       "      <td>0.699585</td>\n",
       "      <td>0.698479</td>\n",
       "      <td>0.960456</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>0.685236</td>\n",
       "      <td>0.690360</td>\n",
       "      <td>0.694246</td>\n",
       "      <td>0.694216</td>\n",
       "      <td>0.685508</td>\n",
       "      <td>0.682246</td>\n",
       "      <td>0.687901</td>\n",
       "      <td>0.689574</td>\n",
       "      <td>0.860703</td>\n",
       "      <td>0.801236</td>\n",
       "      <td>...</td>\n",
       "      <td>0.698560</td>\n",
       "      <td>0.698201</td>\n",
       "      <td>0.691345</td>\n",
       "      <td>0.699323</td>\n",
       "      <td>0.699038</td>\n",
       "      <td>0.827916</td>\n",
       "      <td>0.954331</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>0.688035</td>\n",
       "      <td>0.690379</td>\n",
       "      <td>0.688829</td>\n",
       "      <td>0.691500</td>\n",
       "      <td>0.685510</td>\n",
       "      <td>0.685711</td>\n",
       "      <td>0.684255</td>\n",
       "      <td>0.690167</td>\n",
       "      <td>0.886172</td>\n",
       "      <td>0.804386</td>\n",
       "      <td>...</td>\n",
       "      <td>0.702254</td>\n",
       "      <td>0.698676</td>\n",
       "      <td>0.698214</td>\n",
       "      <td>0.703386</td>\n",
       "      <td>0.702963</td>\n",
       "      <td>0.831521</td>\n",
       "      <td>0.880014</td>\n",
       "      <td>0.941582</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>0.683845</td>\n",
       "      <td>0.689289</td>\n",
       "      <td>0.685486</td>\n",
       "      <td>0.687532</td>\n",
       "      <td>0.683954</td>\n",
       "      <td>0.683999</td>\n",
       "      <td>0.683132</td>\n",
       "      <td>0.690396</td>\n",
       "      <td>0.822307</td>\n",
       "      <td>0.793323</td>\n",
       "      <td>...</td>\n",
       "      <td>0.696911</td>\n",
       "      <td>0.697188</td>\n",
       "      <td>0.690900</td>\n",
       "      <td>0.698664</td>\n",
       "      <td>0.702934</td>\n",
       "      <td>0.848660</td>\n",
       "      <td>0.829399</td>\n",
       "      <td>0.838053</td>\n",
       "      <td>0.984710</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>0.686044</td>\n",
       "      <td>0.690321</td>\n",
       "      <td>0.692043</td>\n",
       "      <td>0.693004</td>\n",
       "      <td>0.681497</td>\n",
       "      <td>0.685688</td>\n",
       "      <td>0.687570</td>\n",
       "      <td>0.688119</td>\n",
       "      <td>0.822173</td>\n",
       "      <td>0.785671</td>\n",
       "      <td>...</td>\n",
       "      <td>0.700835</td>\n",
       "      <td>0.698553</td>\n",
       "      <td>0.691451</td>\n",
       "      <td>0.699616</td>\n",
       "      <td>0.704693</td>\n",
       "      <td>0.836758</td>\n",
       "      <td>0.829312</td>\n",
       "      <td>0.830575</td>\n",
       "      <td>0.877518</td>\n",
       "      <td>0.978957</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56 rows × 56 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0         1         2         3         4         5         6   \\\n",
       "0   0.934816  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "1   0.783949  0.977619  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "2   0.750807  0.738941  0.950710  0.000000  0.000000  0.000000  0.000000   \n",
       "3   0.748160  0.738134  0.851631  0.936520  0.000000  0.000000  0.000000   \n",
       "4   0.854389  0.784220  0.749937  0.749702  0.949022  0.000000  0.000000   \n",
       "5   0.840005  0.778977  0.745428  0.743418  0.844897  0.921687  0.000000   \n",
       "6   0.784556  0.799953  0.744342  0.741278  0.782100  0.780424  0.951716   \n",
       "7   0.804715  0.840063  0.755351  0.745741  0.805334  0.794352  0.842448   \n",
       "8   0.685939  0.689590  0.691690  0.690281  0.684042  0.684396  0.685473   \n",
       "9   0.683152  0.694775  0.691760  0.691238  0.688755  0.688081  0.695693   \n",
       "10  0.685014  0.695168  0.695434  0.694043  0.688178  0.688559  0.691181   \n",
       "11  0.728886  0.727167  0.727936  0.741250  0.730548  0.730278  0.721784   \n",
       "12  0.688921  0.696588  0.692216  0.692601  0.688108  0.684220  0.687794   \n",
       "13  0.725224  0.720202  0.722864  0.730578  0.720414  0.723844  0.718734   \n",
       "14  0.729473  0.726418  0.725197  0.738215  0.731152  0.727414  0.722559   \n",
       "15  0.744255  0.747013  0.773161  0.764520  0.744402  0.740711  0.742413   \n",
       "16  0.697606  0.702401  0.704359  0.702809  0.701534  0.701398  0.695016   \n",
       "17  0.788325  0.866975  0.736418  0.738782  0.784557  0.784291  0.794389   \n",
       "18  0.784287  0.851129  0.737207  0.733792  0.779929  0.777619  0.796996   \n",
       "19  0.737965  0.736958  0.769602  0.766593  0.739857  0.737867  0.734781   \n",
       "20  0.687067  0.693814  0.690586  0.696170  0.691335  0.690163  0.688544   \n",
       "21  0.687116  0.690694  0.692040  0.694352  0.689987  0.687206  0.688456   \n",
       "22  0.680353  0.690178  0.688861  0.690286  0.685481  0.683610  0.684907   \n",
       "23  0.743997  0.737109  0.762498  0.764336  0.736986  0.741156  0.728631   \n",
       "24  0.738776  0.729953  0.758641  0.757234  0.731889  0.731237  0.728438   \n",
       "25  0.702732  0.706165  0.708429  0.715292  0.703266  0.700912  0.707462   \n",
       "26  0.713155  0.711962  0.722878  0.723102  0.713194  0.707450  0.718164   \n",
       "27  0.684091  0.690592  0.688112  0.689219  0.682467  0.685541  0.688200   \n",
       "28  0.698752  0.692118  0.697944  0.703711  0.698021  0.695307  0.694548   \n",
       "29  0.705598  0.695493  0.712660  0.712546  0.700052  0.699860  0.701532   \n",
       "30  0.706860  0.715474  0.717571  0.714976  0.710234  0.703376  0.715087   \n",
       "31  0.718824  0.714309  0.723794  0.721167  0.711160  0.707423  0.716006   \n",
       "32  0.701668  0.697649  0.704354  0.710482  0.707100  0.698657  0.694239   \n",
       "33  0.691024  0.691129  0.698760  0.706687  0.697803  0.695513  0.694895   \n",
       "34  0.700370  0.706583  0.710043  0.708987  0.700352  0.699349  0.700087   \n",
       "35  0.681056  0.691002  0.684735  0.685174  0.683047  0.684746  0.683687   \n",
       "36  0.690633  0.692940  0.690036  0.689574  0.686112  0.688727  0.686761   \n",
       "37  0.687736  0.692359  0.693352  0.694434  0.689820  0.694187  0.690536   \n",
       "38  0.690687  0.690091  0.690112  0.694178  0.689018  0.689289  0.680350   \n",
       "39  0.692125  0.691724  0.691936  0.692440  0.686848  0.689465  0.690511   \n",
       "40  0.684560  0.691371  0.686024  0.687559  0.687008  0.686352  0.685443   \n",
       "41  0.703343  0.702358  0.703257  0.705792  0.705602  0.697844  0.707694   \n",
       "42  0.705778  0.701213  0.703342  0.707115  0.704742  0.699392  0.703710   \n",
       "43  0.694266  0.687586  0.691015  0.692222  0.690423  0.690612  0.685751   \n",
       "44  0.688326  0.688645  0.691236  0.691835  0.688623  0.685965  0.686672   \n",
       "45  0.689960  0.690364  0.689147  0.689936  0.693764  0.694116  0.682763   \n",
       "46  0.693766  0.689802  0.693989  0.694470  0.693055  0.695302  0.688101   \n",
       "47  0.692385  0.688524  0.697319  0.695939  0.690428  0.688607  0.686674   \n",
       "48  0.693111  0.692672  0.694074  0.691459  0.688377  0.698549  0.691526   \n",
       "49  0.691011  0.688959  0.694905  0.693554  0.689306  0.691381  0.685588   \n",
       "50  0.693976  0.694550  0.691529  0.692475  0.688289  0.689564  0.687903   \n",
       "51  0.679432  0.690538  0.684203  0.686743  0.682930  0.681391  0.683244   \n",
       "52  0.685236  0.690360  0.694246  0.694216  0.685508  0.682246  0.687901   \n",
       "53  0.688035  0.690379  0.688829  0.691500  0.685510  0.685711  0.684255   \n",
       "54  0.683845  0.689289  0.685486  0.687532  0.683954  0.683999  0.683132   \n",
       "55  0.686044  0.690321  0.692043  0.693004  0.681497  0.685688  0.687570   \n",
       "\n",
       "          7         8         9   ...        46        47        48        49  \\\n",
       "0   0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "1   0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "2   0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "3   0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "4   0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "5   0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "6   0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "7   0.977010  0.000000  0.000000  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "8   0.684307  0.981559  0.000000  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "9   0.692221  0.792899  0.968273  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "10  0.690430  0.793459  0.849706  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "11  0.720931  0.688170  0.690929  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "12  0.691487  0.835340  0.811437  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "13  0.724085  0.681713  0.688961  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "14  0.731567  0.688550  0.686971  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "15  0.748835  0.688991  0.687687  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "16  0.702711  0.685002  0.690607  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "17  0.843431  0.690093  0.693465  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "18  0.827492  0.690166  0.693032  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "19  0.747455  0.685368  0.690778  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "20  0.694346  0.823948  0.802855  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "21  0.692051  0.835392  0.809665  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "22  0.682722  0.819087  0.774057  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "23  0.741744  0.687240  0.693568  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "24  0.743948  0.684695  0.682731  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "25  0.709255  0.689281  0.690929  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "26  0.717054  0.687752  0.691585  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "27  0.688216  0.816531  0.788674  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "28  0.701628  0.688789  0.686925  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "29  0.702778  0.694551  0.700835  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "30  0.725350  0.682435  0.690065  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "31  0.723137  0.691859  0.690037  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "32  0.705609  0.686181  0.683526  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "33  0.698695  0.679965  0.693369  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "34  0.703738  0.684724  0.690792  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "35  0.689509  0.817576  0.785813  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "36  0.690451  0.729768  0.731972  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "37  0.690528  0.743410  0.748876  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "38  0.685460  0.741756  0.736080  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "39  0.689152  0.739334  0.740906  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "40  0.688599  0.703583  0.700832  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "41  0.709830  0.687566  0.687421  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "42  0.701941  0.681906  0.684698  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "43  0.688394  0.703370  0.700512  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "44  0.687833  0.706672  0.697084  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "45  0.690901  0.692914  0.707947  ...  0.000000  0.000000  0.000000  0.000000   \n",
       "46  0.695158  0.697632  0.705065  ...  0.947391  0.000000  0.000000  0.000000   \n",
       "47  0.693175  0.698229  0.700515  ...  0.769756  0.942668  0.000000  0.000000   \n",
       "48  0.691099  0.694301  0.699509  ...  0.772508  0.753439  0.979212  0.000000   \n",
       "49  0.688846  0.701178  0.703580  ...  0.785098  0.767045  0.795907  0.906631   \n",
       "50  0.688923  0.703577  0.702598  ...  0.787864  0.777159  0.806227  0.842220   \n",
       "51  0.689697  0.823738  0.791087  ...  0.700246  0.694669  0.695050  0.699585   \n",
       "52  0.689574  0.860703  0.801236  ...  0.698560  0.698201  0.691345  0.699323   \n",
       "53  0.690167  0.886172  0.804386  ...  0.702254  0.698676  0.698214  0.703386   \n",
       "54  0.690396  0.822307  0.793323  ...  0.696911  0.697188  0.690900  0.698664   \n",
       "55  0.688119  0.822173  0.785671  ...  0.700835  0.698553  0.691451  0.699616   \n",
       "\n",
       "          50        51        52        53        54        55  \n",
       "0   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "1   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "2   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "3   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "4   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "5   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "6   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "7   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "8   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "9   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "10  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "11  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "12  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "13  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "14  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "15  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "16  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "17  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "18  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "19  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "20  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "21  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "22  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "23  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "24  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "25  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "26  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "27  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "28  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "29  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "30  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "31  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "32  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "33  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "34  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "35  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "36  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "37  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "38  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "39  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "40  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "41  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "42  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "43  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "44  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "45  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "46  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "47  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "48  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "49  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "50  0.953763  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "51  0.698479  0.960456  0.000000  0.000000  0.000000  0.000000  \n",
       "52  0.699038  0.827916  0.954331  0.000000  0.000000  0.000000  \n",
       "53  0.702963  0.831521  0.880014  0.941582  0.000000  0.000000  \n",
       "54  0.702934  0.848660  0.829399  0.838053  0.984710  0.000000  \n",
       "55  0.704693  0.836758  0.829312  0.830575  0.877518  0.978957  \n",
       "\n",
       "[56 rows x 56 columns]"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_difference(cluster, genome, 4, proc = 8, n=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "running-deputy",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
