{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "amended-victim",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sys\n",
    "import os\n",
    "import re\n",
    "import csv\n",
    "import collections as co\n",
    "import itertools as it\n",
    "import umap\n",
    "import hdbscan\n",
    "import time \n",
    "import progressbar\n",
    "import math\n",
    "import random\n",
    "import fileinput\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.transforms as transforms\n",
    "import scipy.spatial.distance as ssd\n",
    "import seaborn as sns\n",
    "\n",
    "from Bio import Align, AlignIO, Phylo\n",
    "from Bio.Align.Applications import MafftCommandline\n",
    "from Bio.Seq import Seq \n",
    "from Bio.Phylo.TreeConstruction import DistanceTreeConstructor, DistanceMatrix, DistanceCalculator\n",
    "from ete3 import PhyloTree, Tree, faces, AttrFace, CircleFace, TreeStyle, NodeStyle, TextFace, SequenceFace\n",
    "from IPython.display import display\n",
    "from matplotlib import colors \n",
    "from scipy.cluster import hierarchy\n",
    "from sklearn.preprocessing import normalize\n",
    "from sklearn.decomposition import PCA\n",
    "from statistics import mean\n",
    "from scipy import stats\n",
    "from kneed import DataGenerator, KneeLocator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "neither-brunei",
   "metadata": {},
   "outputs": [],
   "source": [
    "class frequency(object):\n",
    "    \n",
    "    def __init__(self, k = 7, convert = 0):\n",
    "    \n",
    "        self.k = k\n",
    "        self.convert = convert\n",
    "        self.exist = co.defaultdict(int) \n",
    "        self.keys = list(self.exist.keys())\n",
    "        self.col = len(self.keys)\n",
    "        self.row = 0\n",
    "        self.matrix = np.empty((self.row, self.col, ),dtype = \"float64\")\n",
    "        self.amino = co.defaultdict(str, {\n",
    "            'AAA':'K', 'AAC':'N', 'AAG':'K', 'AAT':'N',\n",
    "            'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "            'AGA':'R', 'AGC':'S', 'AGG':'R', 'AGT':'S',\n",
    "            'ATA':'I', 'ATC':'I', 'ATG':'M', 'ATT':'I',\n",
    "            'CAA':'Q', 'CAC':'H', 'CAG':'Q', 'CAT':'H',\n",
    "            'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "            'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "            'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "            'GAA':'E', 'GAC':'D', 'GAG':'E', 'GAT':'D',\n",
    "            'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "            'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "            'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',    \n",
    "            'TAA':'Y', 'TAC':'*', 'TAG':'*', 'TAT':'Y',\n",
    "            'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "            'TGA':'*', 'TGC':'C', 'TGG':'W', 'TGT':'C',\n",
    "            'TTA':'L', 'TTC':'F', 'TTG':'L', 'TTT':'F'\n",
    "        })\n",
    "                \n",
    "    def translate(self, read):\n",
    "    \n",
    "        chain = ''\n",
    "\n",
    "        for i in range(len(read) - 2):\n",
    "            trip = read[i:i+3]\n",
    "            chain += self.amino[trip]\n",
    "\n",
    "        return(chain)\n",
    "    \n",
    "    \n",
    "    def adjust_to_data(self, infile):\n",
    "    \n",
    "        self.row = infile.shape[0]\n",
    "            \n",
    "        for line, read in infile.itertuples(index=True, name=None):\n",
    "\n",
    "            if self.convert == 1:\n",
    "                seq = self.translate(read)\n",
    "                del read\n",
    "\n",
    "                num = len(seq) - self.k + 1\n",
    "\n",
    "                for i in range(num):\n",
    "                    kmer = seq[i:i+self.k]\n",
    "                    self.exist[kmer] = 0\n",
    "\n",
    "            else:\n",
    "                seq = read\n",
    "                del read\n",
    "\n",
    "                num = len(seq) - self.k + 1\n",
    "\n",
    "                if re.match('^[ACGT]*$', seq): \n",
    "                    for i in range(num):\n",
    "                        kmer = seq[i:i+self.k]\n",
    "                        self.exist[kmer] = 0\n",
    "                else:\n",
    "                    for i in range(num):\n",
    "                        kmer = seq[i:i+self.k]\n",
    "                        if re.match('^[ACGT]*$', kmer): \n",
    "                            self.exist[kmer] = 0\n",
    "            \n",
    "        self.keys = list(self.exist.keys())\n",
    "        self.col = len(self.keys)\n",
    "        self.matrix = np.empty((self.row, self.col, ), dtype=\"float64\")\n",
    "        \n",
    "        del seq\n",
    "    \n",
    "    \n",
    "    def calculate_frequence(self, infile):\n",
    "        \n",
    "        for line, read in infile.itertuples(index=True, name=None): \n",
    "                 \n",
    "            if self.convert == 1:\n",
    "                seq = self.translate(read)\n",
    "                del read\n",
    "\n",
    "                counts = self.exist.copy()\n",
    "                num = len(seq) - self.k + 1\n",
    "\n",
    "                for i in range(num):\n",
    "                    kmer = seq[i:i+self.k]\n",
    "                    counts[kmer] += 1\n",
    "\n",
    "            else:\n",
    "                seq = read\n",
    "                del read\n",
    "\n",
    "                counts = self.exist.copy()\n",
    "                num = len(seq) - self.k + 1\n",
    "\n",
    "                if re.match('^[ACGT]*$', seq): \n",
    "                    for i in range(num):\n",
    "                        kmer = seq[i:i+self.k]\n",
    "                        counts[kmer] += 1\n",
    "                else:\n",
    "                    for i in range(num):\n",
    "                        kmer = seq[i:i+self.k]\n",
    "                        if re.match('^[ACGT]*$', kmer): \n",
    "                            counts[kmer] += 1\n",
    "\n",
    "            vector = np.array(list(counts.values()), dtype = \"float64\")\n",
    "\n",
    "            self.matrix[line] = vector/vector.sum()\n",
    "            \n",
    "            counts.clear()\n",
    "            del vector\n",
    "            del seq\n",
    "            del counts\n",
    "    \n",
    "    \n",
    "    def get_keys(self):\n",
    "        \n",
    "        return(self.keys)\n",
    "    \n",
    "    def get_matrix(self):\n",
    "        \n",
    "        return(self.matrix)\n",
    "\n",
    "\n",
    "# In[3]:\n",
    "\n",
    "\n",
    "def subsplit(subtype):\n",
    "    if re.match('^[H][0-9]+N[0-9]+$', subtype): \n",
    "        H = re.search('[H][0-9]+', subtype).group(0)\n",
    "        N = re.search('[N][0-9]+', subtype).group(0)\n",
    "    else:\n",
    "        H = ''#np.nan\n",
    "        N = ''#np.nan\n",
    "    return(H, N)\n",
    "\n",
    "\n",
    "# In[4]:\n",
    "\n",
    "\n",
    "def convert_fasta(infile, outfile):\n",
    "\n",
    "    #with some help from https://stackoverflow.com/questions/50856538/how-to-convert-multiline-fasta-files-to-singleline-fasta-files-without-biopython\n",
    "    \n",
    "    with open(infile) as f_input, open(outfile, 'w') as f_output:\n",
    "\n",
    "        block = []\n",
    "\n",
    "        for line in f_input:\n",
    "    \n",
    "            if line.startswith('>'):\n",
    "\n",
    "                if block:\n",
    "                    f_output.write(new_line + ',' + \"\".join(block) + '\\n')\n",
    "                    block = []\n",
    "                \n",
    "                new_line = line.strip().replace('|', ',')\n",
    "            else:\n",
    "                block.append(line.strip())\n",
    "    \n",
    "        if block:\n",
    "            f_output.write(new_line + ',' + \"\".join(block) + '\\n')\n",
    "\n",
    "\n",
    "# In[5]:\n",
    "\n",
    "\n",
    "def get_vectors(sequence, accession, metric, neigh, comp, pca):\n",
    "\n",
    "    freq = frequency(k = 7, convert = 0)\n",
    "    freq.adjust_to_data(sequence)\n",
    "    freq.calculate_frequence(sequence)\n",
    "\n",
    "    #Kevins approach\n",
    "    #freq = frequency(k = 7)\n",
    "    #freq.calculate_frequence(sequence)\n",
    "    \n",
    "    if pca:\n",
    "        matrix = freq.get_matrix()\n",
    "        pca_init = PCA(n_components=pca)\n",
    "        pca_matr = pca_init.fit_transform(matrix)\n",
    "        explained_var = pca_init.explained_variance_ratio_.sum()\n",
    "        #pca_matr = freq.get_matrix()\n",
    "\n",
    "        dataframe_pca = pd.concat([accession, pd.DataFrame(pca_matr)], axis=1, copy = False, ignore_index = False).set_index('accession')\n",
    "    else:\n",
    "        pca_matr = freq.get_matrix()\n",
    "        dataframe_pca = pd.DataFrame()\n",
    "        explained_var = -1\n",
    "    \n",
    "    \n",
    "    if (pca and not comp) or (pca <= comp):\n",
    "        reduced = pca_matr\n",
    "\n",
    "    elif neigh and comp:\n",
    "        reduced = umap.UMAP(\n",
    "            n_neighbors = neigh,\n",
    "            min_dist = 0.0,\n",
    "            n_components = comp,\n",
    "            random_state = 42,\n",
    "            #metric = 'euclidean',\n",
    "        ).fit_transform(pca_matr)\n",
    "        \n",
    "    else:\n",
    "        reduced = pca_matr\n",
    "    \n",
    "    #approximation of cosine by l2 normalization (https://github.com/scikit-learn-contrib/hdbscan/issues/69)\n",
    "    if metric == 'cosine':\n",
    "        cos_red = normalize(reduced, norm='l2')\n",
    "        dataframe = pd.concat([accession, pd.DataFrame(cos_red)], axis=1, copy = False, ignore_index = False).set_index('accession')\n",
    "    else:\n",
    "        dataframe = pd.concat([accession, pd.DataFrame(reduced)], axis=1, copy = False, ignore_index = False).set_index('accession')\n",
    "    \n",
    "    return(dataframe, dataframe_pca, explained_var)\n",
    "\n",
    "\n",
    "# In[6]:\n",
    "\n",
    "\n",
    "def get_elbow(dataframe, extra, accession, metric, min_clust, sample, kneedle):\n",
    "    \n",
    "    #with some help from https://joernhees.de/blog/2015/08/26/scipy-hierarchical-clustering-and-dendrogram-tutorial/\n",
    "    \n",
    "    if metric == 'cosine':\n",
    "        metric = 'euclidean'\n",
    "    \n",
    "    clusterer_best = hdbscan.HDBSCAN(\n",
    "        min_samples = sample,\n",
    "        min_cluster_size = min_clust,\n",
    "        gen_min_span_tree = True,\n",
    "        metric = metric,\n",
    "    ).fit(dataframe)\n",
    "    \n",
    "    label = clusterer_best.labels_\n",
    "    label_list = label.tolist()\n",
    "    \n",
    "    if -1 in label_list:\n",
    "        label_list.remove(-1)\n",
    "    \n",
    "    n_cluster = len(set(label_list))\n",
    "    \n",
    "    linkage = clusterer_best.single_linkage_tree_.to_pandas()\n",
    "    numpy_linkage = linkage.drop(columns=['parent']).to_numpy()\n",
    "    \n",
    "    dist = numpy_linkage[:, 2]\n",
    "    dist_rev = dist[::-1]\n",
    "    \n",
    "    if kneedle == -1:\n",
    "        dist_area = dist_rev[0:n_cluster]\n",
    "    else:\n",
    "        dist_area = dist_rev[0:kneedle]\n",
    "    \n",
    "    idxs = np.arange(1, len(dist_area) + 1)\n",
    "\n",
    "    kn = KneeLocator(idxs, dist_area,\n",
    "        curve='convex',\n",
    "        direction='decreasing',\n",
    "        interp_method='polynomial',\n",
    "        online=True,\n",
    "        S = 1.0,\n",
    "    )\n",
    "    \n",
    "    n_cluster_raw = kn.knee\n",
    "    n_cluster_norm = kn.norm_knee\n",
    "    \n",
    "    #linkage.set_index('parent', inplace = True)\n",
    "    elbow = pd.DataFrame({'n_cluster': idxs, \n",
    "                          'distance': dist_area, \n",
    "                          'x_normalized':kn.x_normalized, \n",
    "                          'y_normalized':kn.y_normalized, \n",
    "                          'x_difference':kn.x_difference, \n",
    "                          'y_difference':kn.y_difference}).set_index('n_cluster')\n",
    "    epsilon_best = elbow.loc[n_cluster_raw][['distance']].item()\n",
    "    linkage.set_index('parent', inplace = True)\n",
    "    \n",
    "    return(linkage, elbow, epsilon_best, n_cluster_raw, n_cluster_norm)\n",
    "\n",
    "\n",
    "# In[7]:\n",
    "\n",
    "\n",
    "def get_cluster(epsilon_best, dataframe, extra, accession, metric, min_clust, sample):\n",
    "    \n",
    "    #approximation see above\n",
    "    if metric == 'cosine':\n",
    "        metric = 'euclidean'\n",
    "    \n",
    "    clusterer_best = hdbscan.HDBSCAN(\n",
    "        min_samples = sample,\n",
    "        min_cluster_size = min_clust,\n",
    "        cluster_selection_epsilon = epsilon_best,\n",
    "        gen_min_span_tree=True,\n",
    "        metric = metric,\n",
    "    ).fit(dataframe)\n",
    "\n",
    "    label = clusterer_best.labels_\n",
    "    label_list = label.tolist()\n",
    "    \n",
    "    unclustered = label_list.count(-1)\n",
    "    if -1 in label_list:\n",
    "        label_list.remove(-1)\n",
    "    \n",
    "    n_cluster = len(set(label_list))\n",
    "    \n",
    "    cluster = pd.concat([pd.DataFrame(label, columns = ['cluster']), extra, accession], axis=1, copy = False).set_index('accession')\n",
    "    #linkage = clusterer_best.single_linkage_tree_.to_pandas()\n",
    "    #linkage.set_index('parent', inplace = True)\n",
    "    \n",
    "    N_unmatch = cluster.query('cluster != -1').groupby('cluster').filter(lambda x: x['N'].replace('', np.nan).nunique() > 1).cluster.nunique()\n",
    "    H_unmatch = cluster.query('cluster != -1').groupby('cluster').filter(lambda x: x['H'].replace('', np.nan).nunique() > 1).cluster.nunique() \n",
    "    \n",
    "    cluster.insert(2, 'centroid', False)\n",
    "\n",
    "    num = cluster['cluster'].max()+1\n",
    "    values = [True]*num\n",
    "    accessions = []\n",
    "\n",
    "    for i in range(num):\n",
    "\n",
    "        query = cluster.query('cluster == @i')\n",
    "        match = query.index.values.tolist()\n",
    "        sub = dataframe.loc[match]\n",
    "\n",
    "        if not sub.empty:\n",
    "\n",
    "            tuple_min = (0, '')\n",
    "            for name, row in sub.iterrows():\n",
    "\n",
    "                if not tuple_min[1]:\n",
    "                    dist_min = ssd.cdist([row], sub, metric).mean()\n",
    "                    tuple_min = (dist_min, name)\n",
    "                else:\n",
    "                    dist_min = ssd.cdist([row], sub, metric).mean()\n",
    "                    if dist_min < tuple_min[0]:\n",
    "                        tuple_min = (dist_min, name)\n",
    "\n",
    "            accessions.append(tuple_min[1])\n",
    "\n",
    "    cluster.update(pd.DataFrame(values, columns=['centroid'], index = accessions))\n",
    "    \n",
    "    return(cluster, n_cluster, unclustered, H_unmatch, N_unmatch)\n",
    "\n",
    "\n",
    "# In[8]:\n",
    "\n",
    "\n",
    "def plot_density(density, segments, outpath, render, accuracy, suffix = ''):\n",
    "\n",
    "    for seg in segments:\n",
    "             \n",
    "        with sns.axes_style(\"darkgrid\"):\n",
    "            fig, ax = plt.subplots(figsize=(4,4))\n",
    "            sns.histplot(data = density.query('segment == 4'), x = \"size\", kde=False, log_scale=False, element=\"step\", fill = True, bins = round(density.query('segment == 4')['size'].max()*accuracy), ax = ax)\n",
    "            plt.xlabel(\"log(Count)\")\n",
    "            plt.ylabel(\"#Cluster\")\n",
    "            plt.tight_layout()\n",
    "            plt.close(fig)\n",
    "            fig.savefig(outpath + 'Cluster_Distribution_Segment_' + str(seg) + str(suffix) + '.' + render)\n",
    "    \n",
    "        with sns.axes_style(\"darkgrid\"):\n",
    "            fig, ax = plt.subplots(figsize=(4,4))\n",
    "            sns.histplot(data = density.query('segment == 4'), x = \"size\", kde=False, log_scale=True, element=\"step\", fill = True, bins = round(density.query('segment == 4')['size'].max()*accuracy), ax = ax)\n",
    "            plt.xlabel(\"log(Count)\")\n",
    "            plt.ylabel(\"#Cluster\")\n",
    "            plt.tight_layout()\n",
    "            plt.close(fig)\n",
    "            fig.savefig(outpath + 'Cluster_Distribution_Log_Segment_' + str(seg) + str(suffix) + '.' + render)\n",
    "\n",
    "# In[9]:\n",
    "\n",
    "\n",
    "def plot_elbow(elbow, information, segments, outfolder, render, suffix = ''):\n",
    "\n",
    "        \n",
    "    for seg in segments:\n",
    "    \n",
    "        with sns.axes_style(\"darkgrid\"):\n",
    "            fig, ax = plt.subplots(figsize=(4,4))\n",
    "        \n",
    "            sns.lineplot(data = elbow.query('segment == @seg'), x = \"n_cluster\", y = \"distance\", ax = ax)\n",
    "\n",
    "            y_max = elbow.query('segment == @seg')['distance'].max().item()\n",
    "            x_max = elbow.query('segment == @seg')['n_cluster'].max().item()\n",
    "\n",
    "            y_note = information.query('segment == @seg')['epsilon'].item()\n",
    "            x_note = information.query('segment == @seg')['n_cluster_raw'].item()\n",
    "            z_note = information.query('segment == @seg')['n_cluster'].item()\n",
    "\n",
    "            ax.axhline(y = y_note, alpha=1, color=\"y\", label = f\"Distance = {y_note:0.3f}\")\n",
    "            ax.axvline(x = x_note, alpha=1, color=\"r\", label = f\"#Cluster = {x_note:0.0f}\")\n",
    "            ax.axvline(x = z_note, alpha=1, color=\"g\", label = f\"#Cluster = {z_note:0.0f}\")\n",
    "\n",
    "            #ax.text(x_max, y_note, f\"{y_note:0.3f}\", va='bottom', ha='right', color = \"r\", alpha=1, size=10)\n",
    "            #ax.text(x_note, y_max, f\"{x_note:0.0f}\", rotation='vertical', va='top', ha='left', color = \"r\", alpha=1, size=10)\n",
    "            #ax.text(z_note, y_max, f\"{z_note:0.0f}\", rotation='vertical', va='top', ha='right', color = \"g\", alpha=1, size=10)\n",
    "\n",
    "            plt.xlabel(\"#Cluster\")\n",
    "            plt.ylabel(\"Distance\")\n",
    "        \n",
    "            plt.legend()\n",
    "            plt.tight_layout()\n",
    "            plt.close(fig)\n",
    "            fig.savefig(outfolder + 'Cluster_Elbow_Knee_Segment_' + str(seg) + str(suffix) + '.' + render)\n",
    "\n",
    "        \n",
    "        with sns.axes_style(\"darkgrid\"):\n",
    "            fig, ax = plt.subplots(figsize=(4,4))\n",
    "        \n",
    "            sns.lineplot(data = elbow.query('segment == @seg'), x = \"x_normalized\", y = \"y_normalized\", ax = ax)\n",
    "            sns.lineplot(data = elbow.query('segment == @seg'), x = \"x_difference\", y = \"y_difference\", color=\"g\", ax = ax)\n",
    "        \n",
    "            z_note = information.query('segment == @seg')['n_cluster_norm'].item()\n",
    "            ax.axvline(x = z_note, alpha=1, color=\"r\", label = f\"norm(#Cluster) = {z_note:0.3f}\")\n",
    "            #ax.text(z_note, y_max*0.0, f\"{z_note:0.3f}\", rotation='vertical', va='bottom', ha='left', color = \"r\", alpha=1, size=10)\n",
    "\n",
    "            plt.xlabel(\"norm(#Cluster)\")\n",
    "            plt.ylabel(\"Difference\")\n",
    "        \n",
    "            plt.legend()\n",
    "            plt.tight_layout()\n",
    "            plt.close(fig)\n",
    "            fig.savefig(outfolder + 'Cluster_Knee_Segment_' + str(seg) + str(suffix) + '.' + render)\n",
    "\n",
    "\n",
    "# In[10]:\n",
    "\n",
    "\n",
    "#Modified Version of the code presented on https://github.com/scipy/scipy/issues/8274\n",
    "def getNewick(node, newick, parentdist, leaf_names):\n",
    "    \n",
    "    if node.is_leaf():\n",
    "        return(leaf_names[node.id] + \":\" + np.format_float_positional(parentdist - node.dist) + newick)\n",
    "        #return \"%s:%.5f%s\" % (leaf_names[node.id], parentdist - node.dist, newick)\n",
    "    \n",
    "    else:\n",
    "        if len(newick) > 0:\n",
    "            newick = \"):\" + np.format_float_positional(parentdist - node.dist) + newick\n",
    "            #newick = \"):%.5f%s\" % (parentdist - node.dist, newick)\n",
    "        else:\n",
    "            newick = \");\"\n",
    "        newick = getNewick(node.get_left(), newick, node.dist, leaf_names)\n",
    "        newick = getNewick(node.get_right(), \",\" + newick, node.dist, leaf_names)\n",
    "        #newick = getNewick(node.get_right(), \",%s\" % (newick), node.dist, leaf_names)\n",
    "        \n",
    "        newick = \"(\" + newick\n",
    "        #newick = \"(%s\" % (newick)\n",
    "        return(newick)\n",
    "\n",
    "\n",
    "# In[11]:\n",
    "\n",
    "\n",
    "def alignment(fasta, threads, outfasta, outmsa, treeout):    \n",
    "\n",
    "    fasta.to_csv(outfasta, header=None, index=True, sep='\\n', mode='w')\n",
    "\n",
    "    mafft_cline = MafftCommandline(input=outfasta, thread=threads,  treeout = treeout)\n",
    "    #mafft_cline = MafftCommandline(input=outfasta, thread=threads,  treeout = False)\n",
    "    stdout, stderr = mafft_cline()\n",
    "\n",
    "    #os.remove('tmp.fasta') \n",
    "    with open(outmsa, \"w\") as handle:\n",
    "        handle.write(stdout)\n",
    "    \n",
    "    return(stdout)\n",
    "\n",
    "\n",
    "# In[12]:\n",
    "\n",
    "\n",
    "def fill_gaps(group, protein):\n",
    "    \n",
    "    set_protein = set(group[protein].tolist())\n",
    "    if '' in set_protein:\n",
    "        set_protein.discard('')\n",
    "    if len(set_protein) == 1:\n",
    "        group['protein'] = list(set_protein)[0]\n",
    "    else:\n",
    "        group['protein'] = group[protein]\n",
    "        group['protein'].replace('', 'mixed', inplace = True)\n",
    "        group['protein'].fillna('mixed', inplace = True)\n",
    "        \n",
    "    return(group)\n",
    "\n",
    "\n",
    "# In[13]:\n",
    "\n",
    "\n",
    "def get_acc(dictionary_tree, list_value):\n",
    "    list_accessions = []\n",
    "    for key, value in dictionary_tree.items():\n",
    "         if value in list_value:\n",
    "            list_accessions.append(key)\n",
    "    return(list_accessions)\n",
    "\n",
    "\n",
    "# In[14]:\n",
    "\n",
    "\n",
    "def layout(node):\n",
    "    \n",
    "    if node.is_leaf():\n",
    "        N = AttrFace(\"name\", fsize=10)\n",
    "        faces.add_face_to_node(N, node, 0, position=\"aligned\")\n",
    "    \n",
    "        M = AttrFace(\"size\", fsize=10, text_prefix='|Size ', text_suffix='|')\n",
    "        faces.add_face_to_node(M, node, 1, position=\"aligned\")    \n",
    "\n",
    "\n",
    "# In[15]:\n",
    "\n",
    "\n",
    "def get_tree(cluster, upload, segment, prot, list_color_hex, list_prune = [], list_focus = [], collapse = True, path_phylo = '', msa_dict = co.defaultdict(str), area = [], linkage = pd.DataFrame()):\n",
    "\n",
    "    if prot:\n",
    "    #Dataframes Creation and Correction\n",
    "        dataframe_tree_query = cluster.query('segment == @segment').join(upload.set_index('accession')[['host', 'date']])[['cluster', prot, 'host', 'date']]\n",
    "        dataframe_tree_corr = dataframe_tree_query.groupby('cluster').apply(fill_gaps, protein = prot).drop(prot, axis = 1)\n",
    "        dictionary_tree = dataframe_tree_corr.to_dict()\n",
    "        list_tree_label = dataframe_tree_corr.index.values\n",
    "    else:\n",
    "        dataframe_tree_query = cluster.query('segment == @segment').join(upload.set_index('accession')[['host', 'date']])[['cluster', 'host', 'date']]\n",
    "        dataframe_tree_query['protein'] = 'mixed'\n",
    "        dictionary_tree = dataframe_tree_query.to_dict()\n",
    "        list_tree_label = dataframe_tree_query.index.values\n",
    "        \n",
    "    #Tree Backbone Creation:\n",
    "    tree_style = TreeStyle()\n",
    "    \n",
    "    if not path_phylo and not linkage.empty:\n",
    "        array_tree_link = linkage.query('segment == @segment').drop(columns=['segment', 'parent']).to_numpy()\n",
    "        tree_hierarchy = hierarchy.to_tree(array_tree_link, False)\n",
    "        tree_newick = getNewick(tree_hierarchy, \"\", tree_hierarchy.dist, list_tree_label)\n",
    "        tree = Tree(tree_newick)\n",
    "    \n",
    "    elif path_phylo and linkage.empty:    \n",
    "        tree = Tree(path_phylo)\n",
    "        for leaf in tree.iter_leaves():\n",
    "            leaf_name = leaf.name.split('_')[1]\n",
    "            if not leaf_name == 'Cluster':\n",
    "                leaf.name = '>' + leaf_name\n",
    "            else:\n",
    "                leaf.name = leaf.name.split('_')[2]\n",
    "                \n",
    "    else:\n",
    "        print('Linkage Tree and Phylo Tree both or none given. Empty Tree generated.')\n",
    "        tree = Tree()\n",
    "        return(tree, tree_style)\n",
    "        \n",
    "    #General Appereance Settings:\n",
    "    for node in tree.traverse():\n",
    "        node.img_style[\"vt_line_width\"] = 1\n",
    "        node.img_style[\"hz_line_width\"] = 1\n",
    "        node.img_style[\"vt_line_type\"] = 0\n",
    "        node.img_style[\"hz_line_type\"] = 0\n",
    "        node.img_style[\"fgcolor\"] = \"#000000\"\n",
    "        node.img_style[\"shape\"] = \"circle\"\n",
    "        node.img_style[\"vt_line_color\"] = \"#000000\"\n",
    "        node.img_style[\"hz_line_color\"] = \"#000000\"\n",
    "        node.img_style['bgcolor'] = 'lightgrey'\n",
    "        node.protein = 'mixed'\n",
    "\n",
    "    #List and Dictionary Creation\n",
    "    #dict_color_protein = co.defaultdict(str)\n",
    "    list_cluster = list(set(dictionary_tree['cluster'].values()))\n",
    "    list_protein = list(set(dictionary_tree['protein'].values()))\n",
    "    list_host = list(set(dictionary_tree['host'].values()))\n",
    "    list_color = []\n",
    "    list_cluster_existing = []\n",
    "    \n",
    "    #Initial Leaf Settings:\n",
    "    for leaf in tree.iter_leaves():\n",
    "        if leaf.name.startswith('>'):\n",
    "            leaf.host = dictionary_tree['host'].get(leaf.name, \"none\")\n",
    "            leaf.date = dictionary_tree['date'].get(leaf.name, \"none\")\n",
    "            leaf.cluster = dictionary_tree['cluster'].get(leaf.name, \"none\")\n",
    "            leaf.protein = dictionary_tree['protein'].get(leaf.name, \"mixed\")\n",
    "            leaf.name = leaf.name[1:]\n",
    "            leaf.accession = True\n",
    "        else:\n",
    "            leaf_member = get_acc(dictionary_tree['cluster'], [int(leaf.name)])\n",
    "            leaf.protein = dictionary_tree['protein'].get(leaf_member[0], \"mixed\")\n",
    "            leaf.size = len(leaf_member)\n",
    "            leaf.accession = False\n",
    "    \n",
    "    #Cluster Node Creation and Settings:\n",
    "    if -1 in list_cluster:\n",
    "        list_cluster.remove(-1)\n",
    "    \n",
    "    for cluster in list_cluster:\n",
    "        try:\n",
    "            list_member = tree.search_nodes(cluster=cluster)\n",
    "            node_cluster = tree.get_common_ancestor(list_member)\n",
    "            node_cluster.name = str(cluster)\n",
    "            node_cluster.accession = False\n",
    "            node_cluster.size = len(list_member)\n",
    "            list_cluster_existing.append(cluster)\n",
    "        except:\n",
    "            pass\n",
    "            \n",
    "    #Serotype Coloring (maybe include mixed in the coloring):\n",
    "    if 'mixed' in list_protein:\n",
    "        list_protein.remove('mixed')\n",
    "\n",
    "    if prot:\n",
    "\n",
    "        for protein in list_protein:\n",
    "\n",
    "            for tree_protein in tree.get_monophyletic(values=[protein], target_attr = \"protein\"):\n",
    "\n",
    "                color = list_color_hex[int(protein[1:])-1]\n",
    "                #dict_color_protein[color] = protein\n",
    "                #tree_protein.protein = protein\n",
    "                #tree_protein.img_style['bgcolor'] = color\n",
    "\n",
    "                for node_protein in tree_protein.traverse():\n",
    "\n",
    "                    node_protein.img_style['bgcolor'] = color\n",
    "                    node_protein.protein = protein\n",
    "\n",
    "    #Host Monophyletic Lines Drawing:\n",
    "    for host in list_host:\n",
    "        for tree_host in tree.get_monophyletic(values=[host], target_attr = \"host\"):\n",
    "            \n",
    "            #tree_host.img_style[\"vt_line_type\"] = 1\n",
    "            #tree_host.img_style[\"hz_line_type\"] = 1\n",
    "            #tree_host.host = host\n",
    "            \n",
    "            for node_host in tree_host.traverse():\n",
    "                node_host.img_style[\"vt_line_type\"] = 1\n",
    "                node_host.img_style[\"hz_line_type\"] = 1\n",
    "                node_host.host = host\n",
    "\n",
    "    #Pruning according to Settings:\n",
    "    if not list_prune:\n",
    "        if not list_focus:\n",
    "            if collapse == True:\n",
    "                list_prune = list(map(str, list_cluster_existing))\n",
    "        else:\n",
    "            list_prune = [accession[1:] for accession in get_acc(dictionary_tree['cluster'], list_focus)]\n",
    "            if collapse == False:\n",
    "                list_prune.extend(list(map(str, list(set(list_cluster_existing).difference(set(list_focus))))))\n",
    "    \n",
    "    if list_prune:\n",
    "        try:\n",
    "            tree.prune(list_prune)\n",
    "        except:\n",
    "            print('No pruning performed.')\n",
    "\n",
    "    for leaf in tree:\n",
    "        if leaf.accession == True:\n",
    "            leaf.add_face(AttrFace(\"cluster\", fsize=10, text_prefix='Cluster: '), 0, position=\"aligned\")\n",
    "            leaf.add_face(AttrFace(\"name\", fsize=10, text_prefix='|', text_suffix='|'), 1, position=\"aligned\")\n",
    "            leaf.add_face(AttrFace(\"date\", fsize=10, text_prefix='Date: '), 2, position=\"aligned\")\n",
    "            if msa_dict:\n",
    "                seq = msa_dict.get(leaf.name, \"\")\n",
    "                if area:\n",
    "                    area_seq = \"...\".join([seq[start:end] for start, end in area])\n",
    "                    leaf.add_face(SequenceFace(area_seq), 3, position=\"aligned\")\n",
    "                else:\n",
    "                    leaf.add_face(SequenceFace(seq), 3, position=\"aligned\")\n",
    "        else:\n",
    "            leaf.add_face(AttrFace(\"name\", fsize=10, text_prefix='Cluster: '), 0, position=\"aligned\")\n",
    "            leaf.add_face(AttrFace(\"size\", fsize=10, text_prefix='|Size: ', text_suffix='|'), 1, position=\"aligned\")\n",
    "            if msa_dict:\n",
    "                seq = msa_dict.get(leaf.name, \"\")\n",
    "                if area:\n",
    "                    area_seq = \"...\".join([seq[start:end] for start, end in area])\n",
    "                    leaf.add_face(SequenceFace(area_seq), 2, position=\"aligned\")\n",
    "                else:\n",
    "                    leaf.add_face(SequenceFace(seq), 2, position=\"aligned\")\n",
    "        \n",
    "        list_color.append(leaf.img_style['bgcolor'])\n",
    "\n",
    "    if prot:\n",
    "        set_color = set(list_color)\n",
    "        column = 0\n",
    "        for color_hex in list_color_hex:\n",
    "            if color_hex in set_color:\n",
    "                face_protein = TextFace(prot + str(list_color_hex.index(color_hex)+1))\n",
    "                face_protein.margin_top = 10\n",
    "                face_protein.margin_right = 10\n",
    "                face_protein.margin_left = 10\n",
    "                face_protein.margin_bottom = 10\n",
    "                face_protein.border.width = 1\n",
    "                face_protein.background.color = color_hex\n",
    "                tree_style.legend.add_face(face_protein, column=column)\n",
    "                column += 1\n",
    "        tree_style.legend_position = 1\n",
    "            \n",
    "    tree_style.show_leaf_name = False\n",
    "    tree_style.draw_guiding_lines = True\n",
    "    tree_style.guiding_lines_color = 'black'\n",
    "    \n",
    "    if not msa_dict:\n",
    "        tree_style.mode = \"c\"\n",
    "\n",
    "    return(tree, tree_style, tree_newick)\n",
    "\n",
    "\n",
    "def pairmsa(nuc1, nuc2):\n",
    "        \n",
    "    seq1 = Seq(nuc1.item())\n",
    "    seq2 = Seq(nuc2.item()) \n",
    "\n",
    "    aligner = Align.PairwiseAligner()\n",
    "    alignments = aligner.align(seq1, seq2) \n",
    "    \n",
    "    #alignments = pairwise2.align.globalmx(seq1, seq2, 1, -1) \n",
    "    \n",
    "    return(alignments[0].score)\n",
    "\n",
    "def sample_difference(cluster, genome, segment, tree, threshold = -1, n = -1, index = []):\n",
    "    \n",
    "    if not index:\n",
    "        num = cluster['cluster'].max()+1\n",
    "        index = [i for i in range(num)]\n",
    "    else:\n",
    "        num = len(index)\n",
    "    \n",
    "    pair = np.empty([num, num])\n",
    "    \n",
    "    for x, j in enumerate(index):\n",
    "        for y, k in enumerate(index):\n",
    "            \n",
    "            if j != k:\n",
    "                dist = tree.get_distance(str(j), str(k))\n",
    "            else:\n",
    "                dist = 0.0\n",
    "            \n",
    "            if threshold == -1 or dist <= threshold:\n",
    "            \n",
    "                if n == -1:\n",
    "                    sample_x = cluster.query('segment == @segment & cluster == @j & centroid == True').join(genome)[['genome']]\n",
    "                    sample_y = cluster.query('segment == @segment & cluster == @k & centroid == True').join(genome)[['genome']]\n",
    "\n",
    "                else:\n",
    "                    query_x = cluster.query('segment == @segment & cluster == @j').join(genome)[['genome']]\n",
    "                    if len(query_x) >= n:\n",
    "                        sample_x = query_x.sample(n=n, random_state=42)\n",
    "                    else:\n",
    "                        sample_x = query_x\n",
    "\n",
    "                    query_y = cluster.query('segment == @segment & cluster == @k').join(genome)[['genome']]\n",
    "                    if len(query_y) >= n:\n",
    "                        sample_y = query_y.sample(n=n, random_state=42)\n",
    "                    else:\n",
    "                        sample_y = query_y\n",
    "\n",
    "                dist = ssd.cdist(sample_x, sample_y, metric = pairmsa)\n",
    "\n",
    "                dist_mean = pd.DataFrame(dist, dtype = 'float64').mean().mean()\n",
    "                len_mean = (sample_x['genome'].str.len().mean() + sample_y['genome'].str.len().mean())/2\n",
    "                \n",
    "                pair[x,y] = dist_mean/len_mean\n",
    "                \n",
    "            else:\n",
    "                pair[x,y] = 0.0\n",
    "            \n",
    "    dataframe_pair = pd.DataFrame(pair, index = index, columns = index)\n",
    "\n",
    "    return(dataframe_pair)\n",
    "\n",
    "\n",
    "def plot_difference(proof, segments, outpath, render):\n",
    "\n",
    "    with sns.axes_style(\"darkgrid\"):\n",
    "        fig, ax = plt.subplots(figsize=(8,6))\n",
    "        ax = sns.heatmap(proof, linewidths=.5)\n",
    "        plt.xlabel(\"Cluster\")\n",
    "        plt.ylabel(\"Cluster\")\n",
    "        plt.tight_layout()\n",
    "        plt.close(fig)\n",
    "        fig.savefig(outpath + 'Cluster_Difference_Segment_' + str(segments) + '.' + render)\n",
    "\n",
    "# In[23]:\n",
    "\n",
    "\n",
    "def main(infile, outfolder, segments, custom, metric, min_clust, sample, umap_neigh, umap_comp, pca_comp, kneedle, render, epsilon):#, recreate):\n",
    "\n",
    "    Hs = ['H1', 'H2', 'H3', 'H4', 'H5', 'H6', 'H7', 'H8', 'H9', 'H10', 'H11', 'H12', 'H13', 'H14', 'H15', 'H16', 'H17', 'H18']\n",
    "    Ns = ['N1', 'N2', 'N3', 'N4', 'N5', 'N6', 'N7', 'N8', 'N9', 'N10', 'N11']\n",
    "    tree_hex = ['#FF0029','#377EB8','#66A61E','#984EA3','#00D2D5','#FF7F00','#AF8D00','#7F80CD','#B3E900','#C42E60','#A65628','#F781BF','#8DD3C7','#BEBADA','#FB8072','#80B1D3','#FDB462','#FCCDE5']\n",
    "\n",
    "    try:\n",
    "        if not outfolder.endswith(\"/\"):\n",
    "            outfolder += \"/\"\n",
    "\n",
    "        if not os.path.isdir(outfolder):\n",
    "            os.mkdir(outfolder)\n",
    "\n",
    "        #if recreate == True:\n",
    "        for f in os.listdir(outfolder):\n",
    "            os.remove(os.path.join(outfolder, f))\n",
    "    except:\n",
    "        sys.exit('Outfolder not reachable. Exit.')\n",
    "\n",
    "    #Dict Creation\n",
    "    dataframe_pca_dict = co.defaultdict(list)\n",
    "    dataframe_dict = co.defaultdict(list)\n",
    "    cluster_dict = co.defaultdict(list)\n",
    "    information_dict = co.defaultdict(list)\n",
    "    linkage_dict = co.defaultdict(list)\n",
    "    elbow_dict = co.defaultdict(list)\n",
    "\n",
    "    #Upload Fasta\n",
    "    try:\n",
    "        upload = pd.read_csv('fasta.csv', sep = ',', na_filter = False, header = None)\n",
    "\n",
    "    except:\n",
    "        try:\n",
    "            #Convert the Fasta to Csv\n",
    "            convert_fasta(infile, 'fasta.csv')\n",
    "            upload = pd.read_csv('fasta.csv', sep = ',', na_filter = False, header = None)\n",
    "        except:\n",
    "            sys.exit('Infile not reachable or wrong format. Exit.')\n",
    "\n",
    "    upload.columns = custom\n",
    "    upload.query('curation == \"Pass\"', inplace = True)\n",
    "    upload[['H', 'N']] = upload.subtype.apply(lambda x: subsplit(x)).tolist()\n",
    "    #upload.to_csv(outfolder + 'upload.csv', index=False, header=True, sep=',', mode='w')\n",
    "\n",
    "    genome = upload[['accession', 'genome']].copy()\n",
    "    genome.set_index('accession', inplace = True)\n",
    "\n",
    "    try:\n",
    "\n",
    "        #dataframe_pca = pd.read_csv(outfolder + 'dataframe_pca.csv', sep = ',', na_filter = False, header = 0, index_col = 0)\n",
    "        #dataframe = pd.read_csv(outfolder + 'dataframe.csv', sep = ',', na_filter = False, header = 0, index_col = 0)\n",
    "        cluster = pd.read_csv(outfolder + 'cluster.csv', sep = ',', na_filter = False, header = 0, index_col = 0)\n",
    "        linkage = pd.read_csv(outfolder + 'linkage.csv', sep = ',', na_filter = False, header = 0)\n",
    "        information = pd.read_csv(outfolder + 'information.csv', sep = ',', na_filter = False, header = 0, index_col = 0)\n",
    "        density = pd.read_csv(outfolder + 'density.csv', sep = ',', na_filter = False, header = 0)\n",
    "        elbow = pd.read_csv(outfolder + 'elbow.csv', sep = ',', na_filter = False, header = 0)\n",
    "\n",
    "    except:\n",
    "\n",
    "        #Progress Variables\n",
    "        widgets = [' [', progressbar.Timer(format = 'elapsed time: %(elapsed)s'), '] ', progressbar.Bar('#'),' (', progressbar.ETA(), ') ', ] \n",
    "        bar = progressbar.ProgressBar(max_value = len(segments), widgets = widgets).start() \n",
    "\n",
    "        #Pipeline\n",
    "        for progress, segment in enumerate(segments, 1):\n",
    "\n",
    "            subset = upload.query('segment == @segment').reset_index()\n",
    "            sequence = subset[['genome']].copy()\n",
    "            accession = subset[['accession']].copy()\n",
    "            extra = subset[['H', 'N']].copy()\n",
    "\n",
    "            dataframe_dict[segment], dataframe_pca_dict[segment], explained_var = get_vectors(sequence, accession, metric, umap_neigh, umap_comp, pca_comp)\n",
    "            linkage_dict[segment], elbow_dict[segment], epsilon_best, n_cluster_raw, n_cluster_norm = get_elbow(dataframe_dict[segment], extra, accession, metric, min_clust, sample, kneedle)\n",
    "\n",
    "            cluster_dict[segment], n_cluster, unclustered, H_unmatch, N_unmatch = get_cluster(epsilon_best, dataframe_dict[segment], extra, accession, metric, min_clust, sample)\n",
    "            information_dict[segment] = [n_cluster, n_cluster_raw, n_cluster_norm, unclustered, H_unmatch, N_unmatch, epsilon_best, explained_var]\n",
    "\n",
    "            bar.update(progress)\n",
    "\n",
    "        bar.finish()    \n",
    "\n",
    "        ########################################################################################################\n",
    "\n",
    "        #dataframe_pca = pd.concat(dataframe_pca_dict)\n",
    "        #dataframe_pca.index.set_names([\"segment\", \"accession\"], inplace=True)\n",
    "        #dataframe_pca.reset_index(level = \"segment\", inplace=True, drop = True)\n",
    "\n",
    "        #dataframe = pd.concat(dataframe_dict)\n",
    "        #dataframe.index.set_names([\"segment\", \"accession\"], inplace=True)\n",
    "        #dataframe.reset_index(level = \"segment\", inplace=True, drop = True)\n",
    "\n",
    "        cluster = pd.concat(cluster_dict)\n",
    "        cluster.index.set_names([\"segment\", \"accession\"], inplace=True)\n",
    "        cluster.reset_index(level = \"segment\", inplace=True)\n",
    "\n",
    "        linkage = pd.concat(linkage_dict)\n",
    "        linkage.index.set_names([\"segment\", 'parent'], inplace=True)\n",
    "        linkage.reset_index(level = [\"segment\", 'parent'], inplace=True)\n",
    "\n",
    "        information = pd.DataFrame.from_dict(information_dict, orient='index', columns=['n_cluster', 'n_cluster_raw', 'n_cluster_norm', 'unclustered', 'H_unmatch', 'N_unmatch', 'epsilon', 'explained_var'])\n",
    "        information.index.set_names([\"segment\"], inplace=True)\n",
    "\n",
    "        elbow = pd.concat(elbow_dict)\n",
    "        elbow.index.set_names(['segment', 'n_cluster'], inplace=True)\n",
    "        elbow.reset_index(level = [\"segment\", \"n_cluster\"], inplace=True)\n",
    "\n",
    "        density = pd.DataFrame(cluster.value_counts(subset=['segment', 'cluster']), columns = ['size'])\n",
    "        density.reset_index(level = [\"segment\", \"cluster\"], inplace=True)\n",
    "\n",
    "        ########################################################################################################\n",
    "\n",
    "        #correcting the KNOWN errors in the database\n",
    "        if '>MK237334' in cluster.index:\n",
    "            cluster.loc['>MK237334','H'] = \"H1\"\n",
    "\n",
    "        if '>MH346657' in cluster.index:\n",
    "            cluster.loc['>MH346657','N'] = \"N1\"\n",
    "\n",
    "        if '>MK237837' in cluster.index:\n",
    "            cluster.loc['>MK237837','N'] = \"N1\"\n",
    "\n",
    "        if '>MG925516' in cluster.index:\n",
    "            cluster.loc['>MG925516','N'] = \"N2\"\n",
    "\n",
    "        ########################################################################################################\n",
    "\n",
    "        #dataframe_pca.to_csv(outfolder + 'dataframe_pca.csv', index=True, header=True, sep=',', mode='w')\n",
    "        #dataframe.to_csv(outfolder + 'dataframe.csv', index=True, header=True, sep=',', mode='w')\n",
    "\n",
    "        cluster.to_csv(outfolder + 'cluster.csv', index=True, header=True, sep=',', mode='w')\n",
    "        #genome.to_csv(outfolder + 'genome.csv', index=True, header=True, sep=',', mode='w')\n",
    "        information.to_csv(outfolder + 'information.csv', index=True, header=True, sep=',', mode='w')\n",
    "\n",
    "        elbow.to_csv(outfolder + 'elbow.csv', index=False, header=True, sep=',', mode='w')\n",
    "        density.to_csv(outfolder + 'density.csv', index=False, header=True, sep=',', mode='w')\n",
    "        linkage.to_csv(outfolder + 'linkage.csv', index=False, header=True, sep=',', mode='w')\n",
    "\n",
    "        ########################################################################################################    \n",
    "\n",
    "    plot_elbow(elbow, information, segments, outfolder, render)\n",
    "    plot_density(density, segments, outfolder, render, 1)\n",
    "\n",
    "    anti_dict = {1:None, 2:None, 3:None, 4:'H', 5:None, 6:'N', 7:None, 8:None}\n",
    "\n",
    "    for seg in segments:\n",
    "\n",
    "        tree, ts, newick = get_tree(cluster, upload, seg, anti_dict[seg], tree_hex, linkage = linkage)\n",
    "        tree.render(outfolder + 'Clustertree_Segment_' + str(seg) + '.' + render, tree_style = ts)\n",
    "        with open(outfolder + 'Newick_Segment_' + str(seg) + '.txt', 'w') as f:\n",
    "            f.write(newick)\n",
    "        proof = sample_difference(cluster, genome, seg, tree)\n",
    "        proof.to_csv(outfolder + 'Pairwise_Segment_' + str(seg) + '.csv', index=True, header=True, sep=',', mode='w')\n",
    "        plot_difference(proof, 4, outfolder, render)\n",
    "\n",
    "    os.remove(\"fasta.csv\") \n",
    "\t\n",
    "# In[29]:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "introductory-hollow",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_difference(proof, segments, outpath, render):\n",
    "\n",
    "    with sns.axes_style(\"darkgrid\"):\n",
    "        fig, ax = plt.subplots(figsize=(8,6))\n",
    "        ax = sns.heatmap(proof, linewidths=.5)\n",
    "        plt.xlabel(\"Cluster\")\n",
    "        plt.ylabel(\"Cluster\")\n",
    "        plt.tight_layout()\n",
    "        plt.close(fig)\n",
    "        fig.savefig(outpath + 'Cluster_Difference_Segment_' + str(segments) + '.' + render)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "elder-michigan",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pairmsa(nuc1, nuc2):\n",
    "        \n",
    "    seq1 = Seq(nuc1.item())\n",
    "    seq2 = Seq(nuc2.item()) \n",
    "\n",
    "    aligner = Align.PairwiseAligner()\n",
    "    aligner.mode = 'global'\n",
    "    alignments = aligner.align(seq1, seq2) \n",
    "    \n",
    "    #alignments = pairwise2.align.globalmx(seq1, seq2, 1, -1) \n",
    "    \n",
    "    return(alignments[0].score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "official-happening",
   "metadata": {},
   "outputs": [],
   "source": [
    "outfolder = 'Results/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "residential-elevation",
   "metadata": {},
   "outputs": [],
   "source": [
    "convert_fasta('../../Desktop/A.fasta', '../../Desktop/fasta.csv')\n",
    "upload = pd.read_csv('../../Desktop/fasta.csv', sep = ',', na_filter = False, header = None)\n",
    "\n",
    "upload.columns = ['accession', 'strain', 'segment', 'protein', 'genus', 'subtype', 'date', 'host', 'curation', 'genome']\n",
    "upload.query('curation == \"Pass\"', inplace = True)\n",
    "upload[['H', 'N']] = upload.subtype.apply(lambda x: subsplit(x)).tolist()\n",
    "\n",
    "genome = upload[['accession', 'genome']].copy()\n",
    "genome.set_index('accession', inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "welcome-tissue",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster = pd.read_csv(outfolder + 'cluster.csv', sep = ',', na_filter = False, header = 0, index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "potential-limitation",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import pairwise2\n",
    "from multiprocessing import Pool\n",
    "import functools\n",
    "\n",
    "def pairmsa2(nuc1, nuc2):\n",
    "        \n",
    "    seq1 = Seq(nuc1.item())\n",
    "    seq2 = Seq(nuc2.item()) \n",
    "\n",
    "    alignments = pairwise2.align.globalxx(seq1, seq2, score_only = True) \n",
    "    \n",
    "    return(alignments)\n",
    "\n",
    "def worker(x, y, j, k, n, cl):\n",
    "           \n",
    "    if n == -1:\n",
    "        sample_x = cl.query('cluster == @j & centroid == True').join(genome)[['genome']]\n",
    "        sample_y = cl.query('cluster == @k & centroid == True').join(genome)[['genome']]\n",
    "\n",
    "    else:\n",
    "        query_x = cl.query('cluster == @j').join(genome)[['genome']]\n",
    "        if len(query_x) >= n:\n",
    "            sample_x = query_x.sample(n=n, random_state=42)\n",
    "        else:\n",
    "            sample_x = query_x\n",
    "\n",
    "        query_y = cl.query('cluster == @k').join(genome)[['genome']]\n",
    "        if len(query_y) >= n:\n",
    "            sample_y = query_y.sample(n=n, random_state=42)\n",
    "        else:\n",
    "            sample_y = query_y\n",
    "\n",
    "    dist_mean = ssd.cdist(sample_x, sample_y, metric = pairmsa2).mean()\n",
    "    len_mean = (sample_x['genome'].str.len().mean() + sample_y['genome'].str.len().mean())/2\n",
    "\n",
    "    return((x,y,dist_mean/len_mean))\n",
    "\n",
    "def sample_difference(cluster, genome, segment, proc = 8, n = -1):\n",
    "\n",
    "    cl = cluster.query('segment == @segment') \n",
    "    names = list(set(cl['cluster']).difference({-1}))\n",
    "    num = len(names)\n",
    "        \n",
    "    params = []\n",
    "    dist_matrix = np.zeros([num, num])\n",
    "\n",
    "    for x, j in enumerate(names):\n",
    "        for y, k in enumerate(names[:names.index(j)+1]):\n",
    "            params.append((x, y, j, k, n, cl))\n",
    "\n",
    "    with Pool(processes=proc) as pool:\n",
    "        result = pool.starmap(worker, params)\n",
    "\n",
    "    for x, y, dist in result:\n",
    "        dist_matrix[x,y] = dist\n",
    "    \n",
    "    dist_dataframe = pd.DataFrame(dist_matrix, index = names, columns = names)\n",
    "    \n",
    "    return(dist_dataframe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "steady-programmer",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_x = cluster.query('segment == 4 & cluster == 9').sample(n=3, random_state=42).join(genome)[['genome']]\n",
    "sample_y = cluster.query('segment == 4 & cluster == 12').sample(n=20, random_state=42).join(genome)[['genome']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "listed-doubt",
   "metadata": {},
   "outputs": [],
   "source": [
    "dist = ssd.cdist(sample_x, sample_y, metric = pairmsa2)\n",
    "dist_mean = dist.mean()\n",
    "len_mean = (sample_x['genome'].str.len().mean() + sample_y['genome'].str.len().mean())/2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "personalized-regular",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1729.8695652173913"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "grave-necklace",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.815090</td>\n",
       "      <td>0.805841</td>\n",
       "      <td>0.810466</td>\n",
       "      <td>0.815090</td>\n",
       "      <td>0.809888</td>\n",
       "      <td>0.812778</td>\n",
       "      <td>0.806997</td>\n",
       "      <td>0.812200</td>\n",
       "      <td>0.812778</td>\n",
       "      <td>0.814512</td>\n",
       "      <td>0.810466</td>\n",
       "      <td>0.805841</td>\n",
       "      <td>0.809888</td>\n",
       "      <td>0.811622</td>\n",
       "      <td>0.809888</td>\n",
       "      <td>0.811044</td>\n",
       "      <td>0.813934</td>\n",
       "      <td>0.814512</td>\n",
       "      <td>0.809888</td>\n",
       "      <td>0.809888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.803529</td>\n",
       "      <td>0.800060</td>\n",
       "      <td>0.800638</td>\n",
       "      <td>0.801216</td>\n",
       "      <td>0.797170</td>\n",
       "      <td>0.799482</td>\n",
       "      <td>0.793701</td>\n",
       "      <td>0.802373</td>\n",
       "      <td>0.801216</td>\n",
       "      <td>0.802373</td>\n",
       "      <td>0.796592</td>\n",
       "      <td>0.790233</td>\n",
       "      <td>0.797170</td>\n",
       "      <td>0.802951</td>\n",
       "      <td>0.794858</td>\n",
       "      <td>0.797748</td>\n",
       "      <td>0.802951</td>\n",
       "      <td>0.802951</td>\n",
       "      <td>0.796592</td>\n",
       "      <td>0.796014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.807575</td>\n",
       "      <td>0.802951</td>\n",
       "      <td>0.803529</td>\n",
       "      <td>0.806419</td>\n",
       "      <td>0.802951</td>\n",
       "      <td>0.804685</td>\n",
       "      <td>0.794858</td>\n",
       "      <td>0.805263</td>\n",
       "      <td>0.806419</td>\n",
       "      <td>0.806997</td>\n",
       "      <td>0.798326</td>\n",
       "      <td>0.794858</td>\n",
       "      <td>0.800638</td>\n",
       "      <td>0.806419</td>\n",
       "      <td>0.796592</td>\n",
       "      <td>0.800638</td>\n",
       "      <td>0.806419</td>\n",
       "      <td>0.808153</td>\n",
       "      <td>0.796592</td>\n",
       "      <td>0.798904</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         0         1         2         3         4         5         6   \\\n",
       "0  0.815090  0.805841  0.810466  0.815090  0.809888  0.812778  0.806997   \n",
       "1  0.803529  0.800060  0.800638  0.801216  0.797170  0.799482  0.793701   \n",
       "2  0.807575  0.802951  0.803529  0.806419  0.802951  0.804685  0.794858   \n",
       "\n",
       "         7         8         9         10        11        12        13  \\\n",
       "0  0.812200  0.812778  0.814512  0.810466  0.805841  0.809888  0.811622   \n",
       "1  0.802373  0.801216  0.802373  0.796592  0.790233  0.797170  0.802951   \n",
       "2  0.805263  0.806419  0.806997  0.798326  0.794858  0.800638  0.806419   \n",
       "\n",
       "         14        15        16        17        18        19  \n",
       "0  0.809888  0.811044  0.813934  0.814512  0.809888  0.809888  \n",
       "1  0.794858  0.797748  0.802951  0.802951  0.796592  0.796014  \n",
       "2  0.796592  0.800638  0.806419  0.808153  0.796592  0.798904  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(dist)/sample_x.append(sample_y)['genome'].str.len().mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "expired-diesel",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.820710</td>\n",
       "      <td>0.811397</td>\n",
       "      <td>0.816053</td>\n",
       "      <td>0.820710</td>\n",
       "      <td>0.815471</td>\n",
       "      <td>0.818382</td>\n",
       "      <td>0.812561</td>\n",
       "      <td>0.817800</td>\n",
       "      <td>0.818382</td>\n",
       "      <td>0.820128</td>\n",
       "      <td>0.816053</td>\n",
       "      <td>0.811397</td>\n",
       "      <td>0.815471</td>\n",
       "      <td>0.817217</td>\n",
       "      <td>0.815471</td>\n",
       "      <td>0.816635</td>\n",
       "      <td>0.819546</td>\n",
       "      <td>0.820128</td>\n",
       "      <td>0.815471</td>\n",
       "      <td>0.815471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.809069</td>\n",
       "      <td>0.805576</td>\n",
       "      <td>0.806158</td>\n",
       "      <td>0.806740</td>\n",
       "      <td>0.802666</td>\n",
       "      <td>0.804994</td>\n",
       "      <td>0.799173</td>\n",
       "      <td>0.807904</td>\n",
       "      <td>0.806740</td>\n",
       "      <td>0.807904</td>\n",
       "      <td>0.802084</td>\n",
       "      <td>0.795681</td>\n",
       "      <td>0.802666</td>\n",
       "      <td>0.808486</td>\n",
       "      <td>0.800338</td>\n",
       "      <td>0.803248</td>\n",
       "      <td>0.808486</td>\n",
       "      <td>0.808486</td>\n",
       "      <td>0.802084</td>\n",
       "      <td>0.801502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.813143</td>\n",
       "      <td>0.808486</td>\n",
       "      <td>0.809069</td>\n",
       "      <td>0.811979</td>\n",
       "      <td>0.808486</td>\n",
       "      <td>0.810233</td>\n",
       "      <td>0.800338</td>\n",
       "      <td>0.810815</td>\n",
       "      <td>0.811979</td>\n",
       "      <td>0.812561</td>\n",
       "      <td>0.803830</td>\n",
       "      <td>0.800338</td>\n",
       "      <td>0.806158</td>\n",
       "      <td>0.811979</td>\n",
       "      <td>0.802084</td>\n",
       "      <td>0.806158</td>\n",
       "      <td>0.811979</td>\n",
       "      <td>0.813725</td>\n",
       "      <td>0.802084</td>\n",
       "      <td>0.804412</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         0         1         2         3         4         5         6   \\\n",
       "0  0.820710  0.811397  0.816053  0.820710  0.815471  0.818382  0.812561   \n",
       "1  0.809069  0.805576  0.806158  0.806740  0.802666  0.804994  0.799173   \n",
       "2  0.813143  0.808486  0.809069  0.811979  0.808486  0.810233  0.800338   \n",
       "\n",
       "         7         8         9         10        11        12        13  \\\n",
       "0  0.817800  0.818382  0.820128  0.816053  0.811397  0.815471  0.817217   \n",
       "1  0.807904  0.806740  0.807904  0.802084  0.795681  0.802666  0.808486   \n",
       "2  0.810815  0.811979  0.812561  0.803830  0.800338  0.806158  0.811979   \n",
       "\n",
       "         14        15        16        17        18        19  \n",
       "0  0.815471  0.816635  0.819546  0.820128  0.815471  0.815471  \n",
       "1  0.800338  0.803248  0.808486  0.808486  0.802084  0.801502  \n",
       "2  0.802084  0.806158  0.811979  0.813725  0.802084  0.804412  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(dist)/len_mean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "environmental-brooklyn",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1734.05"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_y['genome'].str.len().mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "domestic-album",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1718.025"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len_mean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "grand-champagne",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "flexible-denial",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "painful-intellectual",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'DataFrame' object is not callable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-37-7ca20b0a757f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0msample_difference\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcluster\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgenome\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m4\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mproc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m8\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m: 'DataFrame' object is not callable"
     ]
    }
   ],
   "source": [
    "sample_difference(cluster, genome, 4, proc = 8, n=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "annual-conservation",
   "metadata": {},
   "outputs": [],
   "source": [
    "upload = pd.read_csv('../../Desktop/fasta.csv', sep = ',', na_filter = False, header = None)\n",
    "upload.columns = ['accession', 'strain', 'segment', 'protein', 'genus', 'subtype', 'date', 'host', 'curation', 'genome']\n",
    "upload.query('curation == \"Pass\"', inplace = True)\n",
    "\n",
    "genome = upload[['accession', 'genome']].copy()\n",
    "genome.set_index('accession', inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "funded-chinese",
   "metadata": {},
   "outputs": [],
   "source": [
    "upload.set_index('accession', inplace = True)\n",
    "upload.to_csv('../../Desktop/extra.csv', index=True, header=True, sep=',', mode='w')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "annoying-samba",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster = pd.read_csv('Results/cluster.csv', sep = ',', na_filter = False, header = 0, index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "devoted-princess",
   "metadata": {},
   "outputs": [],
   "source": [
    "investigation = cluster.join(upload, lsuffix='_caller', rsuffix='_other')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "saving-metropolitan",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "strain                        A/environment/Chile/C20076/2016\n",
       "segment                                                     4\n",
       "protein                                                    HA\n",
       "genus                                                       A\n",
       "subtype                                                 H13N9\n",
       "date                                               12/13/2016\n",
       "host                                              Environment\n",
       "curation                                                 Pass\n",
       "genome      ATGGAAGTTCCAGTACTCGCCCTCTTAATGTTAACCAGTATATGCG...\n",
       "Name: >MH488120, dtype: object"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#str1 = \n",
    "upload.loc['>MH488120']#['genome']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "authentic-english",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "strain                 A/white_backed_stilt/Chile/C20090/2016\n",
       "segment                                                     4\n",
       "protein                                                    HA\n",
       "genus                                                       A\n",
       "subtype                                                 H13N9\n",
       "date                                               12/13/2016\n",
       "host                                       Black_Winged_Stilt\n",
       "curation                                                 Pass\n",
       "genome      ATGGAAGTTCCAGTACTCGCCCTCTTAATGTTAACCAGTATATGCG...\n",
       "Name: >MH498671, dtype: object"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#str2 = \n",
    "upload.loc['>MH498671']#['genome']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "standard-comment",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str1 == str2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "available-equity",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>segment</th>\n",
       "      <th>protein</th>\n",
       "      <th>genus</th>\n",
       "      <th>subtype</th>\n",
       "      <th>date</th>\n",
       "      <th>host</th>\n",
       "      <th>curation</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>&gt;MH498772</th>\n",
       "      <td>A/environment/Chile/C20173/2016</td>\n",
       "      <td>7</td>\n",
       "      <td>M1</td>\n",
       "      <td>A</td>\n",
       "      <td>H13N9</td>\n",
       "      <td>12/13/2016</td>\n",
       "      <td>Environment</td>\n",
       "      <td>Pass</td>\n",
       "      <td>ATGAGTCTTCTAACCGAGGTCGAAACGTATGTTCTCTCTATCATTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;MH499024</th>\n",
       "      <td>A/environment/Chile/C20173/2016</td>\n",
       "      <td>8</td>\n",
       "      <td>NS1</td>\n",
       "      <td>A</td>\n",
       "      <td>H13N9</td>\n",
       "      <td>12/13/2016</td>\n",
       "      <td>Environment</td>\n",
       "      <td>Pass</td>\n",
       "      <td>ATGGATTCAAACACCGTGTCAAGCTTTCAGGTAGACTGCTTTCTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;MH499206</th>\n",
       "      <td>A/environment/Chile/C20173/2016</td>\n",
       "      <td>5</td>\n",
       "      <td>NP</td>\n",
       "      <td>A</td>\n",
       "      <td>H13N9</td>\n",
       "      <td>12/13/2016</td>\n",
       "      <td>Environment</td>\n",
       "      <td>Pass</td>\n",
       "      <td>ATGGCGTCTCAAGGCACCAAACGATCCTATGAGCAGATGGAAACTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;MH498744</th>\n",
       "      <td>A/environment/Chile/C20173/2016</td>\n",
       "      <td>3</td>\n",
       "      <td>PA</td>\n",
       "      <td>A</td>\n",
       "      <td>H13N9</td>\n",
       "      <td>12/13/2016</td>\n",
       "      <td>Environment</td>\n",
       "      <td>Pass</td>\n",
       "      <td>ATGGAAGACTTTGTGCGACAATGCTTCAATCCAATGGTTGTCGAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;MH498989</th>\n",
       "      <td>A/environment/Chile/C20173/2016</td>\n",
       "      <td>6</td>\n",
       "      <td>NA</td>\n",
       "      <td>A</td>\n",
       "      <td>H13N9</td>\n",
       "      <td>12/13/2016</td>\n",
       "      <td>Environment</td>\n",
       "      <td>Pass</td>\n",
       "      <td>ATGAATCCAAATCAGAAGATTCTATGCACTTCTGCCACCGCTATTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;MH499064</th>\n",
       "      <td>A/environment/Chile/C20173/2016</td>\n",
       "      <td>1</td>\n",
       "      <td>PB2</td>\n",
       "      <td>A</td>\n",
       "      <td>H13N9</td>\n",
       "      <td>12/13/2016</td>\n",
       "      <td>Environment</td>\n",
       "      <td>Pass</td>\n",
       "      <td>ATGGAGAGAATAAAAGAACTAAGAGATCTAATGTCACAGTCTCGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;MH499234</th>\n",
       "      <td>A/environment/Chile/C20173/2016</td>\n",
       "      <td>2</td>\n",
       "      <td>PB1</td>\n",
       "      <td>A</td>\n",
       "      <td>H13N9</td>\n",
       "      <td>12/13/2016</td>\n",
       "      <td>Environment</td>\n",
       "      <td>Pass</td>\n",
       "      <td>ATGGATGTCAATCCGACTTTACTTTTCTTGAAAGTGCCAGCGCAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;MH498694</th>\n",
       "      <td>A/environment/Chile/C20173/2016</td>\n",
       "      <td>4</td>\n",
       "      <td>HA</td>\n",
       "      <td>A</td>\n",
       "      <td>H13N9</td>\n",
       "      <td>12/13/2016</td>\n",
       "      <td>Environment</td>\n",
       "      <td>Pass</td>\n",
       "      <td>ATGGAAGTTCCAGTACTCGCCCTCTTAATGTTAACCAGTATATGCG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    strain  segment protein genus subtype  \\\n",
       "accession                                                                   \n",
       ">MH498772  A/environment/Chile/C20173/2016        7      M1     A   H13N9   \n",
       ">MH499024  A/environment/Chile/C20173/2016        8     NS1     A   H13N9   \n",
       ">MH499206  A/environment/Chile/C20173/2016        5      NP     A   H13N9   \n",
       ">MH498744  A/environment/Chile/C20173/2016        3      PA     A   H13N9   \n",
       ">MH498989  A/environment/Chile/C20173/2016        6      NA     A   H13N9   \n",
       ">MH499064  A/environment/Chile/C20173/2016        1     PB2     A   H13N9   \n",
       ">MH499234  A/environment/Chile/C20173/2016        2     PB1     A   H13N9   \n",
       ">MH498694  A/environment/Chile/C20173/2016        4      HA     A   H13N9   \n",
       "\n",
       "                 date         host curation  \\\n",
       "accession                                     \n",
       ">MH498772  12/13/2016  Environment     Pass   \n",
       ">MH499024  12/13/2016  Environment     Pass   \n",
       ">MH499206  12/13/2016  Environment     Pass   \n",
       ">MH498744  12/13/2016  Environment     Pass   \n",
       ">MH498989  12/13/2016  Environment     Pass   \n",
       ">MH499064  12/13/2016  Environment     Pass   \n",
       ">MH499234  12/13/2016  Environment     Pass   \n",
       ">MH498694  12/13/2016  Environment     Pass   \n",
       "\n",
       "                                                      genome  \n",
       "accession                                                     \n",
       ">MH498772  ATGAGTCTTCTAACCGAGGTCGAAACGTATGTTCTCTCTATCATTC...  \n",
       ">MH499024  ATGGATTCAAACACCGTGTCAAGCTTTCAGGTAGACTGCTTTCTTT...  \n",
       ">MH499206  ATGGCGTCTCAAGGCACCAAACGATCCTATGAGCAGATGGAAACTG...  \n",
       ">MH498744  ATGGAAGACTTTGTGCGACAATGCTTCAATCCAATGGTTGTCGAGC...  \n",
       ">MH498989  ATGAATCCAAATCAGAAGATTCTATGCACTTCTGCCACCGCTATTG...  \n",
       ">MH499064  ATGGAGAGAATAAAAGAACTAAGAGATCTAATGTCACAGTCTCGCA...  \n",
       ">MH499234  ATGGATGTCAATCCGACTTTACTTTTCTTGAAAGTGCCAGCGCAAA...  \n",
       ">MH498694  ATGGAAGTTCCAGTACTCGCCCTCTTAATGTTAACCAGTATATGCG...  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "upload.query('strain == \"A/environment/Chile/C20173/2016\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "biological-drink",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>46</th>\n",
       "      <th>47</th>\n",
       "      <th>48</th>\n",
       "      <th>49</th>\n",
       "      <th>50</th>\n",
       "      <th>51</th>\n",
       "      <th>52</th>\n",
       "      <th>53</th>\n",
       "      <th>54</th>\n",
       "      <th>55</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.946766</td>\n",
       "      <td>0.784484</td>\n",
       "      <td>0.752536</td>\n",
       "      <td>0.748091</td>\n",
       "      <td>0.855079</td>\n",
       "      <td>0.835843</td>\n",
       "      <td>0.785578</td>\n",
       "      <td>0.804805</td>\n",
       "      <td>0.685704</td>\n",
       "      <td>0.683377</td>\n",
       "      <td>...</td>\n",
       "      <td>0.691221</td>\n",
       "      <td>0.695433</td>\n",
       "      <td>0.693244</td>\n",
       "      <td>0.690111</td>\n",
       "      <td>0.694085</td>\n",
       "      <td>0.679390</td>\n",
       "      <td>0.685788</td>\n",
       "      <td>0.688496</td>\n",
       "      <td>0.684604</td>\n",
       "      <td>0.685460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.784484</td>\n",
       "      <td>0.962751</td>\n",
       "      <td>0.739544</td>\n",
       "      <td>0.739005</td>\n",
       "      <td>0.783708</td>\n",
       "      <td>0.775224</td>\n",
       "      <td>0.799280</td>\n",
       "      <td>0.837051</td>\n",
       "      <td>0.688697</td>\n",
       "      <td>0.693427</td>\n",
       "      <td>...</td>\n",
       "      <td>0.689997</td>\n",
       "      <td>0.690206</td>\n",
       "      <td>0.692046</td>\n",
       "      <td>0.687958</td>\n",
       "      <td>0.694247</td>\n",
       "      <td>0.689242</td>\n",
       "      <td>0.689887</td>\n",
       "      <td>0.688691</td>\n",
       "      <td>0.689435</td>\n",
       "      <td>0.688622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.752536</td>\n",
       "      <td>0.739544</td>\n",
       "      <td>0.943797</td>\n",
       "      <td>0.852779</td>\n",
       "      <td>0.748857</td>\n",
       "      <td>0.743702</td>\n",
       "      <td>0.744241</td>\n",
       "      <td>0.754887</td>\n",
       "      <td>0.691477</td>\n",
       "      <td>0.691642</td>\n",
       "      <td>...</td>\n",
       "      <td>0.687751</td>\n",
       "      <td>0.694049</td>\n",
       "      <td>0.694393</td>\n",
       "      <td>0.694040</td>\n",
       "      <td>0.691360</td>\n",
       "      <td>0.683802</td>\n",
       "      <td>0.693029</td>\n",
       "      <td>0.688316</td>\n",
       "      <td>0.685412</td>\n",
       "      <td>0.689313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.748091</td>\n",
       "      <td>0.739005</td>\n",
       "      <td>0.852779</td>\n",
       "      <td>0.927736</td>\n",
       "      <td>0.746276</td>\n",
       "      <td>0.741655</td>\n",
       "      <td>0.741183</td>\n",
       "      <td>0.745799</td>\n",
       "      <td>0.690116</td>\n",
       "      <td>0.691273</td>\n",
       "      <td>...</td>\n",
       "      <td>0.691283</td>\n",
       "      <td>0.693461</td>\n",
       "      <td>0.690232</td>\n",
       "      <td>0.692321</td>\n",
       "      <td>0.691567</td>\n",
       "      <td>0.683870</td>\n",
       "      <td>0.691567</td>\n",
       "      <td>0.689718</td>\n",
       "      <td>0.687169</td>\n",
       "      <td>0.690127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.855079</td>\n",
       "      <td>0.783708</td>\n",
       "      <td>0.748857</td>\n",
       "      <td>0.746276</td>\n",
       "      <td>0.940973</td>\n",
       "      <td>0.838161</td>\n",
       "      <td>0.783518</td>\n",
       "      <td>0.806429</td>\n",
       "      <td>0.685338</td>\n",
       "      <td>0.690110</td>\n",
       "      <td>...</td>\n",
       "      <td>0.693681</td>\n",
       "      <td>0.691866</td>\n",
       "      <td>0.688448</td>\n",
       "      <td>0.687570</td>\n",
       "      <td>0.688384</td>\n",
       "      <td>0.682856</td>\n",
       "      <td>0.685787</td>\n",
       "      <td>0.686444</td>\n",
       "      <td>0.685081</td>\n",
       "      <td>0.681831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.835843</td>\n",
       "      <td>0.775224</td>\n",
       "      <td>0.743702</td>\n",
       "      <td>0.741655</td>\n",
       "      <td>0.838161</td>\n",
       "      <td>0.918365</td>\n",
       "      <td>0.778696</td>\n",
       "      <td>0.791158</td>\n",
       "      <td>0.684342</td>\n",
       "      <td>0.687866</td>\n",
       "      <td>...</td>\n",
       "      <td>0.693449</td>\n",
       "      <td>0.693463</td>\n",
       "      <td>0.697271</td>\n",
       "      <td>0.689534</td>\n",
       "      <td>0.688879</td>\n",
       "      <td>0.681259</td>\n",
       "      <td>0.681684</td>\n",
       "      <td>0.683977</td>\n",
       "      <td>0.684534</td>\n",
       "      <td>0.685016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.785578</td>\n",
       "      <td>0.799280</td>\n",
       "      <td>0.744241</td>\n",
       "      <td>0.741183</td>\n",
       "      <td>0.783518</td>\n",
       "      <td>0.778696</td>\n",
       "      <td>0.935086</td>\n",
       "      <td>0.840219</td>\n",
       "      <td>0.685485</td>\n",
       "      <td>0.695672</td>\n",
       "      <td>...</td>\n",
       "      <td>0.682206</td>\n",
       "      <td>0.688386</td>\n",
       "      <td>0.691809</td>\n",
       "      <td>0.682970</td>\n",
       "      <td>0.688128</td>\n",
       "      <td>0.682887</td>\n",
       "      <td>0.687947</td>\n",
       "      <td>0.683822</td>\n",
       "      <td>0.683033</td>\n",
       "      <td>0.685883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.804805</td>\n",
       "      <td>0.837051</td>\n",
       "      <td>0.754887</td>\n",
       "      <td>0.745799</td>\n",
       "      <td>0.806429</td>\n",
       "      <td>0.791158</td>\n",
       "      <td>0.840219</td>\n",
       "      <td>0.978459</td>\n",
       "      <td>0.684914</td>\n",
       "      <td>0.692240</td>\n",
       "      <td>...</td>\n",
       "      <td>0.691860</td>\n",
       "      <td>0.695408</td>\n",
       "      <td>0.691224</td>\n",
       "      <td>0.687652</td>\n",
       "      <td>0.689009</td>\n",
       "      <td>0.689360</td>\n",
       "      <td>0.689880</td>\n",
       "      <td>0.690392</td>\n",
       "      <td>0.690296</td>\n",
       "      <td>0.686976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.685704</td>\n",
       "      <td>0.688697</td>\n",
       "      <td>0.691477</td>\n",
       "      <td>0.690116</td>\n",
       "      <td>0.685338</td>\n",
       "      <td>0.684342</td>\n",
       "      <td>0.685485</td>\n",
       "      <td>0.684914</td>\n",
       "      <td>0.980630</td>\n",
       "      <td>0.794265</td>\n",
       "      <td>...</td>\n",
       "      <td>0.694188</td>\n",
       "      <td>0.699058</td>\n",
       "      <td>0.695524</td>\n",
       "      <td>0.701246</td>\n",
       "      <td>0.703924</td>\n",
       "      <td>0.824791</td>\n",
       "      <td>0.860579</td>\n",
       "      <td>0.891979</td>\n",
       "      <td>0.824422</td>\n",
       "      <td>0.820957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.683377</td>\n",
       "      <td>0.693427</td>\n",
       "      <td>0.691642</td>\n",
       "      <td>0.691273</td>\n",
       "      <td>0.690110</td>\n",
       "      <td>0.687866</td>\n",
       "      <td>0.695672</td>\n",
       "      <td>0.692240</td>\n",
       "      <td>0.794265</td>\n",
       "      <td>0.968273</td>\n",
       "      <td>...</td>\n",
       "      <td>0.707710</td>\n",
       "      <td>0.705450</td>\n",
       "      <td>0.699447</td>\n",
       "      <td>0.704009</td>\n",
       "      <td>0.702509</td>\n",
       "      <td>0.792738</td>\n",
       "      <td>0.799169</td>\n",
       "      <td>0.803361</td>\n",
       "      <td>0.793323</td>\n",
       "      <td>0.787099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.686960</td>\n",
       "      <td>0.694550</td>\n",
       "      <td>0.694589</td>\n",
       "      <td>0.692494</td>\n",
       "      <td>0.689357</td>\n",
       "      <td>0.688192</td>\n",
       "      <td>0.691176</td>\n",
       "      <td>0.691728</td>\n",
       "      <td>0.792825</td>\n",
       "      <td>0.847368</td>\n",
       "      <td>...</td>\n",
       "      <td>0.707091</td>\n",
       "      <td>0.703671</td>\n",
       "      <td>0.703952</td>\n",
       "      <td>0.707958</td>\n",
       "      <td>0.700062</td>\n",
       "      <td>0.781437</td>\n",
       "      <td>0.791131</td>\n",
       "      <td>0.795788</td>\n",
       "      <td>0.781567</td>\n",
       "      <td>0.781458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>0.746761</td>\n",
       "      <td>0.747820</td>\n",
       "      <td>0.776813</td>\n",
       "      <td>0.767814</td>\n",
       "      <td>0.744930</td>\n",
       "      <td>0.741248</td>\n",
       "      <td>0.745283</td>\n",
       "      <td>0.748863</td>\n",
       "      <td>0.688190</td>\n",
       "      <td>0.687629</td>\n",
       "      <td>...</td>\n",
       "      <td>0.696358</td>\n",
       "      <td>0.700379</td>\n",
       "      <td>0.698689</td>\n",
       "      <td>0.696961</td>\n",
       "      <td>0.695355</td>\n",
       "      <td>0.682961</td>\n",
       "      <td>0.690327</td>\n",
       "      <td>0.685242</td>\n",
       "      <td>0.689469</td>\n",
       "      <td>0.687254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.689367</td>\n",
       "      <td>0.696623</td>\n",
       "      <td>0.692582</td>\n",
       "      <td>0.693592</td>\n",
       "      <td>0.687156</td>\n",
       "      <td>0.682622</td>\n",
       "      <td>0.687031</td>\n",
       "      <td>0.690567</td>\n",
       "      <td>0.836459</td>\n",
       "      <td>0.810842</td>\n",
       "      <td>...</td>\n",
       "      <td>0.704498</td>\n",
       "      <td>0.705530</td>\n",
       "      <td>0.696778</td>\n",
       "      <td>0.705075</td>\n",
       "      <td>0.709342</td>\n",
       "      <td>0.806203</td>\n",
       "      <td>0.843295</td>\n",
       "      <td>0.851791</td>\n",
       "      <td>0.808390</td>\n",
       "      <td>0.809583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.738975</td>\n",
       "      <td>0.737661</td>\n",
       "      <td>0.770044</td>\n",
       "      <td>0.766586</td>\n",
       "      <td>0.742682</td>\n",
       "      <td>0.739870</td>\n",
       "      <td>0.734704</td>\n",
       "      <td>0.746967</td>\n",
       "      <td>0.685760</td>\n",
       "      <td>0.690778</td>\n",
       "      <td>...</td>\n",
       "      <td>0.698059</td>\n",
       "      <td>0.692023</td>\n",
       "      <td>0.697127</td>\n",
       "      <td>0.692961</td>\n",
       "      <td>0.693790</td>\n",
       "      <td>0.687649</td>\n",
       "      <td>0.685804</td>\n",
       "      <td>0.686139</td>\n",
       "      <td>0.685366</td>\n",
       "      <td>0.684207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.729346</td>\n",
       "      <td>0.726935</td>\n",
       "      <td>0.729278</td>\n",
       "      <td>0.739978</td>\n",
       "      <td>0.730523</td>\n",
       "      <td>0.729589</td>\n",
       "      <td>0.721662</td>\n",
       "      <td>0.721272</td>\n",
       "      <td>0.688738</td>\n",
       "      <td>0.690219</td>\n",
       "      <td>...</td>\n",
       "      <td>0.697438</td>\n",
       "      <td>0.697647</td>\n",
       "      <td>0.695963</td>\n",
       "      <td>0.701424</td>\n",
       "      <td>0.695834</td>\n",
       "      <td>0.685959</td>\n",
       "      <td>0.688320</td>\n",
       "      <td>0.686133</td>\n",
       "      <td>0.684935</td>\n",
       "      <td>0.685964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>0.699031</td>\n",
       "      <td>0.700607</td>\n",
       "      <td>0.705223</td>\n",
       "      <td>0.703808</td>\n",
       "      <td>0.702863</td>\n",
       "      <td>0.700715</td>\n",
       "      <td>0.694441</td>\n",
       "      <td>0.701940</td>\n",
       "      <td>0.685355</td>\n",
       "      <td>0.690034</td>\n",
       "      <td>...</td>\n",
       "      <td>0.689627</td>\n",
       "      <td>0.691342</td>\n",
       "      <td>0.685565</td>\n",
       "      <td>0.695097</td>\n",
       "      <td>0.693885</td>\n",
       "      <td>0.686691</td>\n",
       "      <td>0.687956</td>\n",
       "      <td>0.687628</td>\n",
       "      <td>0.685799</td>\n",
       "      <td>0.685002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>0.787265</td>\n",
       "      <td>0.862260</td>\n",
       "      <td>0.735654</td>\n",
       "      <td>0.735596</td>\n",
       "      <td>0.786371</td>\n",
       "      <td>0.782351</td>\n",
       "      <td>0.793290</td>\n",
       "      <td>0.840812</td>\n",
       "      <td>0.691302</td>\n",
       "      <td>0.694984</td>\n",
       "      <td>...</td>\n",
       "      <td>0.691189</td>\n",
       "      <td>0.687254</td>\n",
       "      <td>0.692511</td>\n",
       "      <td>0.687656</td>\n",
       "      <td>0.691336</td>\n",
       "      <td>0.689206</td>\n",
       "      <td>0.687066</td>\n",
       "      <td>0.688369</td>\n",
       "      <td>0.691705</td>\n",
       "      <td>0.687898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0.785883</td>\n",
       "      <td>0.850610</td>\n",
       "      <td>0.737231</td>\n",
       "      <td>0.732695</td>\n",
       "      <td>0.781559</td>\n",
       "      <td>0.775408</td>\n",
       "      <td>0.797154</td>\n",
       "      <td>0.827275</td>\n",
       "      <td>0.689083</td>\n",
       "      <td>0.692608</td>\n",
       "      <td>...</td>\n",
       "      <td>0.690803</td>\n",
       "      <td>0.684944</td>\n",
       "      <td>0.681766</td>\n",
       "      <td>0.685216</td>\n",
       "      <td>0.687815</td>\n",
       "      <td>0.686353</td>\n",
       "      <td>0.686374</td>\n",
       "      <td>0.684489</td>\n",
       "      <td>0.694876</td>\n",
       "      <td>0.686506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>0.744493</td>\n",
       "      <td>0.736539</td>\n",
       "      <td>0.763029</td>\n",
       "      <td>0.762969</td>\n",
       "      <td>0.737843</td>\n",
       "      <td>0.741652</td>\n",
       "      <td>0.730041</td>\n",
       "      <td>0.741371</td>\n",
       "      <td>0.687557</td>\n",
       "      <td>0.693474</td>\n",
       "      <td>...</td>\n",
       "      <td>0.689003</td>\n",
       "      <td>0.685504</td>\n",
       "      <td>0.694887</td>\n",
       "      <td>0.689427</td>\n",
       "      <td>0.688593</td>\n",
       "      <td>0.683211</td>\n",
       "      <td>0.679507</td>\n",
       "      <td>0.681351</td>\n",
       "      <td>0.683566</td>\n",
       "      <td>0.684728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>0.739538</td>\n",
       "      <td>0.730021</td>\n",
       "      <td>0.756737</td>\n",
       "      <td>0.754560</td>\n",
       "      <td>0.730421</td>\n",
       "      <td>0.731175</td>\n",
       "      <td>0.726913</td>\n",
       "      <td>0.741969</td>\n",
       "      <td>0.683686</td>\n",
       "      <td>0.683625</td>\n",
       "      <td>...</td>\n",
       "      <td>0.686078</td>\n",
       "      <td>0.690045</td>\n",
       "      <td>0.686927</td>\n",
       "      <td>0.692093</td>\n",
       "      <td>0.692752</td>\n",
       "      <td>0.683864</td>\n",
       "      <td>0.682368</td>\n",
       "      <td>0.685348</td>\n",
       "      <td>0.685245</td>\n",
       "      <td>0.682432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>0.725886</td>\n",
       "      <td>0.721147</td>\n",
       "      <td>0.722897</td>\n",
       "      <td>0.730430</td>\n",
       "      <td>0.723369</td>\n",
       "      <td>0.723808</td>\n",
       "      <td>0.717401</td>\n",
       "      <td>0.723879</td>\n",
       "      <td>0.682195</td>\n",
       "      <td>0.688072</td>\n",
       "      <td>...</td>\n",
       "      <td>0.694673</td>\n",
       "      <td>0.694982</td>\n",
       "      <td>0.687648</td>\n",
       "      <td>0.691892</td>\n",
       "      <td>0.695720</td>\n",
       "      <td>0.681714</td>\n",
       "      <td>0.679431</td>\n",
       "      <td>0.681809</td>\n",
       "      <td>0.681251</td>\n",
       "      <td>0.682600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>0.729812</td>\n",
       "      <td>0.727624</td>\n",
       "      <td>0.725639</td>\n",
       "      <td>0.736433</td>\n",
       "      <td>0.731806</td>\n",
       "      <td>0.728020</td>\n",
       "      <td>0.722257</td>\n",
       "      <td>0.731727</td>\n",
       "      <td>0.687446</td>\n",
       "      <td>0.687051</td>\n",
       "      <td>...</td>\n",
       "      <td>0.690380</td>\n",
       "      <td>0.695031</td>\n",
       "      <td>0.686062</td>\n",
       "      <td>0.691193</td>\n",
       "      <td>0.691277</td>\n",
       "      <td>0.684785</td>\n",
       "      <td>0.691776</td>\n",
       "      <td>0.692421</td>\n",
       "      <td>0.684997</td>\n",
       "      <td>0.684389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>0.688081</td>\n",
       "      <td>0.693663</td>\n",
       "      <td>0.689697</td>\n",
       "      <td>0.695544</td>\n",
       "      <td>0.691301</td>\n",
       "      <td>0.689248</td>\n",
       "      <td>0.689314</td>\n",
       "      <td>0.694659</td>\n",
       "      <td>0.823860</td>\n",
       "      <td>0.802935</td>\n",
       "      <td>...</td>\n",
       "      <td>0.697877</td>\n",
       "      <td>0.701213</td>\n",
       "      <td>0.695666</td>\n",
       "      <td>0.704055</td>\n",
       "      <td>0.704118</td>\n",
       "      <td>0.804665</td>\n",
       "      <td>0.837047</td>\n",
       "      <td>0.836126</td>\n",
       "      <td>0.816541</td>\n",
       "      <td>0.807831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>0.687456</td>\n",
       "      <td>0.692161</td>\n",
       "      <td>0.691181</td>\n",
       "      <td>0.692000</td>\n",
       "      <td>0.689011</td>\n",
       "      <td>0.686394</td>\n",
       "      <td>0.687625</td>\n",
       "      <td>0.691858</td>\n",
       "      <td>0.834804</td>\n",
       "      <td>0.810995</td>\n",
       "      <td>...</td>\n",
       "      <td>0.700565</td>\n",
       "      <td>0.703602</td>\n",
       "      <td>0.693407</td>\n",
       "      <td>0.700896</td>\n",
       "      <td>0.701608</td>\n",
       "      <td>0.815719</td>\n",
       "      <td>0.842639</td>\n",
       "      <td>0.847676</td>\n",
       "      <td>0.818929</td>\n",
       "      <td>0.815907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>0.681035</td>\n",
       "      <td>0.690320</td>\n",
       "      <td>0.687548</td>\n",
       "      <td>0.690133</td>\n",
       "      <td>0.684358</td>\n",
       "      <td>0.683188</td>\n",
       "      <td>0.683958</td>\n",
       "      <td>0.683327</td>\n",
       "      <td>0.818352</td>\n",
       "      <td>0.774057</td>\n",
       "      <td>...</td>\n",
       "      <td>0.694224</td>\n",
       "      <td>0.694599</td>\n",
       "      <td>0.690264</td>\n",
       "      <td>0.699980</td>\n",
       "      <td>0.696235</td>\n",
       "      <td>0.840703</td>\n",
       "      <td>0.822159</td>\n",
       "      <td>0.821998</td>\n",
       "      <td>0.852339</td>\n",
       "      <td>0.854867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>0.714099</td>\n",
       "      <td>0.711805</td>\n",
       "      <td>0.725097</td>\n",
       "      <td>0.721855</td>\n",
       "      <td>0.714047</td>\n",
       "      <td>0.705997</td>\n",
       "      <td>0.717345</td>\n",
       "      <td>0.717395</td>\n",
       "      <td>0.688709</td>\n",
       "      <td>0.691080</td>\n",
       "      <td>...</td>\n",
       "      <td>0.684539</td>\n",
       "      <td>0.688905</td>\n",
       "      <td>0.687441</td>\n",
       "      <td>0.689074</td>\n",
       "      <td>0.697398</td>\n",
       "      <td>0.686018</td>\n",
       "      <td>0.689500</td>\n",
       "      <td>0.693489</td>\n",
       "      <td>0.685827</td>\n",
       "      <td>0.692284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>0.704652</td>\n",
       "      <td>0.707612</td>\n",
       "      <td>0.709023</td>\n",
       "      <td>0.712349</td>\n",
       "      <td>0.703589</td>\n",
       "      <td>0.701030</td>\n",
       "      <td>0.707807</td>\n",
       "      <td>0.709802</td>\n",
       "      <td>0.690794</td>\n",
       "      <td>0.691121</td>\n",
       "      <td>...</td>\n",
       "      <td>0.685254</td>\n",
       "      <td>0.686822</td>\n",
       "      <td>0.683413</td>\n",
       "      <td>0.687667</td>\n",
       "      <td>0.689050</td>\n",
       "      <td>0.692592</td>\n",
       "      <td>0.694898</td>\n",
       "      <td>0.693303</td>\n",
       "      <td>0.695722</td>\n",
       "      <td>0.692336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>0.684815</td>\n",
       "      <td>0.689857</td>\n",
       "      <td>0.687512</td>\n",
       "      <td>0.689807</td>\n",
       "      <td>0.683009</td>\n",
       "      <td>0.684797</td>\n",
       "      <td>0.687056</td>\n",
       "      <td>0.687818</td>\n",
       "      <td>0.818804</td>\n",
       "      <td>0.788277</td>\n",
       "      <td>...</td>\n",
       "      <td>0.697725</td>\n",
       "      <td>0.701389</td>\n",
       "      <td>0.693778</td>\n",
       "      <td>0.700816</td>\n",
       "      <td>0.703718</td>\n",
       "      <td>0.835858</td>\n",
       "      <td>0.821228</td>\n",
       "      <td>0.830261</td>\n",
       "      <td>0.876224</td>\n",
       "      <td>0.925893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>0.706640</td>\n",
       "      <td>0.714567</td>\n",
       "      <td>0.718418</td>\n",
       "      <td>0.715357</td>\n",
       "      <td>0.711410</td>\n",
       "      <td>0.701832</td>\n",
       "      <td>0.714158</td>\n",
       "      <td>0.724998</td>\n",
       "      <td>0.682503</td>\n",
       "      <td>0.690065</td>\n",
       "      <td>...</td>\n",
       "      <td>0.686010</td>\n",
       "      <td>0.684952</td>\n",
       "      <td>0.691400</td>\n",
       "      <td>0.684625</td>\n",
       "      <td>0.689600</td>\n",
       "      <td>0.681713</td>\n",
       "      <td>0.682932</td>\n",
       "      <td>0.685321</td>\n",
       "      <td>0.681082</td>\n",
       "      <td>0.679388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>0.720643</td>\n",
       "      <td>0.715139</td>\n",
       "      <td>0.724594</td>\n",
       "      <td>0.719920</td>\n",
       "      <td>0.712772</td>\n",
       "      <td>0.705993</td>\n",
       "      <td>0.716913</td>\n",
       "      <td>0.723758</td>\n",
       "      <td>0.692098</td>\n",
       "      <td>0.691069</td>\n",
       "      <td>...</td>\n",
       "      <td>0.688618</td>\n",
       "      <td>0.690367</td>\n",
       "      <td>0.690208</td>\n",
       "      <td>0.691416</td>\n",
       "      <td>0.694258</td>\n",
       "      <td>0.687068</td>\n",
       "      <td>0.694342</td>\n",
       "      <td>0.692729</td>\n",
       "      <td>0.690266</td>\n",
       "      <td>0.691319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>0.702260</td>\n",
       "      <td>0.698761</td>\n",
       "      <td>0.705390</td>\n",
       "      <td>0.709909</td>\n",
       "      <td>0.705661</td>\n",
       "      <td>0.699265</td>\n",
       "      <td>0.696191</td>\n",
       "      <td>0.706438</td>\n",
       "      <td>0.684854</td>\n",
       "      <td>0.683758</td>\n",
       "      <td>...</td>\n",
       "      <td>0.678749</td>\n",
       "      <td>0.685358</td>\n",
       "      <td>0.677763</td>\n",
       "      <td>0.686263</td>\n",
       "      <td>0.686510</td>\n",
       "      <td>0.689406</td>\n",
       "      <td>0.691110</td>\n",
       "      <td>0.687476</td>\n",
       "      <td>0.687762</td>\n",
       "      <td>0.686205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>0.692094</td>\n",
       "      <td>0.690951</td>\n",
       "      <td>0.699184</td>\n",
       "      <td>0.705660</td>\n",
       "      <td>0.696749</td>\n",
       "      <td>0.695473</td>\n",
       "      <td>0.696026</td>\n",
       "      <td>0.699360</td>\n",
       "      <td>0.680619</td>\n",
       "      <td>0.693531</td>\n",
       "      <td>...</td>\n",
       "      <td>0.678996</td>\n",
       "      <td>0.678189</td>\n",
       "      <td>0.680845</td>\n",
       "      <td>0.681603</td>\n",
       "      <td>0.682485</td>\n",
       "      <td>0.683838</td>\n",
       "      <td>0.690910</td>\n",
       "      <td>0.685294</td>\n",
       "      <td>0.685084</td>\n",
       "      <td>0.682508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>0.701636</td>\n",
       "      <td>0.705187</td>\n",
       "      <td>0.708791</td>\n",
       "      <td>0.707977</td>\n",
       "      <td>0.700041</td>\n",
       "      <td>0.698274</td>\n",
       "      <td>0.699800</td>\n",
       "      <td>0.703497</td>\n",
       "      <td>0.683888</td>\n",
       "      <td>0.690808</td>\n",
       "      <td>...</td>\n",
       "      <td>0.683789</td>\n",
       "      <td>0.684805</td>\n",
       "      <td>0.681069</td>\n",
       "      <td>0.688336</td>\n",
       "      <td>0.689574</td>\n",
       "      <td>0.686909</td>\n",
       "      <td>0.689174</td>\n",
       "      <td>0.687942</td>\n",
       "      <td>0.689297</td>\n",
       "      <td>0.685733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>0.700601</td>\n",
       "      <td>0.692580</td>\n",
       "      <td>0.699437</td>\n",
       "      <td>0.701523</td>\n",
       "      <td>0.697640</td>\n",
       "      <td>0.693573</td>\n",
       "      <td>0.696412</td>\n",
       "      <td>0.702463</td>\n",
       "      <td>0.688935</td>\n",
       "      <td>0.688403</td>\n",
       "      <td>...</td>\n",
       "      <td>0.692649</td>\n",
       "      <td>0.694007</td>\n",
       "      <td>0.683007</td>\n",
       "      <td>0.688940</td>\n",
       "      <td>0.690106</td>\n",
       "      <td>0.687514</td>\n",
       "      <td>0.693279</td>\n",
       "      <td>0.688500</td>\n",
       "      <td>0.689263</td>\n",
       "      <td>0.687698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>0.707263</td>\n",
       "      <td>0.696748</td>\n",
       "      <td>0.711143</td>\n",
       "      <td>0.709049</td>\n",
       "      <td>0.700812</td>\n",
       "      <td>0.700207</td>\n",
       "      <td>0.700826</td>\n",
       "      <td>0.702936</td>\n",
       "      <td>0.693789</td>\n",
       "      <td>0.699791</td>\n",
       "      <td>...</td>\n",
       "      <td>0.699269</td>\n",
       "      <td>0.699436</td>\n",
       "      <td>0.694390</td>\n",
       "      <td>0.698000</td>\n",
       "      <td>0.698361</td>\n",
       "      <td>0.695674</td>\n",
       "      <td>0.697508</td>\n",
       "      <td>0.693083</td>\n",
       "      <td>0.694169</td>\n",
       "      <td>0.693324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>0.681399</td>\n",
       "      <td>0.690096</td>\n",
       "      <td>0.685056</td>\n",
       "      <td>0.685519</td>\n",
       "      <td>0.681712</td>\n",
       "      <td>0.684156</td>\n",
       "      <td>0.683094</td>\n",
       "      <td>0.689466</td>\n",
       "      <td>0.819249</td>\n",
       "      <td>0.785813</td>\n",
       "      <td>...</td>\n",
       "      <td>0.696249</td>\n",
       "      <td>0.699174</td>\n",
       "      <td>0.691973</td>\n",
       "      <td>0.702660</td>\n",
       "      <td>0.698318</td>\n",
       "      <td>0.866762</td>\n",
       "      <td>0.830911</td>\n",
       "      <td>0.834396</td>\n",
       "      <td>0.845400</td>\n",
       "      <td>0.832100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>0.703307</td>\n",
       "      <td>0.702821</td>\n",
       "      <td>0.701724</td>\n",
       "      <td>0.703205</td>\n",
       "      <td>0.705374</td>\n",
       "      <td>0.696667</td>\n",
       "      <td>0.706469</td>\n",
       "      <td>0.709612</td>\n",
       "      <td>0.686685</td>\n",
       "      <td>0.687720</td>\n",
       "      <td>...</td>\n",
       "      <td>0.680832</td>\n",
       "      <td>0.682901</td>\n",
       "      <td>0.684193</td>\n",
       "      <td>0.689249</td>\n",
       "      <td>0.687465</td>\n",
       "      <td>0.687267</td>\n",
       "      <td>0.690843</td>\n",
       "      <td>0.685676</td>\n",
       "      <td>0.685183</td>\n",
       "      <td>0.684767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>0.707554</td>\n",
       "      <td>0.700339</td>\n",
       "      <td>0.703640</td>\n",
       "      <td>0.704405</td>\n",
       "      <td>0.704596</td>\n",
       "      <td>0.699038</td>\n",
       "      <td>0.704195</td>\n",
       "      <td>0.703607</td>\n",
       "      <td>0.681905</td>\n",
       "      <td>0.685667</td>\n",
       "      <td>...</td>\n",
       "      <td>0.682304</td>\n",
       "      <td>0.691095</td>\n",
       "      <td>0.687009</td>\n",
       "      <td>0.690618</td>\n",
       "      <td>0.693596</td>\n",
       "      <td>0.685791</td>\n",
       "      <td>0.687276</td>\n",
       "      <td>0.679027</td>\n",
       "      <td>0.681728</td>\n",
       "      <td>0.679890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>0.692182</td>\n",
       "      <td>0.692357</td>\n",
       "      <td>0.691755</td>\n",
       "      <td>0.687601</td>\n",
       "      <td>0.685918</td>\n",
       "      <td>0.687281</td>\n",
       "      <td>0.686365</td>\n",
       "      <td>0.690050</td>\n",
       "      <td>0.727783</td>\n",
       "      <td>0.731722</td>\n",
       "      <td>...</td>\n",
       "      <td>0.701411</td>\n",
       "      <td>0.700268</td>\n",
       "      <td>0.697766</td>\n",
       "      <td>0.699078</td>\n",
       "      <td>0.702562</td>\n",
       "      <td>0.726067</td>\n",
       "      <td>0.729644</td>\n",
       "      <td>0.729324</td>\n",
       "      <td>0.724008</td>\n",
       "      <td>0.725392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>0.688785</td>\n",
       "      <td>0.690875</td>\n",
       "      <td>0.693429</td>\n",
       "      <td>0.693209</td>\n",
       "      <td>0.689832</td>\n",
       "      <td>0.693138</td>\n",
       "      <td>0.689112</td>\n",
       "      <td>0.689911</td>\n",
       "      <td>0.741501</td>\n",
       "      <td>0.747157</td>\n",
       "      <td>...</td>\n",
       "      <td>0.702323</td>\n",
       "      <td>0.705508</td>\n",
       "      <td>0.700474</td>\n",
       "      <td>0.702111</td>\n",
       "      <td>0.703942</td>\n",
       "      <td>0.736845</td>\n",
       "      <td>0.744486</td>\n",
       "      <td>0.740859</td>\n",
       "      <td>0.731997</td>\n",
       "      <td>0.734582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>0.692229</td>\n",
       "      <td>0.690079</td>\n",
       "      <td>0.690717</td>\n",
       "      <td>0.694002</td>\n",
       "      <td>0.688640</td>\n",
       "      <td>0.688014</td>\n",
       "      <td>0.681660</td>\n",
       "      <td>0.685623</td>\n",
       "      <td>0.740425</td>\n",
       "      <td>0.736576</td>\n",
       "      <td>...</td>\n",
       "      <td>0.701457</td>\n",
       "      <td>0.705491</td>\n",
       "      <td>0.696564</td>\n",
       "      <td>0.702807</td>\n",
       "      <td>0.705436</td>\n",
       "      <td>0.730739</td>\n",
       "      <td>0.737616</td>\n",
       "      <td>0.738156</td>\n",
       "      <td>0.731751</td>\n",
       "      <td>0.729335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>0.693813</td>\n",
       "      <td>0.694363</td>\n",
       "      <td>0.692755</td>\n",
       "      <td>0.691739</td>\n",
       "      <td>0.687865</td>\n",
       "      <td>0.688573</td>\n",
       "      <td>0.690376</td>\n",
       "      <td>0.690167</td>\n",
       "      <td>0.740359</td>\n",
       "      <td>0.741916</td>\n",
       "      <td>...</td>\n",
       "      <td>0.704081</td>\n",
       "      <td>0.709053</td>\n",
       "      <td>0.702169</td>\n",
       "      <td>0.705012</td>\n",
       "      <td>0.709534</td>\n",
       "      <td>0.732731</td>\n",
       "      <td>0.741406</td>\n",
       "      <td>0.740407</td>\n",
       "      <td>0.733348</td>\n",
       "      <td>0.727890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>0.685243</td>\n",
       "      <td>0.690668</td>\n",
       "      <td>0.685323</td>\n",
       "      <td>0.687536</td>\n",
       "      <td>0.686926</td>\n",
       "      <td>0.685685</td>\n",
       "      <td>0.685622</td>\n",
       "      <td>0.688523</td>\n",
       "      <td>0.704345</td>\n",
       "      <td>0.700841</td>\n",
       "      <td>...</td>\n",
       "      <td>0.728785</td>\n",
       "      <td>0.735986</td>\n",
       "      <td>0.729591</td>\n",
       "      <td>0.731898</td>\n",
       "      <td>0.739419</td>\n",
       "      <td>0.704444</td>\n",
       "      <td>0.706921</td>\n",
       "      <td>0.703559</td>\n",
       "      <td>0.699184</td>\n",
       "      <td>0.695467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>0.694135</td>\n",
       "      <td>0.687031</td>\n",
       "      <td>0.690617</td>\n",
       "      <td>0.691779</td>\n",
       "      <td>0.691994</td>\n",
       "      <td>0.690121</td>\n",
       "      <td>0.685878</td>\n",
       "      <td>0.688076</td>\n",
       "      <td>0.704280</td>\n",
       "      <td>0.700512</td>\n",
       "      <td>...</td>\n",
       "      <td>0.727824</td>\n",
       "      <td>0.728051</td>\n",
       "      <td>0.725200</td>\n",
       "      <td>0.738422</td>\n",
       "      <td>0.738883</td>\n",
       "      <td>0.698577</td>\n",
       "      <td>0.692087</td>\n",
       "      <td>0.701419</td>\n",
       "      <td>0.701729</td>\n",
       "      <td>0.699648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>0.690455</td>\n",
       "      <td>0.689454</td>\n",
       "      <td>0.692662</td>\n",
       "      <td>0.690318</td>\n",
       "      <td>0.691465</td>\n",
       "      <td>0.686981</td>\n",
       "      <td>0.686985</td>\n",
       "      <td>0.688102</td>\n",
       "      <td>0.707650</td>\n",
       "      <td>0.698083</td>\n",
       "      <td>...</td>\n",
       "      <td>0.728002</td>\n",
       "      <td>0.735047</td>\n",
       "      <td>0.736771</td>\n",
       "      <td>0.740318</td>\n",
       "      <td>0.745256</td>\n",
       "      <td>0.698964</td>\n",
       "      <td>0.705133</td>\n",
       "      <td>0.704192</td>\n",
       "      <td>0.699298</td>\n",
       "      <td>0.698154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>0.692627</td>\n",
       "      <td>0.688207</td>\n",
       "      <td>0.696655</td>\n",
       "      <td>0.695026</td>\n",
       "      <td>0.689197</td>\n",
       "      <td>0.686827</td>\n",
       "      <td>0.686675</td>\n",
       "      <td>0.693425</td>\n",
       "      <td>0.698102</td>\n",
       "      <td>0.700431</td>\n",
       "      <td>...</td>\n",
       "      <td>0.771901</td>\n",
       "      <td>0.771651</td>\n",
       "      <td>0.756025</td>\n",
       "      <td>0.769104</td>\n",
       "      <td>0.775018</td>\n",
       "      <td>0.693354</td>\n",
       "      <td>0.698197</td>\n",
       "      <td>0.699042</td>\n",
       "      <td>0.697305</td>\n",
       "      <td>0.698650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>0.691221</td>\n",
       "      <td>0.689997</td>\n",
       "      <td>0.687751</td>\n",
       "      <td>0.691283</td>\n",
       "      <td>0.693681</td>\n",
       "      <td>0.693449</td>\n",
       "      <td>0.682206</td>\n",
       "      <td>0.691860</td>\n",
       "      <td>0.694188</td>\n",
       "      <td>0.707710</td>\n",
       "      <td>...</td>\n",
       "      <td>0.973500</td>\n",
       "      <td>0.825117</td>\n",
       "      <td>0.776492</td>\n",
       "      <td>0.788097</td>\n",
       "      <td>0.789969</td>\n",
       "      <td>0.698240</td>\n",
       "      <td>0.692139</td>\n",
       "      <td>0.698310</td>\n",
       "      <td>0.698962</td>\n",
       "      <td>0.701215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>0.695433</td>\n",
       "      <td>0.690206</td>\n",
       "      <td>0.694049</td>\n",
       "      <td>0.693461</td>\n",
       "      <td>0.691866</td>\n",
       "      <td>0.693463</td>\n",
       "      <td>0.688386</td>\n",
       "      <td>0.695408</td>\n",
       "      <td>0.699058</td>\n",
       "      <td>0.705450</td>\n",
       "      <td>...</td>\n",
       "      <td>0.825117</td>\n",
       "      <td>0.946274</td>\n",
       "      <td>0.774923</td>\n",
       "      <td>0.785969</td>\n",
       "      <td>0.789483</td>\n",
       "      <td>0.700103</td>\n",
       "      <td>0.699019</td>\n",
       "      <td>0.703583</td>\n",
       "      <td>0.696831</td>\n",
       "      <td>0.700532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>0.693244</td>\n",
       "      <td>0.692046</td>\n",
       "      <td>0.694393</td>\n",
       "      <td>0.690232</td>\n",
       "      <td>0.688448</td>\n",
       "      <td>0.697271</td>\n",
       "      <td>0.691809</td>\n",
       "      <td>0.691224</td>\n",
       "      <td>0.695524</td>\n",
       "      <td>0.699447</td>\n",
       "      <td>...</td>\n",
       "      <td>0.776492</td>\n",
       "      <td>0.774923</td>\n",
       "      <td>0.971693</td>\n",
       "      <td>0.795111</td>\n",
       "      <td>0.807809</td>\n",
       "      <td>0.696179</td>\n",
       "      <td>0.690735</td>\n",
       "      <td>0.697678</td>\n",
       "      <td>0.691569</td>\n",
       "      <td>0.693872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>0.690111</td>\n",
       "      <td>0.687958</td>\n",
       "      <td>0.694040</td>\n",
       "      <td>0.692321</td>\n",
       "      <td>0.687570</td>\n",
       "      <td>0.689534</td>\n",
       "      <td>0.682970</td>\n",
       "      <td>0.687652</td>\n",
       "      <td>0.701246</td>\n",
       "      <td>0.704009</td>\n",
       "      <td>...</td>\n",
       "      <td>0.788097</td>\n",
       "      <td>0.785969</td>\n",
       "      <td>0.795111</td>\n",
       "      <td>0.909894</td>\n",
       "      <td>0.841370</td>\n",
       "      <td>0.700550</td>\n",
       "      <td>0.699980</td>\n",
       "      <td>0.702556</td>\n",
       "      <td>0.698905</td>\n",
       "      <td>0.699560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>0.694085</td>\n",
       "      <td>0.694247</td>\n",
       "      <td>0.691360</td>\n",
       "      <td>0.691567</td>\n",
       "      <td>0.688384</td>\n",
       "      <td>0.688879</td>\n",
       "      <td>0.688128</td>\n",
       "      <td>0.689009</td>\n",
       "      <td>0.703924</td>\n",
       "      <td>0.702509</td>\n",
       "      <td>...</td>\n",
       "      <td>0.789969</td>\n",
       "      <td>0.789483</td>\n",
       "      <td>0.807809</td>\n",
       "      <td>0.841370</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>0.698199</td>\n",
       "      <td>0.698687</td>\n",
       "      <td>0.703544</td>\n",
       "      <td>0.702787</td>\n",
       "      <td>0.704523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>0.679390</td>\n",
       "      <td>0.689242</td>\n",
       "      <td>0.683802</td>\n",
       "      <td>0.683870</td>\n",
       "      <td>0.682856</td>\n",
       "      <td>0.681259</td>\n",
       "      <td>0.682887</td>\n",
       "      <td>0.689360</td>\n",
       "      <td>0.824791</td>\n",
       "      <td>0.792738</td>\n",
       "      <td>...</td>\n",
       "      <td>0.698240</td>\n",
       "      <td>0.700103</td>\n",
       "      <td>0.696179</td>\n",
       "      <td>0.700550</td>\n",
       "      <td>0.698199</td>\n",
       "      <td>0.963780</td>\n",
       "      <td>0.826024</td>\n",
       "      <td>0.830656</td>\n",
       "      <td>0.849111</td>\n",
       "      <td>0.839600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>0.685788</td>\n",
       "      <td>0.689887</td>\n",
       "      <td>0.693029</td>\n",
       "      <td>0.691567</td>\n",
       "      <td>0.685787</td>\n",
       "      <td>0.681684</td>\n",
       "      <td>0.687947</td>\n",
       "      <td>0.689880</td>\n",
       "      <td>0.860579</td>\n",
       "      <td>0.799169</td>\n",
       "      <td>...</td>\n",
       "      <td>0.692139</td>\n",
       "      <td>0.699019</td>\n",
       "      <td>0.690735</td>\n",
       "      <td>0.699980</td>\n",
       "      <td>0.698687</td>\n",
       "      <td>0.826024</td>\n",
       "      <td>0.939680</td>\n",
       "      <td>0.878461</td>\n",
       "      <td>0.827197</td>\n",
       "      <td>0.823202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>0.688496</td>\n",
       "      <td>0.688691</td>\n",
       "      <td>0.688316</td>\n",
       "      <td>0.689718</td>\n",
       "      <td>0.686444</td>\n",
       "      <td>0.683977</td>\n",
       "      <td>0.683822</td>\n",
       "      <td>0.690392</td>\n",
       "      <td>0.891979</td>\n",
       "      <td>0.803361</td>\n",
       "      <td>...</td>\n",
       "      <td>0.698310</td>\n",
       "      <td>0.703583</td>\n",
       "      <td>0.697678</td>\n",
       "      <td>0.702556</td>\n",
       "      <td>0.703544</td>\n",
       "      <td>0.830656</td>\n",
       "      <td>0.878461</td>\n",
       "      <td>0.941551</td>\n",
       "      <td>0.836026</td>\n",
       "      <td>0.830635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>0.684604</td>\n",
       "      <td>0.689435</td>\n",
       "      <td>0.685412</td>\n",
       "      <td>0.687169</td>\n",
       "      <td>0.685081</td>\n",
       "      <td>0.684534</td>\n",
       "      <td>0.683033</td>\n",
       "      <td>0.690296</td>\n",
       "      <td>0.824422</td>\n",
       "      <td>0.793323</td>\n",
       "      <td>...</td>\n",
       "      <td>0.698962</td>\n",
       "      <td>0.696831</td>\n",
       "      <td>0.691569</td>\n",
       "      <td>0.698905</td>\n",
       "      <td>0.702787</td>\n",
       "      <td>0.849111</td>\n",
       "      <td>0.827197</td>\n",
       "      <td>0.836026</td>\n",
       "      <td>0.984710</td>\n",
       "      <td>0.878301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>0.685460</td>\n",
       "      <td>0.688622</td>\n",
       "      <td>0.689313</td>\n",
       "      <td>0.690127</td>\n",
       "      <td>0.681831</td>\n",
       "      <td>0.685016</td>\n",
       "      <td>0.685883</td>\n",
       "      <td>0.686976</td>\n",
       "      <td>0.820957</td>\n",
       "      <td>0.787099</td>\n",
       "      <td>...</td>\n",
       "      <td>0.701215</td>\n",
       "      <td>0.700532</td>\n",
       "      <td>0.693872</td>\n",
       "      <td>0.699560</td>\n",
       "      <td>0.704523</td>\n",
       "      <td>0.839600</td>\n",
       "      <td>0.823202</td>\n",
       "      <td>0.830635</td>\n",
       "      <td>0.878301</td>\n",
       "      <td>0.946323</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56 rows × 56 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0         1         2         3         4         5         6  \\\n",
       "0   0.946766  0.784484  0.752536  0.748091  0.855079  0.835843  0.785578   \n",
       "1   0.784484  0.962751  0.739544  0.739005  0.783708  0.775224  0.799280   \n",
       "2   0.752536  0.739544  0.943797  0.852779  0.748857  0.743702  0.744241   \n",
       "3   0.748091  0.739005  0.852779  0.927736  0.746276  0.741655  0.741183   \n",
       "4   0.855079  0.783708  0.748857  0.746276  0.940973  0.838161  0.783518   \n",
       "5   0.835843  0.775224  0.743702  0.741655  0.838161  0.918365  0.778696   \n",
       "6   0.785578  0.799280  0.744241  0.741183  0.783518  0.778696  0.935086   \n",
       "7   0.804805  0.837051  0.754887  0.745799  0.806429  0.791158  0.840219   \n",
       "8   0.685704  0.688697  0.691477  0.690116  0.685338  0.684342  0.685485   \n",
       "9   0.683377  0.693427  0.691642  0.691273  0.690110  0.687866  0.695672   \n",
       "10  0.686960  0.694550  0.694589  0.692494  0.689357  0.688192  0.691176   \n",
       "11  0.746761  0.747820  0.776813  0.767814  0.744930  0.741248  0.745283   \n",
       "12  0.689367  0.696623  0.692582  0.693592  0.687156  0.682622  0.687031   \n",
       "13  0.738975  0.737661  0.770044  0.766586  0.742682  0.739870  0.734704   \n",
       "14  0.729346  0.726935  0.729278  0.739978  0.730523  0.729589  0.721662   \n",
       "15  0.699031  0.700607  0.705223  0.703808  0.702863  0.700715  0.694441   \n",
       "16  0.787265  0.862260  0.735654  0.735596  0.786371  0.782351  0.793290   \n",
       "17  0.785883  0.850610  0.737231  0.732695  0.781559  0.775408  0.797154   \n",
       "18  0.744493  0.736539  0.763029  0.762969  0.737843  0.741652  0.730041   \n",
       "19  0.739538  0.730021  0.756737  0.754560  0.730421  0.731175  0.726913   \n",
       "20  0.725886  0.721147  0.722897  0.730430  0.723369  0.723808  0.717401   \n",
       "21  0.729812  0.727624  0.725639  0.736433  0.731806  0.728020  0.722257   \n",
       "22  0.688081  0.693663  0.689697  0.695544  0.691301  0.689248  0.689314   \n",
       "23  0.687456  0.692161  0.691181  0.692000  0.689011  0.686394  0.687625   \n",
       "24  0.681035  0.690320  0.687548  0.690133  0.684358  0.683188  0.683958   \n",
       "25  0.714099  0.711805  0.725097  0.721855  0.714047  0.705997  0.717345   \n",
       "26  0.704652  0.707612  0.709023  0.712349  0.703589  0.701030  0.707807   \n",
       "27  0.684815  0.689857  0.687512  0.689807  0.683009  0.684797  0.687056   \n",
       "28  0.706640  0.714567  0.718418  0.715357  0.711410  0.701832  0.714158   \n",
       "29  0.720643  0.715139  0.724594  0.719920  0.712772  0.705993  0.716913   \n",
       "30  0.702260  0.698761  0.705390  0.709909  0.705661  0.699265  0.696191   \n",
       "31  0.692094  0.690951  0.699184  0.705660  0.696749  0.695473  0.696026   \n",
       "32  0.701636  0.705187  0.708791  0.707977  0.700041  0.698274  0.699800   \n",
       "33  0.700601  0.692580  0.699437  0.701523  0.697640  0.693573  0.696412   \n",
       "34  0.707263  0.696748  0.711143  0.709049  0.700812  0.700207  0.700826   \n",
       "35  0.681399  0.690096  0.685056  0.685519  0.681712  0.684156  0.683094   \n",
       "36  0.703307  0.702821  0.701724  0.703205  0.705374  0.696667  0.706469   \n",
       "37  0.707554  0.700339  0.703640  0.704405  0.704596  0.699038  0.704195   \n",
       "38  0.692182  0.692357  0.691755  0.687601  0.685918  0.687281  0.686365   \n",
       "39  0.688785  0.690875  0.693429  0.693209  0.689832  0.693138  0.689112   \n",
       "40  0.692229  0.690079  0.690717  0.694002  0.688640  0.688014  0.681660   \n",
       "41  0.693813  0.694363  0.692755  0.691739  0.687865  0.688573  0.690376   \n",
       "42  0.685243  0.690668  0.685323  0.687536  0.686926  0.685685  0.685622   \n",
       "43  0.694135  0.687031  0.690617  0.691779  0.691994  0.690121  0.685878   \n",
       "44  0.690455  0.689454  0.692662  0.690318  0.691465  0.686981  0.686985   \n",
       "45  0.692627  0.688207  0.696655  0.695026  0.689197  0.686827  0.686675   \n",
       "46  0.691221  0.689997  0.687751  0.691283  0.693681  0.693449  0.682206   \n",
       "47  0.695433  0.690206  0.694049  0.693461  0.691866  0.693463  0.688386   \n",
       "48  0.693244  0.692046  0.694393  0.690232  0.688448  0.697271  0.691809   \n",
       "49  0.690111  0.687958  0.694040  0.692321  0.687570  0.689534  0.682970   \n",
       "50  0.694085  0.694247  0.691360  0.691567  0.688384  0.688879  0.688128   \n",
       "51  0.679390  0.689242  0.683802  0.683870  0.682856  0.681259  0.682887   \n",
       "52  0.685788  0.689887  0.693029  0.691567  0.685787  0.681684  0.687947   \n",
       "53  0.688496  0.688691  0.688316  0.689718  0.686444  0.683977  0.683822   \n",
       "54  0.684604  0.689435  0.685412  0.687169  0.685081  0.684534  0.683033   \n",
       "55  0.685460  0.688622  0.689313  0.690127  0.681831  0.685016  0.685883   \n",
       "\n",
       "           7         8         9  ...        46        47        48        49  \\\n",
       "0   0.804805  0.685704  0.683377  ...  0.691221  0.695433  0.693244  0.690111   \n",
       "1   0.837051  0.688697  0.693427  ...  0.689997  0.690206  0.692046  0.687958   \n",
       "2   0.754887  0.691477  0.691642  ...  0.687751  0.694049  0.694393  0.694040   \n",
       "3   0.745799  0.690116  0.691273  ...  0.691283  0.693461  0.690232  0.692321   \n",
       "4   0.806429  0.685338  0.690110  ...  0.693681  0.691866  0.688448  0.687570   \n",
       "5   0.791158  0.684342  0.687866  ...  0.693449  0.693463  0.697271  0.689534   \n",
       "6   0.840219  0.685485  0.695672  ...  0.682206  0.688386  0.691809  0.682970   \n",
       "7   0.978459  0.684914  0.692240  ...  0.691860  0.695408  0.691224  0.687652   \n",
       "8   0.684914  0.980630  0.794265  ...  0.694188  0.699058  0.695524  0.701246   \n",
       "9   0.692240  0.794265  0.968273  ...  0.707710  0.705450  0.699447  0.704009   \n",
       "10  0.691728  0.792825  0.847368  ...  0.707091  0.703671  0.703952  0.707958   \n",
       "11  0.748863  0.688190  0.687629  ...  0.696358  0.700379  0.698689  0.696961   \n",
       "12  0.690567  0.836459  0.810842  ...  0.704498  0.705530  0.696778  0.705075   \n",
       "13  0.746967  0.685760  0.690778  ...  0.698059  0.692023  0.697127  0.692961   \n",
       "14  0.721272  0.688738  0.690219  ...  0.697438  0.697647  0.695963  0.701424   \n",
       "15  0.701940  0.685355  0.690034  ...  0.689627  0.691342  0.685565  0.695097   \n",
       "16  0.840812  0.691302  0.694984  ...  0.691189  0.687254  0.692511  0.687656   \n",
       "17  0.827275  0.689083  0.692608  ...  0.690803  0.684944  0.681766  0.685216   \n",
       "18  0.741371  0.687557  0.693474  ...  0.689003  0.685504  0.694887  0.689427   \n",
       "19  0.741969  0.683686  0.683625  ...  0.686078  0.690045  0.686927  0.692093   \n",
       "20  0.723879  0.682195  0.688072  ...  0.694673  0.694982  0.687648  0.691892   \n",
       "21  0.731727  0.687446  0.687051  ...  0.690380  0.695031  0.686062  0.691193   \n",
       "22  0.694659  0.823860  0.802935  ...  0.697877  0.701213  0.695666  0.704055   \n",
       "23  0.691858  0.834804  0.810995  ...  0.700565  0.703602  0.693407  0.700896   \n",
       "24  0.683327  0.818352  0.774057  ...  0.694224  0.694599  0.690264  0.699980   \n",
       "25  0.717395  0.688709  0.691080  ...  0.684539  0.688905  0.687441  0.689074   \n",
       "26  0.709802  0.690794  0.691121  ...  0.685254  0.686822  0.683413  0.687667   \n",
       "27  0.687818  0.818804  0.788277  ...  0.697725  0.701389  0.693778  0.700816   \n",
       "28  0.724998  0.682503  0.690065  ...  0.686010  0.684952  0.691400  0.684625   \n",
       "29  0.723758  0.692098  0.691069  ...  0.688618  0.690367  0.690208  0.691416   \n",
       "30  0.706438  0.684854  0.683758  ...  0.678749  0.685358  0.677763  0.686263   \n",
       "31  0.699360  0.680619  0.693531  ...  0.678996  0.678189  0.680845  0.681603   \n",
       "32  0.703497  0.683888  0.690808  ...  0.683789  0.684805  0.681069  0.688336   \n",
       "33  0.702463  0.688935  0.688403  ...  0.692649  0.694007  0.683007  0.688940   \n",
       "34  0.702936  0.693789  0.699791  ...  0.699269  0.699436  0.694390  0.698000   \n",
       "35  0.689466  0.819249  0.785813  ...  0.696249  0.699174  0.691973  0.702660   \n",
       "36  0.709612  0.686685  0.687720  ...  0.680832  0.682901  0.684193  0.689249   \n",
       "37  0.703607  0.681905  0.685667  ...  0.682304  0.691095  0.687009  0.690618   \n",
       "38  0.690050  0.727783  0.731722  ...  0.701411  0.700268  0.697766  0.699078   \n",
       "39  0.689911  0.741501  0.747157  ...  0.702323  0.705508  0.700474  0.702111   \n",
       "40  0.685623  0.740425  0.736576  ...  0.701457  0.705491  0.696564  0.702807   \n",
       "41  0.690167  0.740359  0.741916  ...  0.704081  0.709053  0.702169  0.705012   \n",
       "42  0.688523  0.704345  0.700841  ...  0.728785  0.735986  0.729591  0.731898   \n",
       "43  0.688076  0.704280  0.700512  ...  0.727824  0.728051  0.725200  0.738422   \n",
       "44  0.688102  0.707650  0.698083  ...  0.728002  0.735047  0.736771  0.740318   \n",
       "45  0.693425  0.698102  0.700431  ...  0.771901  0.771651  0.756025  0.769104   \n",
       "46  0.691860  0.694188  0.707710  ...  0.973500  0.825117  0.776492  0.788097   \n",
       "47  0.695408  0.699058  0.705450  ...  0.825117  0.946274  0.774923  0.785969   \n",
       "48  0.691224  0.695524  0.699447  ...  0.776492  0.774923  0.971693  0.795111   \n",
       "49  0.687652  0.701246  0.704009  ...  0.788097  0.785969  0.795111  0.909894   \n",
       "50  0.689009  0.703924  0.702509  ...  0.789969  0.789483  0.807809  0.841370   \n",
       "51  0.689360  0.824791  0.792738  ...  0.698240  0.700103  0.696179  0.700550   \n",
       "52  0.689880  0.860579  0.799169  ...  0.692139  0.699019  0.690735  0.699980   \n",
       "53  0.690392  0.891979  0.803361  ...  0.698310  0.703583  0.697678  0.702556   \n",
       "54  0.690296  0.824422  0.793323  ...  0.698962  0.696831  0.691569  0.698905   \n",
       "55  0.686976  0.820957  0.787099  ...  0.701215  0.700532  0.693872  0.699560   \n",
       "\n",
       "          50        51        52        53        54        55  \n",
       "0   0.694085  0.679390  0.685788  0.688496  0.684604  0.685460  \n",
       "1   0.694247  0.689242  0.689887  0.688691  0.689435  0.688622  \n",
       "2   0.691360  0.683802  0.693029  0.688316  0.685412  0.689313  \n",
       "3   0.691567  0.683870  0.691567  0.689718  0.687169  0.690127  \n",
       "4   0.688384  0.682856  0.685787  0.686444  0.685081  0.681831  \n",
       "5   0.688879  0.681259  0.681684  0.683977  0.684534  0.685016  \n",
       "6   0.688128  0.682887  0.687947  0.683822  0.683033  0.685883  \n",
       "7   0.689009  0.689360  0.689880  0.690392  0.690296  0.686976  \n",
       "8   0.703924  0.824791  0.860579  0.891979  0.824422  0.820957  \n",
       "9   0.702509  0.792738  0.799169  0.803361  0.793323  0.787099  \n",
       "10  0.700062  0.781437  0.791131  0.795788  0.781567  0.781458  \n",
       "11  0.695355  0.682961  0.690327  0.685242  0.689469  0.687254  \n",
       "12  0.709342  0.806203  0.843295  0.851791  0.808390  0.809583  \n",
       "13  0.693790  0.687649  0.685804  0.686139  0.685366  0.684207  \n",
       "14  0.695834  0.685959  0.688320  0.686133  0.684935  0.685964  \n",
       "15  0.693885  0.686691  0.687956  0.687628  0.685799  0.685002  \n",
       "16  0.691336  0.689206  0.687066  0.688369  0.691705  0.687898  \n",
       "17  0.687815  0.686353  0.686374  0.684489  0.694876  0.686506  \n",
       "18  0.688593  0.683211  0.679507  0.681351  0.683566  0.684728  \n",
       "19  0.692752  0.683864  0.682368  0.685348  0.685245  0.682432  \n",
       "20  0.695720  0.681714  0.679431  0.681809  0.681251  0.682600  \n",
       "21  0.691277  0.684785  0.691776  0.692421  0.684997  0.684389  \n",
       "22  0.704118  0.804665  0.837047  0.836126  0.816541  0.807831  \n",
       "23  0.701608  0.815719  0.842639  0.847676  0.818929  0.815907  \n",
       "24  0.696235  0.840703  0.822159  0.821998  0.852339  0.854867  \n",
       "25  0.697398  0.686018  0.689500  0.693489  0.685827  0.692284  \n",
       "26  0.689050  0.692592  0.694898  0.693303  0.695722  0.692336  \n",
       "27  0.703718  0.835858  0.821228  0.830261  0.876224  0.925893  \n",
       "28  0.689600  0.681713  0.682932  0.685321  0.681082  0.679388  \n",
       "29  0.694258  0.687068  0.694342  0.692729  0.690266  0.691319  \n",
       "30  0.686510  0.689406  0.691110  0.687476  0.687762  0.686205  \n",
       "31  0.682485  0.683838  0.690910  0.685294  0.685084  0.682508  \n",
       "32  0.689574  0.686909  0.689174  0.687942  0.689297  0.685733  \n",
       "33  0.690106  0.687514  0.693279  0.688500  0.689263  0.687698  \n",
       "34  0.698361  0.695674  0.697508  0.693083  0.694169  0.693324  \n",
       "35  0.698318  0.866762  0.830911  0.834396  0.845400  0.832100  \n",
       "36  0.687465  0.687267  0.690843  0.685676  0.685183  0.684767  \n",
       "37  0.693596  0.685791  0.687276  0.679027  0.681728  0.679890  \n",
       "38  0.702562  0.726067  0.729644  0.729324  0.724008  0.725392  \n",
       "39  0.703942  0.736845  0.744486  0.740859  0.731997  0.734582  \n",
       "40  0.705436  0.730739  0.737616  0.738156  0.731751  0.729335  \n",
       "41  0.709534  0.732731  0.741406  0.740407  0.733348  0.727890  \n",
       "42  0.739419  0.704444  0.706921  0.703559  0.699184  0.695467  \n",
       "43  0.738883  0.698577  0.692087  0.701419  0.701729  0.699648  \n",
       "44  0.745256  0.698964  0.705133  0.704192  0.699298  0.698154  \n",
       "45  0.775018  0.693354  0.698197  0.699042  0.697305  0.698650  \n",
       "46  0.789969  0.698240  0.692139  0.698310  0.698962  0.701215  \n",
       "47  0.789483  0.700103  0.699019  0.703583  0.696831  0.700532  \n",
       "48  0.807809  0.696179  0.690735  0.697678  0.691569  0.693872  \n",
       "49  0.841370  0.700550  0.699980  0.702556  0.698905  0.699560  \n",
       "50  0.947368  0.698199  0.698687  0.703544  0.702787  0.704523  \n",
       "51  0.698199  0.963780  0.826024  0.830656  0.849111  0.839600  \n",
       "52  0.698687  0.826024  0.939680  0.878461  0.827197  0.823202  \n",
       "53  0.703544  0.830656  0.878461  0.941551  0.836026  0.830635  \n",
       "54  0.702787  0.849111  0.827197  0.836026  0.984710  0.878301  \n",
       "55  0.704523  0.839600  0.823202  0.830635  0.878301  0.946323  \n",
       "\n",
       "[56 rows x 56 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv('Results/Pairwise_Segment_4.csv', sep = ',', na_filter = False, header = 0, index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "finished-charge",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<function __main__.sample_difference(cluster, genome, segment, proc=8, n=-1)>"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_difference #> 0.9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "designing-chuck",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'Series' object has no attribute 'join'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-74-b0b280644ec0>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mcluster\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'>KX101146'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenome\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/Masterthesis/lib/python3.8/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5460\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5461\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5462\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5463\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5464\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'Series' object has no attribute 'join'"
     ]
    }
   ],
   "source": [
    "cluster.loc['>KX101146'].join(genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "specific-poster",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = upload.loc[cluster.query('cluster == 9').index.values.tolist()]['genome'].to_frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "documentary-equity",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1710., 1613., 1607.],\n",
       "       [1613., 1698., 1643.],\n",
       "       [1607., 1643., 1698.]])"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ssd.cdist(x, x, metric = pairmsa2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "marked-setup",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>segment</th>\n",
       "      <th>protein</th>\n",
       "      <th>genus</th>\n",
       "      <th>subtype</th>\n",
       "      <th>date</th>\n",
       "      <th>host</th>\n",
       "      <th>curation</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>&gt;CY076093</th>\n",
       "      <td>A/American_coot/California/20181_006/2007</td>\n",
       "      <td>4</td>\n",
       "      <td>HA</td>\n",
       "      <td>A</td>\n",
       "      <td>H10N3</td>\n",
       "      <td>03/02/2007</td>\n",
       "      <td>American_Coot</td>\n",
       "      <td>Pass</td>\n",
       "      <td>ATGTACAAAATAGTACTAGTACTTGCGCTCCTTGGAGCGGTGCATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;CY097159</th>\n",
       "      <td>A/American_coot/Illinois/3405/2009</td>\n",
       "      <td>4</td>\n",
       "      <td>HA</td>\n",
       "      <td>A</td>\n",
       "      <td>H10N3</td>\n",
       "      <td>10/31/2009</td>\n",
       "      <td>Coot</td>\n",
       "      <td>Pass</td>\n",
       "      <td>ATGTACAAAATAGTACTAGTACTTGCGCTCCTTGGAGCGGTGCATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;CY079404</th>\n",
       "      <td>A/American_coot/Mississippi/09OS615/2009</td>\n",
       "      <td>4</td>\n",
       "      <td>HA</td>\n",
       "      <td>A</td>\n",
       "      <td>H10N3</td>\n",
       "      <td>01/24/2009</td>\n",
       "      <td>Avian</td>\n",
       "      <td>Pass</td>\n",
       "      <td>ATGTACAAAATAGTACTAGTACTTGCGCTTCTTGGAGCGGTGCATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;KY561267</th>\n",
       "      <td>A/American_coot/Missouri/15OS6556/2015</td>\n",
       "      <td>4</td>\n",
       "      <td>HA</td>\n",
       "      <td>A</td>\n",
       "      <td>H10N2</td>\n",
       "      <td>10/31/2015</td>\n",
       "      <td>Coot</td>\n",
       "      <td>Pass</td>\n",
       "      <td>ATGTACAAAGTAGTGCTAGTACTTGCGCTCCTTGGAGCGGTGCATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;CY076165</th>\n",
       "      <td>A/American_coot/Oregon/20589_007/2007</td>\n",
       "      <td>4</td>\n",
       "      <td>HA</td>\n",
       "      <td>A</td>\n",
       "      <td>H3N8</td>\n",
       "      <td>12/17/2007</td>\n",
       "      <td>Avian</td>\n",
       "      <td>Pass</td>\n",
       "      <td>ATAATTCTATTAGACATGAAGGCAATCATTGCACTAAGCTATTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;CY185401</th>\n",
       "      <td>A/common_coot/Republic_of_Georgia/1/2010</td>\n",
       "      <td>4</td>\n",
       "      <td>HA</td>\n",
       "      <td>A</td>\n",
       "      <td>H6N2</td>\n",
       "      <td>01/21/2010</td>\n",
       "      <td>Common_Coot</td>\n",
       "      <td>Pass</td>\n",
       "      <td>GGAAAATGATTGCAATTGTTGTAATAGCAATACTGACAGTAGCCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;CY004515</th>\n",
       "      <td>A/coot/ALB/134/1987</td>\n",
       "      <td>4</td>\n",
       "      <td>HA</td>\n",
       "      <td>A</td>\n",
       "      <td>H6N2</td>\n",
       "      <td>08/13/1987</td>\n",
       "      <td>Common_Coot</td>\n",
       "      <td>Pass</td>\n",
       "      <td>AGCAAAAGCAGGGGAAAATGATTGCAATCATAATACTTGCAATAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;CY136646</th>\n",
       "      <td>A/coot/Alberta/134/1987</td>\n",
       "      <td>4</td>\n",
       "      <td>HA</td>\n",
       "      <td>A</td>\n",
       "      <td>H6N2</td>\n",
       "      <td>08/13/1987</td>\n",
       "      <td>Common_Coot</td>\n",
       "      <td>Pass</td>\n",
       "      <td>GGAAAATGATTGCAATCATAATACTTGCAATAGTGGTCTCTACCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;LC318848</th>\n",
       "      <td>A/coot/Iwate/13/2016</td>\n",
       "      <td>4</td>\n",
       "      <td>HA</td>\n",
       "      <td>A</td>\n",
       "      <td>H5N6</td>\n",
       "      <td>12/22/2016</td>\n",
       "      <td>Coot</td>\n",
       "      <td>Pass</td>\n",
       "      <td>AGCAGGGGTTCAATCTGTCAAAATGGAGAAAATAGTGCTTCTTCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;LC275013</th>\n",
       "      <td>A/coot/Shiga/2501T010/2017</td>\n",
       "      <td>4</td>\n",
       "      <td>HA</td>\n",
       "      <td>A</td>\n",
       "      <td>H5N6</td>\n",
       "      <td>01/04/2017</td>\n",
       "      <td>Coot</td>\n",
       "      <td>Pass</td>\n",
       "      <td>TTCAATCTGTCAAAATGGAGAAAATAGTGCTTCTTCTTGCAGTGGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;CY028243</th>\n",
       "      <td>A/Eurasian_coot/Western_Australia/2727/1979</td>\n",
       "      <td>4</td>\n",
       "      <td>HA</td>\n",
       "      <td>A</td>\n",
       "      <td>H6N2</td>\n",
       "      <td>1979</td>\n",
       "      <td>Eurasian_Coot</td>\n",
       "      <td>Pass</td>\n",
       "      <td>AATGATTGCAATCATAATATTCGCGATAGTGGCCTCTACCAGCAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;KX101146</th>\n",
       "      <td>A/red_fronted_coot/Chile/5/2013</td>\n",
       "      <td>4</td>\n",
       "      <td>HA</td>\n",
       "      <td>A</td>\n",
       "      <td>H3N6</td>\n",
       "      <td>10/24/2013</td>\n",
       "      <td>Coot</td>\n",
       "      <td>Pass</td>\n",
       "      <td>ATGAAGACAATCACTGCTCTGCTGCTATTTGCCCATTGGGCCTACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;MH675632</th>\n",
       "      <td>A/red_gartered_coot/Chile/C16030/2016</td>\n",
       "      <td>4</td>\n",
       "      <td>HA</td>\n",
       "      <td>A</td>\n",
       "      <td>H3N4</td>\n",
       "      <td>09/22/2016</td>\n",
       "      <td>Coot</td>\n",
       "      <td>Pass</td>\n",
       "      <td>ATGAAGACAATCACTGCTCTGGTGCTACTTGCCCATTGGGCCTACA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                strain  segment protein genus  \\\n",
       "accession                                                                       \n",
       ">CY076093    A/American_coot/California/20181_006/2007        4      HA     A   \n",
       ">CY097159           A/American_coot/Illinois/3405/2009        4      HA     A   \n",
       ">CY079404     A/American_coot/Mississippi/09OS615/2009        4      HA     A   \n",
       ">KY561267       A/American_coot/Missouri/15OS6556/2015        4      HA     A   \n",
       ">CY076165        A/American_coot/Oregon/20589_007/2007        4      HA     A   \n",
       ">CY185401     A/common_coot/Republic_of_Georgia/1/2010        4      HA     A   \n",
       ">CY004515                          A/coot/ALB/134/1987        4      HA     A   \n",
       ">CY136646                      A/coot/Alberta/134/1987        4      HA     A   \n",
       ">LC318848                         A/coot/Iwate/13/2016        4      HA     A   \n",
       ">LC275013                   A/coot/Shiga/2501T010/2017        4      HA     A   \n",
       ">CY028243  A/Eurasian_coot/Western_Australia/2727/1979        4      HA     A   \n",
       ">KX101146              A/red_fronted_coot/Chile/5/2013        4      HA     A   \n",
       ">MH675632        A/red_gartered_coot/Chile/C16030/2016        4      HA     A   \n",
       "\n",
       "          subtype        date           host curation  \\\n",
       "accession                                               \n",
       ">CY076093   H10N3  03/02/2007  American_Coot     Pass   \n",
       ">CY097159   H10N3  10/31/2009           Coot     Pass   \n",
       ">CY079404   H10N3  01/24/2009          Avian     Pass   \n",
       ">KY561267   H10N2  10/31/2015           Coot     Pass   \n",
       ">CY076165    H3N8  12/17/2007          Avian     Pass   \n",
       ">CY185401    H6N2  01/21/2010    Common_Coot     Pass   \n",
       ">CY004515    H6N2  08/13/1987    Common_Coot     Pass   \n",
       ">CY136646    H6N2  08/13/1987    Common_Coot     Pass   \n",
       ">LC318848    H5N6  12/22/2016           Coot     Pass   \n",
       ">LC275013    H5N6  01/04/2017           Coot     Pass   \n",
       ">CY028243    H6N2        1979  Eurasian_Coot     Pass   \n",
       ">KX101146    H3N6  10/24/2013           Coot     Pass   \n",
       ">MH675632    H3N4  09/22/2016           Coot     Pass   \n",
       "\n",
       "                                                      genome  \n",
       "accession                                                     \n",
       ">CY076093  ATGTACAAAATAGTACTAGTACTTGCGCTCCTTGGAGCGGTGCATG...  \n",
       ">CY097159  ATGTACAAAATAGTACTAGTACTTGCGCTCCTTGGAGCGGTGCATG...  \n",
       ">CY079404  ATGTACAAAATAGTACTAGTACTTGCGCTTCTTGGAGCGGTGCATG...  \n",
       ">KY561267  ATGTACAAAGTAGTGCTAGTACTTGCGCTCCTTGGAGCGGTGCATG...  \n",
       ">CY076165  ATAATTCTATTAGACATGAAGGCAATCATTGCACTAAGCTATTTTT...  \n",
       ">CY185401  GGAAAATGATTGCAATTGTTGTAATAGCAATACTGACAGTAGCCGG...  \n",
       ">CY004515  AGCAAAAGCAGGGGAAAATGATTGCAATCATAATACTTGCAATAGT...  \n",
       ">CY136646  GGAAAATGATTGCAATCATAATACTTGCAATAGTGGTCTCTACCAG...  \n",
       ">LC318848  AGCAGGGGTTCAATCTGTCAAAATGGAGAAAATAGTGCTTCTTCTT...  \n",
       ">LC275013  TTCAATCTGTCAAAATGGAGAAAATAGTGCTTCTTCTTGCAGTGGT...  \n",
       ">CY028243  AATGATTGCAATCATAATATTCGCGATAGTGGCCTCTACCAGCAAA...  \n",
       ">KX101146  ATGAAGACAATCACTGCTCTGCTGCTATTTGCCCATTGGGCCTACA...  \n",
       ">MH675632  ATGAAGACAATCACTGCTCTGGTGCTACTTGCCCATTGGGCCTACA...  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "upload.query('segment == 4 & strain.str.contains(\"coot\")')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "unlikely-packet",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>segment</th>\n",
       "      <th>cluster</th>\n",
       "      <th>H</th>\n",
       "      <th>centroid</th>\n",
       "      <th>N</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>&gt;MK071380</th>\n",
       "      <td>4</td>\n",
       "      <td>43</td>\n",
       "      <td>H10</td>\n",
       "      <td>False</td>\n",
       "      <td>N7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;MK071396</th>\n",
       "      <td>4</td>\n",
       "      <td>43</td>\n",
       "      <td>H10</td>\n",
       "      <td>False</td>\n",
       "      <td>N7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;MH498655</th>\n",
       "      <td>4</td>\n",
       "      <td>43</td>\n",
       "      <td>H10</td>\n",
       "      <td>True</td>\n",
       "      <td>N7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;MH498733</th>\n",
       "      <td>4</td>\n",
       "      <td>43</td>\n",
       "      <td>H10</td>\n",
       "      <td>False</td>\n",
       "      <td>N7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;MH498637</th>\n",
       "      <td>4</td>\n",
       "      <td>43</td>\n",
       "      <td>H10</td>\n",
       "      <td>False</td>\n",
       "      <td>N7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           segment  cluster    H  centroid   N\n",
       "accession                                     \n",
       ">MK071380        4       43  H10     False  N7\n",
       ">MK071396        4       43  H10     False  N7\n",
       ">MH498655        4       43  H10      True  N7\n",
       ">MH498733        4       43  H10     False  N7\n",
       ">MH498637        4       43  H10     False  N7"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster.query('cluster == 43')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "damaged-immunology",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATGAAGACAATCACTGCTCTGCTGCTATTTGCCCATTGGGCCTACAGTCAAAACCCAATCAATGGCAATGAAACAGCCACACTGTGTTTGGGACACCATGCAGTGGCAAATGGAACGCTAGTGAAAACGATAACCGATGATCAGATCGAAGTAACCAATGCCACGGAACTGGTCCAGAGCACTTCAACAGGGAAGATATGTAACAACCCACACAGAGTCTTGGATGGGAGAGATTGCACTTTGATAGATGCGATGCTAGGAGACCCCCATTGTGATGTCTTCCAGGGTGAAAATTGGGACCTTTTCGTAGAACGAAGCAGTGCCTTCAGCAATTGCTACCCATATGACGTTCCTGATTATGCATCACTCCGGTCTCTTGTGGCATCCTCAGGGACATTAGAGTTCATAACAGAGGGTTTCACATGGACGGGTGTCACCCAAAACGGAGGAAGTAGCGCCTGCAAAAGAGGACCAGCAAATGGTTTCTTTAGCCGACTAAACTGGCTAACAAAATCTGGCAGTGCCTACCCCACACTGAATGTGACAATGCCAAACAATGACAATTTTGACAAACTGTACGTCTGGGGGGTTCATCATCCAAGTACAAACAATGAGCAGACAAGTCTATACGTCCAGGCGTCAGGCAGAATAACAGTCTCAACAAGAAGGACCCAGCAGACAATAATCCCAAATATCGGATCAAGACCATGGGTCAGAGGCCAATCAGGCAGGATAAGCATATACTGGACTATTGTGAAGCCTGGGGATGTCATTGTAATAAACAGCAATGGCAACTTAATTGCGCCGAGGGGATATTTCAAAATGCGAACAGGGAAAAGTTCTATAATGAGATCAGATGCACCCATAGACACTTGTATATCTGAATGCATTACACCGAATGGAAGCATTCCCAACAACAGGCCATTTCAAAATGTGAACAAGATCACATATGGAAACTGCCCCAAGTATGTTAAACAGAATACTTTAAAATTGGCCACTGGAATGAGGAACGTACCAGAGAAGCAAACCAGGGGACTCTTTGGGGCAATAGCAGGTTTTATAGAAAATGGTTGGCAAGGAATGATTGATGGATGGTATGGGTTTCGGCATCAGAATTCAGAAGGAACTGGGCAAGCTGCTGATCTTAAGAGTACCCAAGCGGCCATTGATCAGATCAATGGGAAATTGAACAGAGTCATTGAAAAGACTAATGAGAAATTCCATCAGATAGAGAAGGAATTTTCAGAAGTGGAAGGAAGGATCCAGGACCTGGAGAAATATGTAGAAGACACAAAGATAGACCTATGGTCCTACAATGCAGAGTTGTTAGTAGCCCTGGAGAACCAACACACGATTGACCTAACAGATGCAGAGATGAACAAACTGTTCGAGAAAACTAGAAGGCAATTAAGAGAAAATGCAGAAGACATGGGAAATGGATGTTTCAAAATTTATCACAAATGTGATAATGCATGCATTGAATCAATACGAAATGGAACATATGACCATGACATCTACAGGGACGAAGCATTAAACAACCGGTTTCAAATTAGAGGTGTCGAATTGAAATCAGGCTACAAAGATTGGATACTGTGGATCTCTTTTGCCATATCATGCTTCTTACTCTGTGTTGTTCTATTGGGCTTCATAATGTGGGCCTGCCAGAGAGGCAACATTAGGTGCAACATTTGCATTTAAGTACACTAATAA'"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x['genome'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "directed-rebel",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq1 = Seq(query_x)\n",
    "seq2 = Seq(query_y) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "patent-vector",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 25.3 ms, sys: 0 ns, total: 25.3 ms\n",
      "Wall time: 25.1 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "aligner = Align.PairwiseAligner()\n",
    "aligner.mode = 'global'\n",
    "alignment = aligner.align(seq1, seq2)[0]\n",
    "\n",
    "score = alignment.score/len(str(alignment).split('\\n')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "proud-financing",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1202.0"
      ]
     },
     "execution_count": 231,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alignment.score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "exact-banks",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1202.0"
      ]
     },
     "execution_count": 230,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "pairwise2.align.globalxx(seq1, seq2, score_only = True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "certain-aluminum",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Align.AlignInfo import SummaryInfo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "recognized-enterprise",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'PairwiseAlignment' object is not subscriptable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-158-98f21ae81b67>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mSummaryInfo\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0malignments\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minformation_content\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;31m#.get_column(1)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/Masterthesis/lib/python3.8/site-packages/Bio/Align/AlignInfo.py\u001b[0m in \u001b[0;36minformation_content\u001b[0;34m(self, start, end, e_freq_table, log_base, chars_to_ignore, pseudo_count)\u001b[0m\n\u001b[1;32m    341\u001b[0m         \u001b[0;31m# if no end was specified, then we default to the end of the sequence\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    342\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mend\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 343\u001b[0;31m             \u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0malignment\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    344\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mchars_to_ignore\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    345\u001b[0m             \u001b[0mchars_to_ignore\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: 'PairwiseAlignment' object is not subscriptable"
     ]
    }
   ],
   "source": [
    "SummaryInfo(alignments).information_content()#.get_column(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "simplified-parts",
   "metadata": {},
   "outputs": [],
   "source": [
    "proof = pd.read_csv('Results/Pairwise_Segment_6.csv', sep = ',', na_filter = True, header = 0, index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "drawn-talent",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_difference(proof, segment, outpath, render):\n",
    "\n",
    "    with sns.axes_style(\"darkgrid\"):\n",
    "        fig, ax = plt.subplots(figsize=(8,6))\n",
    "        ax = sns.heatmap(proof, linewidths=.5, vmin = 0, vmax = 1, center = 0.5,cmap=\"RdWtBu\")\n",
    "        plt.xlabel(\"Cluster\")\n",
    "        plt.ylabel(\"Cluster\")\n",
    "        plt.tight_layout()\n",
    "        plt.close(fig)\n",
    "        fig.savefig(outpath + 'Cluster_Difference_Segment_' + str(segment) + '.' + render)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "fabulous-progress",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "'RdWtBu' is not a valid value for name; supported values are 'Accent', 'Accent_r', 'Blues', 'Blues_r', 'BrBG', 'BrBG_r', 'BuGn', 'BuGn_r', 'BuPu', 'BuPu_r', 'CMRmap', 'CMRmap_r', 'Dark2', 'Dark2_r', 'GnBu', 'GnBu_r', 'Greens', 'Greens_r', 'Greys', 'Greys_r', 'OrRd', 'OrRd_r', 'Oranges', 'Oranges_r', 'PRGn', 'PRGn_r', 'Paired', 'Paired_r', 'Pastel1', 'Pastel1_r', 'Pastel2', 'Pastel2_r', 'PiYG', 'PiYG_r', 'PuBu', 'PuBuGn', 'PuBuGn_r', 'PuBu_r', 'PuOr', 'PuOr_r', 'PuRd', 'PuRd_r', 'Purples', 'Purples_r', 'RdBu', 'RdBu_r', 'RdGy', 'RdGy_r', 'RdPu', 'RdPu_r', 'RdYlBu', 'RdYlBu_r', 'RdYlGn', 'RdYlGn_r', 'Reds', 'Reds_r', 'Set1', 'Set1_r', 'Set2', 'Set2_r', 'Set3', 'Set3_r', 'Spectral', 'Spectral_r', 'Wistia', 'Wistia_r', 'YlGn', 'YlGnBu', 'YlGnBu_r', 'YlGn_r', 'YlOrBr', 'YlOrBr_r', 'YlOrRd', 'YlOrRd_r', 'afmhot', 'afmhot_r', 'autumn', 'autumn_r', 'binary', 'binary_r', 'bone', 'bone_r', 'brg', 'brg_r', 'bwr', 'bwr_r', 'cividis', 'cividis_r', 'cool', 'cool_r', 'coolwarm', 'coolwarm_r', 'copper', 'copper_r', 'crest', 'crest_r', 'cubehelix', 'cubehelix_r', 'flag', 'flag_r', 'flare', 'flare_r', 'gist_earth', 'gist_earth_r', 'gist_gray', 'gist_gray_r', 'gist_heat', 'gist_heat_r', 'gist_ncar', 'gist_ncar_r', 'gist_rainbow', 'gist_rainbow_r', 'gist_stern', 'gist_stern_r', 'gist_yarg', 'gist_yarg_r', 'gnuplot', 'gnuplot2', 'gnuplot2_r', 'gnuplot_r', 'gray', 'gray_r', 'hot', 'hot_r', 'hsv', 'hsv_r', 'icefire', 'icefire_r', 'inferno', 'inferno_r', 'jet', 'jet_r', 'magma', 'magma_r', 'mako', 'mako_r', 'nipy_spectral', 'nipy_spectral_r', 'ocean', 'ocean_r', 'pink', 'pink_r', 'plasma', 'plasma_r', 'prism', 'prism_r', 'rainbow', 'rainbow_r', 'rocket', 'rocket_r', 'seismic', 'seismic_r', 'spring', 'spring_r', 'summer', 'summer_r', 'tab10', 'tab10_r', 'tab20', 'tab20_r', 'tab20b', 'tab20b_r', 'tab20c', 'tab20c_r', 'terrain', 'terrain_r', 'turbo', 'turbo_r', 'twilight', 'twilight_r', 'twilight_shifted', 'twilight_shifted_r', 'viridis', 'viridis_r', 'vlag', 'vlag_r', 'winter', 'winter_r'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-22-ab86cbfbbdd4>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mplot_difference\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mproof\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m6\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Results/'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'pdf'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-21-b79358c3daa6>\u001b[0m in \u001b[0;36mplot_difference\u001b[0;34m(proof, segment, outpath, render)\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0;32mwith\u001b[0m \u001b[0msns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maxes_style\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"darkgrid\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m         \u001b[0mfig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0max\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubplots\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfigsize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m8\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m6\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m         \u001b[0max\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mheatmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mproof\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlinewidths\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m.5\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvmin\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvmax\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcenter\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0.5\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mcmap\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"RdWtBu\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m         \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mxlabel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Cluster\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m         \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mylabel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Cluster\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/newlab/lib/python3.8/site-packages/seaborn/_decorators.py\u001b[0m in \u001b[0;36minner_f\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     44\u001b[0m             )\n\u001b[1;32m     45\u001b[0m         \u001b[0mkwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0marg\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marg\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparameters\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 46\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     47\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0minner_f\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     48\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/newlab/lib/python3.8/site-packages/seaborn/matrix.py\u001b[0m in \u001b[0;36mheatmap\u001b[0;34m(data, vmin, vmax, cmap, center, robust, annot, fmt, annot_kws, linewidths, linecolor, cbar, cbar_kws, cbar_ax, square, xticklabels, yticklabels, mask, ax, **kwargs)\u001b[0m\n\u001b[1;32m    533\u001b[0m     \"\"\"\n\u001b[1;32m    534\u001b[0m     \u001b[0;31m# Initialize the plotter object\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 535\u001b[0;31m     plotter = _HeatMapper(data, vmin, vmax, cmap, center, robust, annot, fmt,\n\u001b[0m\u001b[1;32m    536\u001b[0m                           \u001b[0mannot_kws\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcbar\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcbar_kws\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mxticklabels\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    537\u001b[0m                           yticklabels, mask)\n",
      "\u001b[0;32m~/miniconda3/envs/newlab/lib/python3.8/site-packages/seaborn/matrix.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, data, vmin, vmax, cmap, center, robust, annot, fmt, annot_kws, cbar, cbar_kws, xticklabels, yticklabels, mask)\u001b[0m\n\u001b[1;32m    153\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    154\u001b[0m         \u001b[0;31m# Determine good default values for the colormapping\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 155\u001b[0;31m         self._determine_cmap_params(plot_data, vmin, vmax,\n\u001b[0m\u001b[1;32m    156\u001b[0m                                     cmap, center, robust)\n\u001b[1;32m    157\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/newlab/lib/python3.8/site-packages/seaborn/matrix.py\u001b[0m in \u001b[0;36m_determine_cmap_params\u001b[0;34m(self, plot_data, vmin, vmax, cmap, center, robust)\u001b[0m\n\u001b[1;32m    207\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcmap\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0micefire\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    208\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmap\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 209\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcmap\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmpl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_cmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmap\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    210\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmap\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    211\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcmap\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmpl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolors\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mListedColormap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmap\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/newlab/lib/python3.8/site-packages/matplotlib/cm.py\u001b[0m in \u001b[0;36mget_cmap\u001b[0;34m(name, lut)\u001b[0m\n\u001b[1;32m    202\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolors\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mColormap\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    203\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 204\u001b[0;31m     \u001b[0mcbook\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_check_in_list\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msorted\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_cmap_registry\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    205\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mlut\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    206\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0m_cmap_registry\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/newlab/lib/python3.8/site-packages/matplotlib/cbook/__init__.py\u001b[0m in \u001b[0;36m_check_in_list\u001b[0;34m(_values, **kwargs)\u001b[0m\n\u001b[1;32m   2264\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mkwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2265\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2266\u001b[0;31m             raise ValueError(\n\u001b[0m\u001b[1;32m   2267\u001b[0m                 \u001b[0;34m\"{!r} is not a valid value for {}; supported values are {}\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2268\u001b[0m                 .format(v, k, ', '.join(map(repr, values))))\n",
      "\u001b[0;31mValueError\u001b[0m: 'RdWtBu' is not a valid value for name; supported values are 'Accent', 'Accent_r', 'Blues', 'Blues_r', 'BrBG', 'BrBG_r', 'BuGn', 'BuGn_r', 'BuPu', 'BuPu_r', 'CMRmap', 'CMRmap_r', 'Dark2', 'Dark2_r', 'GnBu', 'GnBu_r', 'Greens', 'Greens_r', 'Greys', 'Greys_r', 'OrRd', 'OrRd_r', 'Oranges', 'Oranges_r', 'PRGn', 'PRGn_r', 'Paired', 'Paired_r', 'Pastel1', 'Pastel1_r', 'Pastel2', 'Pastel2_r', 'PiYG', 'PiYG_r', 'PuBu', 'PuBuGn', 'PuBuGn_r', 'PuBu_r', 'PuOr', 'PuOr_r', 'PuRd', 'PuRd_r', 'Purples', 'Purples_r', 'RdBu', 'RdBu_r', 'RdGy', 'RdGy_r', 'RdPu', 'RdPu_r', 'RdYlBu', 'RdYlBu_r', 'RdYlGn', 'RdYlGn_r', 'Reds', 'Reds_r', 'Set1', 'Set1_r', 'Set2', 'Set2_r', 'Set3', 'Set3_r', 'Spectral', 'Spectral_r', 'Wistia', 'Wistia_r', 'YlGn', 'YlGnBu', 'YlGnBu_r', 'YlGn_r', 'YlOrBr', 'YlOrBr_r', 'YlOrRd', 'YlOrRd_r', 'afmhot', 'afmhot_r', 'autumn', 'autumn_r', 'binary', 'binary_r', 'bone', 'bone_r', 'brg', 'brg_r', 'bwr', 'bwr_r', 'cividis', 'cividis_r', 'cool', 'cool_r', 'coolwarm', 'coolwarm_r', 'copper', 'copper_r', 'crest', 'crest_r', 'cubehelix', 'cubehelix_r', 'flag', 'flag_r', 'flare', 'flare_r', 'gist_earth', 'gist_earth_r', 'gist_gray', 'gist_gray_r', 'gist_heat', 'gist_heat_r', 'gist_ncar', 'gist_ncar_r', 'gist_rainbow', 'gist_rainbow_r', 'gist_stern', 'gist_stern_r', 'gist_yarg', 'gist_yarg_r', 'gnuplot', 'gnuplot2', 'gnuplot2_r', 'gnuplot_r', 'gray', 'gray_r', 'hot', 'hot_r', 'hsv', 'hsv_r', 'icefire', 'icefire_r', 'inferno', 'inferno_r', 'jet', 'jet_r', 'magma', 'magma_r', 'mako', 'mako_r', 'nipy_spectral', 'nipy_spectral_r', 'ocean', 'ocean_r', 'pink', 'pink_r', 'plasma', 'plasma_r', 'prism', 'prism_r', 'rainbow', 'rainbow_r', 'rocket', 'rocket_r', 'seismic', 'seismic_r', 'spring', 'spring_r', 'summer', 'summer_r', 'tab10', 'tab10_r', 'tab20', 'tab20_r', 'tab20b', 'tab20b_r', 'tab20c', 'tab20c_r', 'terrain', 'terrain_r', 'turbo', 'turbo_r', 'twilight', 'twilight_r', 'twilight_shifted', 'twilight_shifted_r', 'viridis', 'viridis_r', 'vlag', 'vlag_r', 'winter', 'winter_r'"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAesAAAFpCAYAAAC8iwByAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAATQklEQVR4nO3dX4jVdd7A8c+ZGYSGRE10TqIM4hp4YUnoRbDaMjYMyzQm6YVSLcQO0QbeiBcrtW7aH1jZvYwFG9g2d/PGRNRhI5pYh6CSlo1ZBWGtBmVjTijOWvZHZvw9F8+zPs88Wuc4es58dF4vCPrN+Z5zPnwI355z9FQqiqIIACCtpqkeAAD4YWINAMmJNQAkJ9YAkJxYA0ByYg0AyVWN9fbt2+OBBx6Ihx9++Jq3F0URL774YnR2dkZPT0+cOHHipg8JANNZ1Vg/+uij0dfX9723Dw4OxvDwcLz99tvxwgsvxPPPP38z5wOAaa9qrFetWhWzZs363tsHBgZi/fr1USqVYsWKFXHhwoX44osvbuqQADCd3fBn1pVKJcrl8pXrcrkclUrlRh8WAPgfLTf6ANf6ttJSqVTT/XzRaX2VSmHHDWDP9WfH9WfHjdHUVL2P13LDsS6XyzEyMnLlemRkJObPn1/1fkURce7cVzf69PyA2bNbY3T066ke47Znz/Vnx/Vnx40xb97MSd3vht8G7+joiIMHD0ZRFPHxxx/HzJkza4o1AFCbqq+st27dGseOHYvz58/HmjVrYsuWLTE2NhYREZs3b44HH3wwjh49Gp2dnXHHHXfEyy+/XPehAWA6KU3V/yLz8uXC2+B15m2txrDn+rPj+rPjxpiyt8EBgPoSawBITqwBIDmxBoDkxBoAkhNrAEhOrAEgObEGgOTEGgCSE2sASE6sASA5sQaA5MQaAJITawBITqwBIDmxBoDkxBoAkhNrAEhOrAEgObEGgOTEGgCSE2sASE6sASA5sQaA5MQaAJITawBITqwBIDmxBoDkxBoAkhNrAEhOrAEgObEGgOTEGgCSE2sASE6sASA5sQaA5MQaAJITawBITqwBIDmxBoDkxBoAkhNrAEhOrAEgObEGgOTEGgCSE2sASE6sASA5sQaA5MQaAJITawBITqwBIDmxBoDkxBoAkhNrAEhOrAEgObEGgORqivXg4GB0dXVFZ2dn7Nmz56rbv/zyy3j66adj3bp10d3dHW+++eZNHxQApquqsR4fH49du3ZFX19f9Pf3x5EjR+LUqVMTzvz5z3+OJUuWxKFDh2Lv3r3xm9/8Ji5dulS3oQFgOqka66GhoWhvb49FixbFjBkzoru7OwYGBiacKZVKcfHixSiKIi5evBizZs2KlpaWug0NANNJ1VhXKpUol8tXrtva2qJSqUw489hjj8Unn3wSq1evjnXr1sWzzz4bTU0+DgeAm6Hqy9+iKK76WalUmnD93nvvxbJly+L111+P06dPx5NPPhkrV66MO++883sft1SKmD27dRIjU6vm5iY7bgB7rj87rj87zq1qrMvlcoyMjFy5rlQqMX/+/AlnDhw4EE899VSUSqVob2+PhQsXxqeffhr33nvv9z5uUUSMjn59A6NTzezZrXbcAPZcf3Zcf3bcGPPmzZzU/aq+V718+fIYHh6OM2fOxKVLl6K/vz86OjomnLn77rvj/fffj4iIs2fPxmeffRYLFy6c1EAAwERVX1m3tLTEjh07ore3N8bHx2PDhg2xdOnS2LdvX0REbN68OZ555pnYvn179PT0RFEUsW3btrjrrrvqPjwATAel4lofSjfA5ctFnDv31VQ89bThba3GsOf6s+P6s+PGqNvb4ADA1BJrAEhOrAEgObEGgOTEGgCSE2sASE6sASA5sQaA5MQaAJITawBITqwBIDmxBoDkxBoAkhNrAEhOrAEgObEGgOTEGgCSE2sASE6sASA5sQaA5MQaAJITawBITqwBIDmxBoDkxBoAkhNrAEhOrAEgObEGgOTEGgCSE2sASE6sASA5sQaA5MQaAJITawBITqwBIDmxBoDkxBoAkhNrAEhOrAEgObEGgOTEGgCSE2sASE6sASA5sQaA5MQaAJITawBITqwBIDmxBoDkxBoAkhNrAEhOrAEgObEGgOTEGgCSE2sASE6sASA5sQaA5MQaAJKrKdaDg4PR1dUVnZ2dsWfPnmue+fDDD+ORRx6J7u7uePzxx2/qkAAwnbVUOzA+Ph67du2KP/zhD9HW1hYbN26Mjo6O+NGPfnTlzIULF2Lnzp3R19cXCxYsiHPnztV1aACYTqq+sh4aGor29vZYtGhRzJgxI7q7u2NgYGDCmcOHD0dnZ2csWLAgIiLmzp1bn2kBYBqq+sq6UqlEuVy+ct3W1hZDQ0MTzgwPD8fY2Fg88cQTcfHixfjZz34W69ev/8HHLZUiZs9undzU1KS5ucmOG8Ce68+O68+Oc6sa66IorvpZqVSacD0+Ph4nTpyI1157Lb799tvYtGlT3HfffbF48eIfeNyI0dGvJzEytZo9u9WOG8Ce68+O68+OG2PevJmTul/VWJfL5RgZGblyXalUYv78+VedmTNnTrS2tkZra2usXLkyTp48+YOxBgBqU/Uz6+XLl8fw8HCcOXMmLl26FP39/dHR0THhzNq1a+Ojjz6KsbGx+Oabb2JoaCiWLFlSt6EBYDqp+sq6paUlduzYEb29vTE+Ph4bNmyIpUuXxr59+yIiYvPmzbFkyZJYvXp1rFu3LpqammLjxo1xzz331H14AJgOSsW1PpRugMuXizh37qupeOppw2dQjWHP9WfH9WfHjTHZz6x9gxkAJCfWAJCcWANAcmINAMmJNQAkJ9YAkJxYA0ByYg0AyYk1ACQn1gCQnFgDQHJiDQDJiTUAJCfWAJCcWANAcmINAMmJNQAkJ9YAkJxYA0ByYg0AyYk1ACQn1gCQnFgDQHJiDQDJiTUAJCfWAJCcWANAcmINAMmJNQAkJ9YAkJxYA0ByYg0AyYk1ACQn1gCQnFgDQHJiDQDJiTUAJCfWAJCcWANAcmINAMmJNQAkJ9YAkJxYA0ByYg0AyYk1ACQn1gCQnFgDQHJiDQDJiTUAJCfWAJCcWANAcmINAMmJNQAkJ9YAkJxYA0ByYg0AydUU68HBwejq6orOzs7Ys2fP954bGhqKZcuWxVtvvXXTBgSA6a5qrMfHx2PXrl3R19cX/f39ceTIkTh16tQ1z/32t7+NH//4x3UZFACmq6qxHhoaivb29li0aFHMmDEjuru7Y2Bg4Kpze/fuja6urpg7d25dBgWA6aql2oFKpRLlcvnKdVtbWwwNDV115p133ok//vGP8Y9//KOmJy6VImbPbr3Ocbkezc1NdtwA9lx/dlx/dpxb1VgXRXHVz0ql0oTrl156KbZt2xbNzc01P3FRRIyOfl3zea7f7NmtdtwA9lx/dlx/dtwY8+bNnNT9qsa6XC7HyMjIletKpRLz58+fcOb48eOxdevWiIg4f/58HD16NFpaWuKhhx6a1FAAwP+qGuvly5fH8PBwnDlzJtra2qK/vz9+97vfTTjz7rvvXvn3X/7yl/GTn/xEqAHgJqka65aWltixY0f09vbG+Ph4bNiwIZYuXRr79u2LiIjNmzfXfUgAmM5KxbU+lG6Ay5eLOHfuq6l46mnDZ1CNYc/1Z8f1Z8eNMdnPrH2DGQAkJ9YAkJxYA0ByYg0AyYk1ACQn1gCQnFgDQHJiDQDJiTUAJCfWAJCcWANAcmINAMmJNQAkJ9YAkJxYA0ByYg0AyYk1ACQn1gCQnFgDQHJiDQDJiTUAJCfWAJCcWANAcmINAMmJNQAkJ9YAkJxYA0ByYg0AyYk1ACQn1gCQnFgDQHJiDQDJiTUAJCfWAJCcWANAcmINAMmJNQAkJ9YAkJxYA0ByYg0AyYk1ACQn1gCQnFgDQHJiDQDJiTUAJCfWAJCcWANAcmINAMmJNQAkJ9YAkJxYA0ByYg0AyYk1ACQn1gCQnFgDQHJiDQDJ1RTrwcHB6Orqis7OztizZ89Vtx86dCh6enqip6cnNm3aFCdPnrzpgwLAdFU11uPj47Fr167o6+uL/v7+OHLkSJw6dWrCmYULF8af/vSnOHz4cPziF7+IX/3qV3UbGACmm6qxHhoaivb29li0aFHMmDEjuru7Y2BgYMKZ+++/P2bNmhUREStWrIiRkZH6TAsA01DVWFcqlSiXy1eu29raolKpfO/5/fv3x5o1a27OdABAtFQ7UBTFVT8rlUrXPPvBBx/E/v3744033qj6xKVSxOzZrTWMyGQ1NzfZcQPYc/3Zcf3ZcW5VY10ulye8rV2pVGL+/PlXnTt58mQ899xz8eqrr8acOXOqPnFRRIyOfn2d43I9Zs9uteMGsOf6s+P6s+PGmDdv5qTuV/Vt8OXLl8fw8HCcOXMmLl26FP39/dHR0THhzOeffx5btmyJ3bt3x+LFiyc1CABwbVVfWbe0tMSOHTuit7c3xsfHY8OGDbF06dLYt29fRERs3rw5XnnllRgdHY2dO3dGRERzc3McOHCgvpMDwDRRKq71oXQDXL5cxLlzX03FU08b3tZqDHuuPzuuPztujLq9DQ4ATC2xBoDkxBoAkhNrAEhOrAEgObEGgOTEGgCSE2sASE6sASA5sQaA5MQaAJITawBITqwBIDmxBoDkxBoAkhNrAEhOrAEgObEGgOTEGgCSE2sASE6sASA5sQaA5MQaAJITawBITqwBIDmxBoDkxBoAkhNrAEhOrAEgObEGgOTEGgCSE2sASE6sASA5sQaA5MQaAJITawBITqwBIDmxBoDkxBoAkhNrAEhOrAEgObEGgOTEGgCSE2sASE6sASA5sQaA5MQaAJITawBITqwBIDmxBoDkxBoAkhNrAEhOrAEgObEGgOTEGgCSE2sASK6mWA8ODkZXV1d0dnbGnj17rrq9KIp48cUXo7OzM3p6euLEiRM3fVAAmK6qxnp8fDx27doVfX190d/fH0eOHIlTp05NODM4OBjDw8Px9ttvxwsvvBDPP/98veYFgGmnaqyHhoaivb09Fi1aFDNmzIju7u4YGBiYcGZgYCDWr18fpVIpVqxYERcuXIgvvviibkMDwHRSNdaVSiXK5fKV67a2tqhUKj94plwuX3UGAJiclmoHiqK46melUum6z/x/TU2lmDdvZrWn5wbZcWPYc/3Zcf3ZcV5VX1mXy+UYGRm5cl2pVGL+/Pk/eGZkZOSqMwDA5FSN9fLly2N4eDjOnDkTly5div7+/ujo6JhwpqOjIw4ePBhFUcTHH38cM2fOFGsAuEmqvg3e0tISO3bsiN7e3hgfH48NGzbE0qVLY9++fRERsXnz5njwwQfj6NGj0dnZGXfccUe8/PLLdR8cAKaLUnGtD5wBgDR8gxkAJCfWAJBc3WPtq0rrr9qODx06FD09PdHT0xObNm2KkydPTsGUt7ZqO/6PoaGhWLZsWbz11lsNnO72UcueP/zww3jkkUeiu7s7Hn/88QZPeOurtuMvv/wynn766Vi3bl10d3fHm2++OQVT3tq2b98eDzzwQDz88MPXvH1S3SvqaGxsrFi7dm1x+vTp4rvvvit6enqKf/7znxPO/PWvfy1+/vOfF5cvXy7+/ve/Fxs3bqznSLedWnb8t7/9rRgdHS2K4r/3bcfXp5Yd/+fcE088UfT29hZ/+ctfpmDSW1ste/73v/9d/PSnPy3+9a9/FUVRFGfPnp2KUW9Ztez497//fbF79+6iKIri3LlzxapVq4rvvvtuKsa9ZR07dqw4fvx40d3dfc3bJ9O9ur6y9lWl9VfLju+///6YNWtWRESsWLFiwt+Jp7padhwRsXfv3ujq6oq5c+dOwZS3vlr2fPjw4ejs7IwFCxZERNj1daplx6VSKS5evBhFUcTFixdj1qxZ0dJS9S8O8X+sWrXqyq+51zKZ7tU11r6qtP5q2fH/tX///lizZk0jRrtt1Prf8TvvvBObNm1q9Hi3jVr2PDw8HBcuXIgnnngiHn300Th48GCDp7y11bLjxx57LD755JNYvXp1rFu3Lp599tloavLHm26myXSvrr9dKur0VaX8r+vZ3wcffBD79++PN954o95j3VZq2fFLL70U27Zti+bm5kaNddupZc/j4+Nx4sSJeO211+Lbb7+NTZs2xX333ReLFy9u1Ji3tFp2/N5778WyZcvi9ddfj9OnT8eTTz4ZK1eujDvvvLNRY972JtO9usbaV5XWXy07jog4efJkPPfcc/Hqq6/GnDlzGjniLa+WHR8/fjy2bt0aERHnz5+Po0ePRktLSzz00EMNnfVWVuuvF3PmzInW1tZobW2NlStXxsmTJ8W6RrXs+MCBA/HUU09FqVSK9vb2WLhwYXz66adx7733Nnrc29ZkulfX9zZ8VWn91bLjzz//PLZs2RK7d+/2i9ok1LLjd99998o/XV1d8etf/1qor1Mte167dm189NFHMTY2Ft98800MDQ3FkiVLpmjiW08tO7777rvj/fffj4iIs2fPxmeffRYLFy6cinFvW5PpXl1fWfuq0vqrZcevvPJKjI6Oxs6dOyMiorm5OQ4cODCVY99SatkxN66WPS9ZsuTKZ6lNTU2xcePGuOeee6Z48ltHLTt+5plnYvv27dHT0xNFUcS2bdvirrvumuLJby1bt26NY8eOxfnz52PNmjWxZcuWGBsbi4jJd8/XjQJAcv6IHwAkJ9YAkJxYA0ByYg0AyYk1ACQn1gCQnFgDQHJiDQDJ/Re4GPjDYyCfHgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 576x432 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_difference(proof, 6, 'Results/', 'pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "amber-immunology",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<matplotlib.colors.LinearSegmentedColormap object at 0x7f38f9c2e6d0>\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "bibliographic-curve",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pairmsa(nuc1, nuc2):\n",
    "        \n",
    "    if nuc1[0] == nuc2[0]:\n",
    "        \n",
    "        return(np.nan)\n",
    "        \n",
    "    else:    \n",
    "        \n",
    "        seq1 = Seq(nuc1[1])\n",
    "        seq2 = Seq(nuc2[1]) \n",
    "\n",
    "        aligner = Align.PairwiseAligner()\n",
    "        #aligner.open_gap_score = -0.5\n",
    "        #aligner.extend_gap_score = -0.1\n",
    "        #aligner.target_end_gap_score = 0.0\n",
    "        #aligner.query_end_gap_score = 0.0\n",
    "        aligner.mode = 'global'\n",
    "\n",
    "        alignment = aligner.align(seq1, seq2)[0]\n",
    "        score = alignment.score/len(str(alignment).split('\\n')[0])\n",
    "\n",
    "        #alignments = pairwise2.align.globalxx(seq1, seq2, score_only = True) \n",
    "\n",
    "        return(score)\n",
    "\n",
    "\n",
    "def worker(x, y, j, k, n, cl, gn):\n",
    "           \n",
    "    if n == -1:\n",
    "        sample_x = cl.query('cluster == @j & centroid == True').join(gn)[['genome']].reset_index()\n",
    "        sample_y = cl.query('cluster == @k & centroid == True').join(gn)[['genome']].reset_index()\n",
    "\n",
    "    else:\n",
    "        query_x = cl.query('cluster == @j').join(gn)[['genome']]\n",
    "        if len(query_x) >= n:\n",
    "            sample_x = query_x.sample(n=n).reset_index()\n",
    "        else:\n",
    "            sample_x = query_x.reset_index()\n",
    "\n",
    "        query_y = cl.query('cluster == @k').join(gn)[['genome']]\n",
    "        if len(query_y) >= n:\n",
    "            sample_y = query_y.sample(n=n).reset_index()\n",
    "        else:\n",
    "            sample_y = query_y.reset_index()\n",
    "\n",
    "    dist_mean = np.nanmean(ssd.cdist(sample_x, sample_y, metric = pairmsa))#.mean()\n",
    "    #len_mean = (sample_x['genome'].str.len().mean() + sample_y['genome'].str.len().mean())/2\n",
    "\n",
    "    #return((x,y,dist_mean/len_mean))\n",
    "    return(x,y,dist_mean)\n",
    "\n",
    "\n",
    "def sample_difference(cluster, genome, segment, proc = 8, n = 10):\n",
    "\n",
    "    cl = cluster.query('segment == @segment') \n",
    "    gn = genome.loc[cl.index.values.tolist()]\n",
    "    names = list(set(cl['cluster']).difference({-1}))\n",
    "    num = len(names)\n",
    "        \n",
    "    params = []\n",
    "    dist_matrix = np.empty([num, num])\n",
    "    dist_matrix[:] = np.nan\n",
    "\n",
    "    for x, j in enumerate(names):\n",
    "        for y, k in enumerate(names[:names.index(j)+1]):\n",
    "            params.append((x, y, j, k, n, cl, gn))\n",
    "\n",
    "    with Pool(processes=proc) as pool:\n",
    "        result = pool.starmap(worker, params)\n",
    "\n",
    "    for x, y, dist in result:\n",
    "        dist_matrix[x,y] = dist\n",
    "        #dist_matrix[y,x] = dist\n",
    "    \n",
    "    dist_dataframe = pd.DataFrame(dist_matrix, index = names, columns = names)\n",
    "    \n",
    "    return(dist_dataframe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "radical-handy",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_x = cluster.query('segment == 4 & cluster == 2').join(genome)[['genome']].sample(n=10, random_state=2).reset_index()\n",
    "sample_y = cluster.query('segment == 4 & cluster == 2').join(genome)[['genome']].sample(n=10, random_state=2).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "japanese-college",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "nan"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ssd.cdist(sample_x, sample_y, metric = pairmsa).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "applicable-crack",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.0"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.array([1,2,3,4,5]).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "muslim-criterion",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.0"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.nanmean(np.array([1,2,np.nan,4,5]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "grand-fleet",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;CY116899</td>\n",
       "      <td>AGACAATGACTATCACTTTTCTCATCCTCCTATTCACAATAGTGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&gt;MK236924</td>\n",
       "      <td>GGTTATACCATAGACAACCGAACAAAGACAATGACTATCACTTTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>&gt;CY139721</td>\n",
       "      <td>GGTTATACCATGGACAACCGAACAAAGACAATGACCATCACCTTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>&gt;CY116979</td>\n",
       "      <td>GGTTATACCATGAACAACCAAACAAAGACAATGACTATCACTTTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&gt;MH501497</td>\n",
       "      <td>GGTTATACCATAGACAACCGAACAAAGACAATGACCATCACTTTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>&gt;CY101555</td>\n",
       "      <td>GGTTATACCATGGACAACCGAACAAAGACAATGACTATCACTTTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>&gt;FJ686754</td>\n",
       "      <td>GGTTATACCATGGACAACCAAACAAAGACAATGACTATCACTTTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>&gt;CY045287</td>\n",
       "      <td>GACAATGACCATCATTTTTCTCATCCTCATGTTCACAGTAGTGAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>&gt;CY202794</td>\n",
       "      <td>GGTTATACCATAGACAACCGAACAAAGAAAATGACCATCACTTTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>&gt;LC042003</td>\n",
       "      <td>ATGACTATCACTTTTCTCATCCTCCTGTTCACAGTAGTGAAAGGGG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   accession                                             genome\n",
       "0  >CY116899  AGACAATGACTATCACTTTTCTCATCCTCCTATTCACAATAGTGAA...\n",
       "1  >MK236924  GGTTATACCATAGACAACCGAACAAAGACAATGACTATCACTTTTC...\n",
       "2  >CY139721  GGTTATACCATGGACAACCGAACAAAGACAATGACCATCACCTTTC...\n",
       "3  >CY116979  GGTTATACCATGAACAACCAAACAAAGACAATGACTATCACTTTTC...\n",
       "4  >MH501497  GGTTATACCATAGACAACCGAACAAAGACAATGACCATCACTTTTC...\n",
       "5  >CY101555  GGTTATACCATGGACAACCGAACAAAGACAATGACTATCACTTTTC...\n",
       "6  >FJ686754  GGTTATACCATGGACAACCAAACAAAGACAATGACTATCACTTTTC...\n",
       "7  >CY045287  GACAATGACCATCATTTTTCTCATCCTCATGTTCACAGTAGTGAAA...\n",
       "8  >CY202794  GGTTATACCATAGACAACCGAACAAAGAAAATGACCATCACTTTTC...\n",
       "9  >LC042003  ATGACTATCACTTTTCTCATCCTCCTGTTCACAGTAGTGAAAGGGG..."
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "incorrect-geography",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
