\section{K-Mer Frequency} \label{sec:Frequency}

The FASTA file containing the to be clustered genomes of \gls{IAV} were translated to vectors for clustering in high dimension by counting their 7-mer frequency (\autoref{fig:Vectorization_Pipeline} \textsf{\textbf{A}}). Let $a_n$ be a mathematical sequence of all permutations with repetition of the set $\{A,C,G,T\}$ to the magnitude of seven, thus $4^7$ elements (\autoref{eq:alphabet}) and $s_n$ a mathematical sequence of all genome sequences of one segment of \gls{IAV} in the FASTA file (\autoref{eq:alphabet}).

\begin{empheq}{alignat = -1}
    &a_n &&= ( \text{AAAAAAA} , \text{AAAAAAC} , \text{AAAAAAG} , \ldots , \text{TTTTTTT} &&)\label{eq:alphabet}\\
    &s_n &&= ( \text{GCAAAA...} , \text{GCAAAA...} , \text{AGCAAA...} , \ldots , \text{ATGGC...} &&)\label{eq:sequences}
\end{empheq}

The elements of $a_n$ represent all possible constellations of 7-mers. The vector component $x_{i,j}$ of a genome sequence $s_i$ with the length $l$ were calculated by summing up the number of occurrences of element $a_j$ in all the 7-mers of genome sequence $s_i$ (\autoref{eq:frequency}). 

\begin{empheq}{alignat = -1}
    &x_{i,j} = \sum^{l-7+1}_{k=1} \delta(s_i[k\ldots k+7-1], a_j), \ \text{ with }\delta(s,a)=\left\{ \begin{array}{c}1\text{ if \ensuremath{s=a}}\\0\text{ if \ensuremath{s\neq a}}\end{array}\right.\label{eq:frequency}
\end{empheq}

This calculation was repeated for all of the $4^7$ elements of $a_n$, as well as for all sequences of $s_n$, building the matrix $\mathbf{X} = [ x_{i,j} ]$, holding the k-mer frequency vectors of the genome sequences of $s_n$ (\autoref{eq:full_matrix}).

\begin{empheq}{alignat = -1}
    &\mathbf{X} = \begin{bmatrix}x_{1,1} & x_{1,2} & x_{1,3} & \dots & x_{1,4^7}\\
    x_{2,1} & x_{2,2} & x_{2,3} & \dots & x_{2,4^7}\\
    x_{3,1} & x_{3,2} & x_{3,3} & \dots & x_{3,4^7}\\
    \vdots & \vdots & \vdots & \ddots & \vdots\\
    x_{i,1} & x_{i,2} & x_{i,3} & \dots & x_{i,4^7}
    \end{bmatrix}\label{eq:full_matrix}
\end{empheq}

The entries in the $j$ columns represent the number of occurencences of 7-mer $a_j$ in a genome sequence $s_i$. The rows represent the $i$ genome sequences from $s_n$ with it's frequency numbers. In summary e.g. $x_{1,1}$ represents the number of occurences of the first 7-mer in $a_n$ in the first genome sequence of $s_n$.

A k-mer length of 7 was used to make the method usable on machines with at least 32Gb of RAM while still being as accurate as possible. With 7-mers, $4^7$ possible combinations can occure, which has a lower chance of accidental matching combinations by e.g. \glspl{SNP} then using 6-mers with only $4^6$ different possible constellations. Using 8-mers on the other hand increases the necessary RAM drastically, by building a matrix e.g. for segment 4, $\mathbf{X}$ of size $56617 \times 4^8$ of 64 bit float numbers making $\approx$ 29.684Gb. Adding the necessary space for dimension reduction methods and other calculations 32Gb would be exceeded by far. 