\section{Hierarchical Density-Based Spatial Clustering of Applications with Noise} \label{sec:HDBSCAN}

\gls{HDBSCAN} is an clustering algorithm proposed by \textcite{campello_hierarchical_2015, hutchison_density-based_2013}. It is a novel version of \gls{DBSCAN} but executed with varying values of $\varepsilon$ \autocite{hutchison_density-based_2013}. Thus not one specific threshold is used to define the clusters, but instead clusters of varying densities are extracted based on their stability over epsilon \autocite{mcinnes_hdbscan_2017}. The version of \gls{HDBSCAN} used in this project is the accelerated implementation by the \textbf{hdbscan} package version 0.8.26 of \textcite{mcinnes_accelerated_2017} (\autoref{fig:Clustering_Pipeline} \textsf{\textbf{F}}).

This is the first time \gls{HDBSCAN} with hybrid clustering setting is, to my best knowledge, used with k-mer frequency genome vectors for large scale clustering of \gls{IAV}. Therefore, most settings for \gls{HDBSCAN} were selected by parameter exploration and graphical comparisons in progress of this project and discussed in the following. However, for the selection of all settings, the \href{https://hdbscan.readthedocs.io/en/latest/api.html}{API} was used as backup and reason \autocite{mcinnes_hdbscan_2017}. 

To capture, in the best case, even genomes with single \glspl{SNP} at rare appearing positions and capture them in their own clusters, \colorbox{backcolour}{min\_cluster\_size=2} and \colorbox{backcolour}{min\_samples=1} setting was used as default in the project. These settings are in both cases the smallest values possible, to declare as least points as possible as noise. The default \colorbox{backcolour}{metric='euclidean'} setting was used for all following executions of \gls{HDBSCAN}, with an exception on the precalculated runs, to approximate the use of cosine distance metric as described in \autoref{sec:Normalization}. Therefore, matrix $\mathbf{X}_{\text{PCA}}$ and $\mathbf{Y}_{\text{UMAP}}$ have to be L2-norm normalized as described in \autoref{sec:Normalization} (\autoref{eq:l2_func_x} and \autoref{eq:l2_func_y}). For calculation of the linkage matrix $\mathbf{L}$, standard \gls{HDBSCAN} without $\varepsilon$ was used.

\autoref{eq:HDB} to \autoref{eq:HDB_link_X} demonstrate the use of \gls{HDBSCAN} to calculate the linkage matrix $\mathbf{L}_{\text{PCA}}$ with matrix $\mathbf{X}_{\text{L2}}$ of \autoref{fig:Clustering_Pipeline} pathway \textsf{\textbf{1}} and was performed in the same way for $\mathbf{L}_{\text{UMAP}}$ with $\mathbf{Y}_{\text{L2}}$ of \autoref{fig:Clustering_Pipeline} pathway \textsf{\textbf{2}} \autocite{gower_minimum_1969, mcinnes_hdbscan_2017}.

\begin{empheq}{alignat = -1}
    &\mathbf{X}_{\text{L2}} &&= \text{NORMALIZE}_{\text{L2}}(\mathbf{X}_{\text{PCA}})\label{eq:l2_func_x}\\
    &\mathbf{Y}_{\text{L2}} &&= \text{NORMALIZE}_{\text{L2}}(\mathbf{Y}_{\text{UMAP}})\label{eq:l2_func_y}
\end{empheq}

\begin{empheq}{alignat = -1}
    &\mathbf{L}_{\text{PCA}} &&= \text{HDBSCAN}_{\text{Link}}(\mathbf{X}_{\text{L2}}, \text{min\_cluster\_size}, \text{min\_samples})\label{eq:HDB}\\
    &&&= \text{HDBSCAN}_{\text{Link}}(\mathbf{X}_{\text{L2}}, 2, 1) \label{eq:HDB_link_X}
\end{empheq}

The mutual reachability distance is, thereby, the degree of difference of two vectors used by \gls{HDBSCAN}. The \colorbox{backcolour}{metric='euclidean'} setting mentioned above, is part of it's calculation and the overall result, when the core distances are smaller (\autoref{eq:reach}). The core distance is the minimum radius necessary to include $k$ other points. The standard setting of $k$ is five and not changed in this project \autocite{mcinnes_hdbscan_2017}. 

\begin{empheq}{alignat = -1}
    &d_{\text{mreach}-k}(\mathbf{x},\mathbf{y}) &&= \max \{ \text{core}_k(\mathbf{x}), \text{core}_k(\mathbf{y}), d_{\text{eucl}}(\mathbf{x},\mathbf{y}) \} \label{eq:reach}
\end{empheq}

\begin{figure}[!hbt]
    \centering
    \includegraphics[width=\textwidth]{Graphics/HDB.pdf}
    \caption[Mutual Reachability Calculation]{\textbf{Mutual Reachability Calculation.} A low dimension representation of the calculation \gls{HDBSCAN} performed in this project. To calculate the mutual reachability distance, a radius is measured necessary to include the next five points, as example for vector $\mathbf{x}$ in blue and $\mathbf{y}$ in red. The euclidean distance between these vectors is then calculated and compared to the radii. The maximum of both radii and the euclidean distance is the mutual reachability distance (\autoref{eq:reach}).}
    \label{fig:HDB}
\end{figure}

By combining \gls{HDBSCAN} with \gls{DBSCAN}, some of the disadvantages of using either of these methods can be avoided \autocite{mcinnes_hdbscan_2017, moulavi_density-based_2014}. Since \gls{DBSCAN} is a hierarchical clustering tool, it is dependent on a strict threshold for clustering. Points not included in this threshold value are omitted and not clustered \autocite{ester_density-based_1996, schubert_dbscan_2017}. \gls{HDBSCAN}, on the other hand, tend to create unwanted micro-clusters in areas of high density \autocite{mcinnes_hdbscan_2017}. Using the hybrid approach proposed in \autocite{malzer_hybrid_2020}, a threshold value $\varepsilon$ can be used to extract these high density areas as single clusters with \gls{DBSCAN}, but still use the method of \gls{HDBSCAN} for the otherwise omitted points (\autoref{fig:Hybrid}). This method is useful when having a small cluster size value, while still aiming to cluster high-density areas together exactly suitable for the proposed clustering. Specific strains of \gls{IAV} are sequenced a lot more, thereby probably creating high-density areas that should be clustered together with \gls{DBSCAN}. The small cluster size is, thereby, used to find small clusters of rare sequenced variants by \gls{HDBSCAN}, with possibly important mutations in low-density areas \autocite{malzer_hybrid_2020}.

\begin{figure}[!hbt]
    \centering
    \includegraphics[width=\textwidth]{Graphics/Hybrid.pdf}
    \caption[Hybrid Clustering Threshold]{\textbf{Hybrid Clustering Threshold.} The hybrid clustering differentiate between clusters where points are connected by distances smaller and higher than $\varepsilon$. When the distance is smaller, the \gls{DBSCAN} algorithm is used and clusters are calculated based on this threshold value $\varepsilon$ by combining points with less distance. For single points, having no point in reachable distance and, therefore, impossible to be clustered by \gls{DBSCAN}, other than single sequence cluster, \gls{HDBSCAN} is used to builds cluster with higher threshold if appropriate. The graphic is based on \glqq Combining HDBSCAN* with DBSCAN\grqq{} in the \href{https://hdbscan.readthedocs.io/en/latest/api.html}{API} adapted to the calculations in this project, as two dimensional example.}
    \label{fig:Hybrid}
\end{figure}

To find an appropriate value for $\varepsilon$, two different methods were used and compared (\autoref{fig:Clustering_Pipeline} pathway \textsf{\textbf{3}} and \textsf{\textbf{4}}. Using the first method, the \gls{DBCV} exploration in \autoref{fig:Clustering_Pipeline} \textsf{\textbf{G}}, execution of \gls{HDBSCAN} was repeated with different settings for $\varepsilon$ and compared by the \gls{DBCV} to find the value of $\varepsilon$ that maximizes the \gls{DBCV} \autocite{moulavi_density-based_2014}. For these repeated executions of \gls{HDBSCAN}, \colorbox{backcolour}{gen\_min\_span\_tree=True} setting was also necessary, to be able to calculate the \gls{DBCV} with the minimum spanning tree (\autoref{eq:DBCV_X}) \autocite{moulavi_density-based_2014, gower_minimum_1969}. The exploration was executed for $\mathbf{X}_{\text{L2}}$ and $\mathbf{Y}_{\text{L2}}$ to find the optimal $\varepsilon_{\text{PCA}}$ and $\varepsilon_{\text{UMAP}}$.

\begin{empheq}{alignat = -1}
    &\max_{\substack{0 \leq \varepsilon}} \left(\text{HDBSCAN}_{\text{DBCV}}(\mathbf{X}_{\text{L2}}, 2, 1, \varepsilon)\right) = \text{HDBSCAN}_{\text{DBCV}}(\mathbf{X}_{\text{L2}}, 2, 1, \varepsilon') \Rightarrow \varepsilon' = \varepsilon_{\text{PCA}} \label{eq:DBCV_X}
\end{empheq}

Using the second method, the optimal value for $\varepsilon_{\text{UMAP'}}$ and $\varepsilon_{\text{PCA'}}$ were calculated using the Kneedle Algorithm \autoref{sec:Kneedle} \autocite{halko_finding_2010}.

With the optimal values for $\varepsilon$ found by \gls{DBCV} exploration and the Kneedle Algorithm, as well as with the matrix for the \gls{UMAP} settings and the one for the \gls{PCA} settings, \gls{HDBSCAN} with the hybrid clustering setting was executed four times. Each execution resulted in the mathematical sequence of cluster names $N$ related to the sequence of genomic sequences $S$ (\autoref{eq:HDB_cluster_PK} to \autoref{eq:HDB_cluster_UD}) (\autoref{fig:Clustering_Pipeline} \textsf{\textbf{H}}) \autocite{mcinnes_hdbscan_2017, malzer_hybrid_2020}. To sum it up, with the \acrshort{UMAP}/\acrshort{DBCV} method, the cluster name of the first genomic sequence of $S$ is the first element of mathematical sequence $N_{\text{UMAP}}$. 

\begin{empheq}{alignat = -1}
    &N_{\text{PCA}} &&= \text{HDBSCAN}_{\text{Hybrid}}(\mathbf{X}_{\text{L2}}, 2, 1, \varepsilon_{\text{PCA}}) \label{eq:HDB_cluster_PK}\\
    &N_{\text{UMAP}} &&= \text{HDBSCAN}_{\text{Hybrid}}(\mathbf{Y}_{\text{L2}}, 2, 1, \varepsilon_{\text{UMAP}}) \label{eq:HDB_cluster_UK}\\
    &N_{\text{PCA'}} &&= \text{HDBSCAN}_{\text{Hybrid}}(\mathbf{X}_{\text{L2}}, 2, 1, \varepsilon_{\text{PCA'}}) \label{eq:HDB_cluster_PD}\\
    &N_{\text{UMAP'}} &&= \text{HDBSCAN}_{\text{Hybrid}}(\mathbf{Y}_{\text{L2}}, 2, 1, \varepsilon_{\text{UMAP'}}) \label{eq:HDB_cluster_UD}
\end{empheq}

Important parameters used in this project, including settings varying from the default, are listed below. All available settings with explanation are listed in the \href{https://hdbscan.readthedocs.io/en/latest/api.html}{API}.

\begin{leftbar}
    \textbf{hdbscan.HDBSCAN}
    \begin{nstabbing}
        \qquad\qquad\qquad\qquad\qquad\quad\=\kill

        min\_cluster\_size \> [min. size of a cluster (default: 5)]\\
        
        min\_samples \> [conservativeness of the clustering (default: None)]\\
        
        cluster\_selection\_epsilon \> [merge clusters below the threshold (default: 0.0)]\\
        
        gen\_min\_span\_tree \> [generate the minimum spanning tree (default: False)]\\
        
        metric \> [metric to use for clustering (default: euclidean)]
        %alpha \> (default: 1.0)
    \end{nstabbing}
\end{leftbar}

%Clustering
%DBCV mehr als 50dims abkacken bla
%Metrik
%Hybrid selection
%Needle
%Linkage matrix