\chapter{Results and Discussion} \label{chap:Results_and_Discussion}

For a basic overview of the sequences in the FASTA file (\autoref{chap:Materials_and_Methods}), the number of sequences of a segment related to it's specific subtypes was counted and visualized (\autoref{fig:Frequency_4} and \autoref{fig:Frequency_6}). Hereby, the high number of segment 4 sequences of subtype H1 and H3 in \autoref{fig:Frequency_4} and segment 6 sequences of N1 and N2 in \autoref{fig:Frequency_6} is noticeable. Furthermore, the similarity of the number of sequences of segment 4 subtype H1 and N1 as well as of segment 6 H3 and N2 is also remarkable. The similarity seems to be most likely due to the frequency of subtype combinations H1N1 and H3N2 of the full genomes, which is also in line with the most prominent subtypes of \gls{IAV} mentioned in \textcite{deng_simplified_2015}. 

\begin{figure}[!hbt]
    \centering
    %\includegraphics[width=\dimexpr\textwidth-2\fboxsep-2\fboxrule,fbox]{PCA/Data_Overview_Segment_4_H.pdf}
    \includegraphics[width=\textwidth]{PCA/Data_Overview_Segment_4_H.pdf}
    \caption[Segment 4 \Acrlong{HA} Antigen Subtype Frequency]{\textbf{Segment 4 \Acrlong{HA} Antigen Subtype Frequency.} The number of occurrences of a segment 4 sequence related to a specific \gls{HA} antigen subtype is visualized. Sequences related to subtypes H1 and H3 are the most frequent ones from segment 4 with a total number of 17627 and 23074. Subtype H5 is also slightly over-represented with 3195 sequences. 1191 sequences from segment 4 in the data have no classification and are listed as \glqq none\grqq{}.}
    \label{fig:Frequency_4}
\end{figure}

\begin{figure}[!hbt]
    \centering
    %\includegraphics[width=\dimexpr\textwidth-2\fboxsep-2\fboxrule,fbox]{PCA/Data_Overview_Segment_6_N.pdf}
    \includegraphics[width=\textwidth]{PCA/Data_Overview_Segment_6_N.pdf}
    \caption[Segment 6 \Acrlong{NA} Antigen Subtype Frequency]{\textbf{Segment 6 \Acrlong{NA} Antigen Subtype Frequency.} The number of occurrences of a segment 6 sequence related to a specific \gls{NA} antigen subtype is visualized. Sequences related to subtypes N1 and N2 are the most frequent ones from segment 6 with a total number of 17945 and 26129. 1305 sequences from segment 6 in the data have no classification and are listed as \glqq none\grqq{}.}
    \label{fig:Frequency_6}
\end{figure}

In addition to the most frequent subtypes a, high number of not classified or \glqq mixed\grqq{} sequences is present in the data. These unclassified strains were maybe sequenced without prior subtype testing or are in fact sequences that don't belong to either of the known subtypes and, thereby, no antibody antigen reaction was observable. It could also be possible, that the sequencing of these sequences is older than the current used classification of \gls{IAV} \autocite{noauthor_revision_1980}. 

As described in \autoref{chap:Materials_and_Methods}, the clustering was performed multiple times to compare the used methods for their quality and efficiency. The four methods are compared directly to each other in the following. For better separation, the four methods were called by their abbreviation, to be found in \autoref{tab:methods}. As a short reminder, PD uses direct reduction to 30 components with \gls{PCA} following $\varepsilon$ exploration using the \gls{DBCV} and PK likewise using the Kneedle Algorithm. Exploration for method UD and UK works equally with reduction to 100 components by \gls{PCA} and to 30 afterwards by \gls{UMAP}, noticeable by the U in the abbreviation instead of a P.   

To choose a reasonable number of components for the \gls{PCA} mentioned in \autoref{sec:PCA}, the k-mer frequency vectors of segment 4 (\autoref{fig:Vectorization_Pipeline}) were reduced by running \gls{PCA} with different settings. The sum of explained variance for every setting were calculated and listed in \autoref{tab:PCA_Dimension}. Because of the high increase of explained variance from 10 to 20 components with 58.7\% to 68.7\% and also to 30 components with 73.9\% explained variance, a value of 30 was used as default in this project. A fairly small value was used, due to increasing computational effort of the \gls{PCA} implementation \textbf{scikit-learn (sklearn)} and also to preserve the usability of Spanning Tree calculation with \gls{HDBSCAN} which is, with the preferred settings, only possible to $\approx$ 50 components. (\autoref{tab:PCA_Dimension}) \autocite{mcinnes_hdbscan_2017, scipy_10_contributors_scipy_2020}. The number of final components for \gls{UMAP} in UD and UK was, therefore, also set to 30 to be comparable to PD and PK.

\begin{table}[!hbt]
    \centering
    \caption[Explained Variance by different PCA settings]{\textbf{Explained Variance by different PCA settings.} The explained variance of specific \gls{PCA} settings used on the k-mer frequency vectors of segment 4. A result of 0.587 as for 10 components represents 58.7\% of the variance is explained by the first 10 components of the vector. The more components are extracted the lower the increase in additional explained variance \autocite{pearson_liii_1901}.}
    \label{tab:PCA_Dimension}
    \pgfplotstabletypeset[
        every head row/.style={
            before row={
                \toprule
            },
            after row={
                \midrule
            },
        },
        every last row/.style={
            after row={
                \bottomrule
            },
        },
        begin table=\begin{tabular*}{.5\textwidth},
        end table=\end{tabular*},
        columns={0,1},
        columns/0/.style={int detect, multicolumn names=l,column name=\textbf{\#Components}, column type=@{\extracolsep{\fill}\hspace{6pt}}r},
        columns/1/.style={multicolumn names=l,column name=\textbf{Explained Variance}, column type=r}
    ]
    {Graphics/PCA.csv}
\end{table}

\begin{figure}[!hbt]
    \centering
    %\begin{adjustbox}{minipage=\dimexpr\textwidth-2\fboxsep-2\fboxrule,fbox}
    \begin{subfigure}[b]{0.475\textwidth}
        \caption[Dimension Reduction with \Acrshort{PCA}]{\textbf{Dimension Reduction with \Acrshort{PCA}}}
        \label{subfig:Normalisation_PCA}            \includegraphics[width=\textwidth]{PCA/Difference_Distance_Calculation.pdf}
    \end{subfigure}
    \hfill
    \begin{subfigure}[b]{0.475\textwidth}
        \caption[Dimension Reduction with \Acrshort{UMAP}]{\textbf{Dimension Reduction with \Acrshort{UMAP}}}
        \label{subfig:Normalisation_UMAP}            \includegraphics[width=\textwidth]{UMAP/Difference_Distance_Calculation.pdf}
    \end{subfigure}
    %\end{adjustbox}
    \caption[Cosine Distance Approximation]{\textbf{Cosine Distance Approximation.} Calculations for the chord and the euclidean distances were performed on subsets for both pipelines and compared. In both cases the difference is exactly matching the proportional increase of squared chord distance divided by two equals the cosine distance as described in \autoref{sec:Normalization}. Crowding of the points of the graph representing the \gls{UMAP} pipeline is explained by the embedding mechanism of the tool itself \autocite{mcinnes_umap_2020}.}
    \label{fig:Normalisation_Methods}
\end{figure}

For a graphical comparison of the demonstrated mathematical proof of the proportional increase of the chord distance to the cosine distance (\autoref{sec:Normalization}), subsets of the real data were used. In \autoref{subfig:Normalisation_PCA}, a subset of 100 k-mer frequency vectors of segment 4 were reduced by \gls{PCA} and compared to each other by chord and euclidean distance as described in \autoref{sec:Normalization} and \autoref{sec:PCA}. All the calculated distances for chord distance were then plotted against the equivalent euclidean distances. The red line indicates the proportional increase as calculated in \autoref{sec:Normalization}. All the points are arranged exactly along the red line which confirms the calculation and usability of the described method. The same was repeated for the settings with the use of the \gls{UMAP} pipeline (\gls{UMAP} posterior to \gls{PCA} with higher component number) resulting in the same graphical confirmation (\autoref{subfig:Normalisation_UMAP}). In \autoref{subfig:Normalisation_UMAP}, all the points are crowded much more to a dense area of the curve. This is by the embedding behavior of \gls{UMAP} and will be discussed in the following.

\input{Thesis/0301_Clustering}

\input{Thesis/0302_Anomalies}

\input{Thesis/0303_K_mer_Representation}

\input{Thesis/0304_Comparison_Clustering}

\input{Thesis/0305_Dimension_Reduction}

\input{Thesis/0306_Serotype_Classification}